// postcss-pxtrans@1.0.1/dist/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }// src/index.ts
var _postcsspluginshared = require('postcss-plugin-shared');

// src/pixel-unit-regex.ts

function pxRegex(units = ["px"]) {
  return _postcsspluginshared.createUnitRegex.call(void 0, {
    units,
    numberPattern: String.raw`\d*\.?\d+`,
    skipVar: false
  });
}

// src/directives.ts
var postcssPlugin = "postcss-pxtrans-directives";
function removeUntilEndif(comment) {
  let next = comment.next();
  while (next) {
    if (next.type === "comment" && next.text.trim() === "#endif") {
      break;
    }
    const temp = next.next();
    next.remove();
    next = temp;
  }
}
function handleIfdef(comment, platform) {
  const wordList = comment.text.split(" ");
  if (!wordList.includes("#ifdef")) {
    return;
  }
  if (wordList.includes(platform)) {
    return;
  }
  removeUntilEndif(comment);
}
function handleIfndef(comment, platform) {
  const wordList = comment.text.split(" ");
  if (!wordList.includes("#ifndef")) {
    return;
  }
  if (!wordList.includes(platform)) {
    return;
  }
  removeUntilEndif(comment);
}
function handleRnEject(comment) {
  if (comment.text !== "postcss-pxtrans rn eject enable") {
    return;
  }
  let next = comment.next();
  while (next) {
    if (next.type === "comment" && next.text === "postcss-pxtrans rn eject disable") {
      break;
    }
    const temp = next.next();
    next.remove();
    next = temp;
  }
}
function createDirectivePlugin(options = {}) {
  const methods = _nullishCoalesce(options.methods, () => ( ["platform", "size"]));
  const platform = _nullishCoalesce(options.platform, () => ( "weapp"));
  const plugin2 = {
    postcssPlugin,
    prepare(result) {
      const state = { skip: false };
      const root = result.root;
      return {
        Comment(comment) {
          if (comment.text === "postcss-pxtrans disable") {
            state.skip = true;
            root.raws.__pxtransSkip = true;
            return;
          }
          if (!methods.includes("platform")) {
            return;
          }
          if (platform === "rn") {
            handleRnEject(comment);
          }
          handleIfdef(comment, platform);
          handleIfndef(comment, platform);
        },
        OnceExit() {
          if (state.skip && root) {
            root.raws.__pxtransSkip = true;
          }
        }
      };
    }
  };
  return plugin2;
}

// src/index.ts
var defaults = {
  methods: ["platform", "size"],
  rootValue: 16,
  unitPrecision: 5,
  selectorBlackList: [],
  propList: ["*"],
  replace: true,
  mediaQuery: false,
  minPixelValue: 0
};
var deviceRatio = {
  375: 2,
  640: 2.34 / 2,
  750: 1,
  828: 1.81 / 2
};
var DEFAULT_WEAPP_OPTIONS = {
  platform: "weapp",
  designWidth: 750,
  deviceRatio
};
var processed = /* @__PURE__ */ Symbol("processed");
var SPECIAL_PIXEL = ["Px", "PX", "pX"];
var SPECIAL_PIXEL_SET = new Set(SPECIAL_PIXEL);
var pxTestRE = /px/i;
function normalizeRootValue(value) {
  if (typeof value === "function") {
    return value;
  }
  if (typeof value === "number") {
    return () => value;
  }
  return () => defaults.rootValue;
}
function createRoundWithPrecision(precision) {
  if (!Number.isFinite(precision) || precision < 0 || precision > 100) {
    return (value) => value;
  }
  const multiplier = 10 ** (precision + 1);
  const divider = multiplier * 0.1;
  return (value) => {
    const wholeNumber = Math.floor(value * multiplier);
    return Math.round(wholeNumber / 10) / divider;
  };
}
function createPxReplace(rootValue, unitPrecision, minPixelValue, onePxTransform, platform, targetUnit, unConvertTargetUnit) {
  const specialUnits = SPECIAL_PIXEL_SET;
  const isHarmony = platform === "harmony";
  const preserveUnit = _nullishCoalesce(unConvertTargetUnit, () => ( targetUnit));
  const roundWithPrecision = createRoundWithPrecision(unitPrecision);
  return function createFileReplace(input) {
    return function replace(m, $1) {
      if (!$1) {
        return m;
      }
      if (isHarmony && specialUnits.has(m.slice(-2))) {
        return `${$1}${preserveUnit}`;
      }
      const pixels = Number.parseFloat($1);
      if (!onePxTransform && pixels === 1) {
        return isHarmony ? `${$1}${preserveUnit}` : m;
      }
      if (pixels < minPixelValue) {
        return isHarmony ? `${$1}${preserveUnit}` : m;
      }
      let val = pixels / rootValue(input, m, $1);
      val = roundWithPrecision(val);
      return `${val}${targetUnit}`;
    };
  };
}
var postcssPlugin2 = "postcss-pxtrans";
function plugin(userOptions = {}) {
  const options = {
    ...DEFAULT_WEAPP_OPTIONS,
    ...userOptions
  };
  const exclude = options.exclude;
  const transUnits = ["px"];
  const minRootSize = _nullishCoalesce(options.minRootSize, () => ( 0));
  const baseFontSize = _nullishCoalesce(options.baseFontSize, () => ( (minRootSize >= 1 ? minRootSize : 20)));
  const designWidth = (input) => typeof options.designWidth === "function" ? options.designWidth(input) : options.designWidth;
  const deviceRatioValue = (input) => _nullishCoalesce(options.deviceRatio[designWidth(input)], () => ( 1));
  let targetUnit;
  let unConvertTargetUnit;
  let computedRootValue;
  const platform = options.platform;
  switch (platform) {
    case "h5": {
      targetUnit = _nullishCoalesce(options.targetUnit, () => ( "rem"));
      switch (targetUnit) {
        case "vw":
        case "vmin":
          computedRootValue = (input) => designWidth(input) / 100;
          break;
        case "px":
          computedRootValue = (input) => 1 / deviceRatioValue(input) * 2;
          break;
        default:
          computedRootValue = (input) => baseFontSize / deviceRatioValue(input) * 2;
          break;
      }
      transUnits.push("rpx");
      break;
    }
    case "rn": {
      computedRootValue = (input) => 1 / deviceRatioValue(input) * 2;
      targetUnit = "px";
      break;
    }
    case "quickapp": {
      computedRootValue = () => 1;
      targetUnit = "px";
      break;
    }
    case "harmony": {
      computedRootValue = (input) => 1 / deviceRatioValue(input);
      targetUnit = "px";
      unConvertTargetUnit = "ch";
      transUnits.push(...SPECIAL_PIXEL);
      break;
    }
    default: {
      targetUnit = _nullishCoalesce(options.targetUnit, () => ( "rpx"));
      if (targetUnit === "rem") {
        computedRootValue = (input) => baseFontSize / deviceRatioValue(input) * 2;
      } else if (targetUnit === "px") {
        computedRootValue = (input) => 1 / deviceRatioValue(input) * 2;
      } else {
        computedRootValue = (input) => 1 / deviceRatioValue(input);
      }
    }
  }
  const resolvedRootValue = typeof userOptions.rootValue !== "undefined" ? userOptions.rootValue : _nullishCoalesce(computedRootValue, () => ( defaults.rootValue));
  const cacheComputedRootValue = typeof userOptions.rootValue === "undefined" && typeof resolvedRootValue === "function" && resolvedRootValue.length <= 1;
  const opts = {
    ...defaults,
    ...options,
    rootValue: resolvedRootValue
  };
  const onePxTransform = typeof options.onePxTransform === "undefined" ? true : options.onePxTransform;
  const pxRgx = pxRegex(transUnits);
  const satisfyPropList = _postcsspluginshared.createAdvancedPropListMatcher.call(void 0, _nullishCoalesce(opts.propList, () => ( defaults.propList)));
  const unitPrecision = _nullishCoalesce(opts.unitPrecision, () => ( defaults.unitPrecision));
  const minPixelValue = _nullishCoalesce(opts.minPixelValue, () => ( defaults.minPixelValue));
  const shouldHandleSize = opts.methods.includes("size");
  const shouldHandleMedia = shouldHandleSize && Boolean(opts.mediaQuery);
  const hasSelectorBlackList = Boolean(_optionalChain([opts, 'access', _ => _.selectorBlackList, 'optionalAccess', _2 => _2.length]));
  return {
    postcssPlugin: postcssPlugin2,
    prepare(result) {
      const root = result.root;
      const input = _optionalChain([root, 'optionalAccess', _3 => _3.source, 'optionalAccess', _4 => _4.input]);
      if (!input) {
        return {};
      }
      root.walkComments((comment) => {
        if (comment.text === "postcss-pxtrans disable") {
          root.raws.__pxtransSkip = true;
          return false;
        }
        return void 0;
      });
      const rootValue = cacheComputedRootValue ? (() => {
        const cached = opts.rootValue(input, "", "");
        return () => cached;
      })() : normalizeRootValue(opts.rootValue);
      const pxReplace = createPxReplace(
        rootValue,
        unitPrecision,
        minPixelValue,
        onePxTransform,
        platform,
        targetUnit,
        unConvertTargetUnit
      )(input);
      const shouldSkip = () => Boolean(_optionalChain([result, 'access', _5 => _5.root, 'optionalAccess', _6 => _6.raws, 'optionalAccess', _7 => _7.__pxtransSkip]));
      const blacklistedCache = hasSelectorBlackList ? /* @__PURE__ */ new WeakMap() : null;
      const isExcluded = exclude && exclude(result.opts.from);
      if (shouldSkip()) {
        return {};
      }
      if (isExcluded) {
        return {};
      }
      return {
        Declaration(decl) {
          if (shouldSkip()) {
            return;
          }
          if (!shouldHandleSize) {
            return;
          }
          if (decl[processed]) {
            return;
          }
          ;
          decl[processed] = true;
          if (!pxTestRE.test(decl.value)) {
            return;
          }
          if (!satisfyPropList(decl.prop)) {
            return;
          }
          const parent = decl.parent;
          let isBlacklisted = false;
          if (hasSelectorBlackList && parent) {
            const cached = _optionalChain([blacklistedCache, 'optionalAccess', _8 => _8.get, 'call', _9 => _9(parent)]);
            if (cached !== void 0) {
              isBlacklisted = cached;
            } else {
              isBlacklisted = _postcsspluginshared.blacklistedSelector.call(void 0, 
                opts.selectorBlackList,
                parent.selector
              );
              _optionalChain([blacklistedCache, 'optionalAccess', _10 => _10.set, 'call', _11 => _11(parent, isBlacklisted)]);
            }
          }
          if (isBlacklisted && platform !== "harmony") {
            return;
          }
          let value;
          if (isBlacklisted) {
            pxRgx.lastIndex = 0;
            value = decl.value.replace(pxRgx, (m, $1) => {
              return $1 ? `${$1}${unConvertTargetUnit}` : m;
            });
          } else {
            pxRgx.lastIndex = 0;
            value = decl.value.replace(pxRgx, pxReplace);
          }
          if (parent && _postcsspluginshared.declarationExists.call(void 0, parent, decl.prop, value)) {
            return;
          }
          if (opts.replace) {
            decl.value = value;
          } else {
            decl.cloneAfter({ value });
          }
        },
        AtRule: {
          media(rule) {
            if (!shouldHandleMedia) {
              return;
            }
            if (shouldSkip()) {
              return;
            }
            if (!pxTestRE.test(rule.params)) {
              return;
            }
            pxRgx.lastIndex = 0;
            rule.params = rule.params.replace(pxRgx, pxReplace);
          }
        }
      };
    }
  };
}
plugin.postcss = true;
var index_default = plugin;



exports.createDirectivePlugin = createDirectivePlugin; exports.default = index_default;
