// @weapp-tailwindcss/postcss@2.1.5/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; } var _class;require('./chunk-GGNOJ77I.js');

// src/handler.ts
var _shared = require('@weapp-tailwindcss/shared');

// src/defaults.ts
function getDefaultOptions(options) {
  return {
    // 参考：https://github.com/postcss/postcss-calc
    cssPresetEnv: {
      features: {
        "cascade-layers": true,
        "is-pseudo-class": {
          specificityMatchingName: "weapp-tw-ig"
        },
        "oklab-function": true,
        "color-mix": true,
        "color-functional-notation": _nullishCoalesce(_optionalChain([options, 'optionalAccess', _ => _.cssPresetEnv, 'optionalAccess', _2 => _2.features, 'optionalAccess', _3 => _3["color-functional-notation"]]), () => ( { preserve: false })),
        // 在 calc 下，这个需要开启
        "custom-properties": _nullishCoalesce(_optionalChain([options, 'optionalAccess', _4 => _4.cssPresetEnv, 'optionalAccess', _5 => _5.features, 'optionalAccess', _6 => _6["custom-properties"]]), () => ( _optionalChain([options, 'optionalAccess', _7 => _7.cssCalc]))) ? { preserve: true } : false
      },
      autoprefixer: {
        add: false
      }
    },
    // 支付宝小程序不支持，所以默认关闭
    cssRemoveProperty: true,
    // cssRemoveAtSupports: true,
    // cssRemoveAtMedia: true,
    cssSelectorReplacement: {
      root: ["page", ".tw-root", "wx-root-portal-content"],
      universal: ["view", "text"]
    }
  };
}

// src/options-resolver.ts


// src/fingerprint.ts
function fingerprintOptions(value, state = { map: /* @__PURE__ */ new WeakMap(), counter: 0 }) {
  if (value === null || value === void 0) {
    return String(value);
  }
  if (typeof value === "function") {
    return `fn:${value.name || "anonymous"}`;
  }
  if (typeof value === "symbol") {
    return `sym:${String(value)}`;
  }
  if (typeof value !== "object") {
    return `${typeof value}:${String(value)}`;
  }
  const objectValue = value;
  const cached = state.map.get(objectValue);
  if (cached) {
    return cached;
  }
  const marker = `ref:${state.counter++}`;
  state.map.set(objectValue, marker);
  if (Array.isArray(objectValue)) {
    const parts2 = objectValue.map((entry) => fingerprintOptions(entry, state));
    return `[${parts2.join(",")}]`;
  }
  const keys = Object.keys(objectValue).sort();
  const parts = keys.map((key) => `${key}:${fingerprintOptions(objectValue[key], state)}`);
  return `{${parts.join(",")}}@${marker}`;
}

// src/options-resolver.ts
var BASE_CACHE_KEY = "base";
function hasOverrides(options) {
  return Boolean(options && Object.keys(options).length > 0);
}
function createOptionsResolver(baseOptions) {
  const cacheByKey = /* @__PURE__ */ new Map();
  const cacheByRef = /* @__PURE__ */ new WeakMap();
  const fingerprintByRef = /* @__PURE__ */ new WeakMap();
  cacheByKey.set(BASE_CACHE_KEY, baseOptions);
  const resolve = (overrides) => {
    if (!hasOverrides(overrides)) {
      return baseOptions;
    }
    const refCached = cacheByRef.get(overrides);
    if (refCached) {
      return refCached;
    }
    let key = fingerprintByRef.get(overrides);
    if (!key) {
      key = fingerprintOptions(overrides);
      fingerprintByRef.set(overrides, key);
    }
    const cached = cacheByKey.get(key);
    if (cached) {
      cacheByRef.set(overrides, cached);
      return cached;
    }
    const merged = _shared.defuOverrideArray.call(void 0, 
      { ...overrides },
      baseOptions
    );
    cacheByKey.set(key, merged);
    cacheByRef.set(overrides, merged);
    return merged;
  };
  return {
    resolve
  };
}

// src/preflight.ts
function createInjectPreflight(options) {
  const result = [];
  if (options && typeof options === "object") {
    const entries = Object.entries(options);
    for (const [prop, value] of entries) {
      if (value !== false) {
        result.push({
          prop,
          value: value.toString()
        });
      }
    }
  }
  return () => {
    return result;
  };
}

// src/processor-cache.ts
var _postcss = require('postcss'); var _postcss2 = _interopRequireDefault(_postcss);

// src/pipeline.ts
var _postcsspresetenv = require('postcss-preset-env'); var _postcsspresetenv2 = _interopRequireDefault(_postcsspresetenv);

// src/plugins/colorFunctionalFallback.ts
var _postcssvalueparser = require('postcss-value-parser'); var _postcssvalueparser2 = _interopRequireDefault(_postcssvalueparser);
var RGB_FUNCTION_NAME = "rgb";
var isSlashDiv = (node) => node.type === "div" && node.value === "/";
function trimNodes(nodes2) {
  let start = 0;
  let end = nodes2.length;
  while (start < end && _optionalChain([nodes2, 'access', _8 => _8[start], 'optionalAccess', _9 => _9.type]) === "space") {
    start += 1;
  }
  while (end > start && _optionalChain([nodes2, 'access', _10 => _10[end - 1], 'optionalAccess', _11 => _11.type]) === "space") {
    end -= 1;
  }
  return nodes2.slice(start, end);
}
function splitColorComponents(nodes2) {
  const parts = [];
  let current = [];
  nodes2.forEach((node) => {
    if (node.type === "div" && node.value === ",") {
      if (current.length) {
        parts.push(current);
        current = [];
      }
      return;
    }
    if (node.type === "space") {
      if (current.length) {
        parts.push(current);
        current = [];
      }
      return;
    }
    current.push(node);
  });
  if (current.length) {
    parts.push(current);
  }
  return parts;
}
var buildFunctionNodes = (content) => _postcssvalueparser2.default.call(void 0, content).nodes;
function convertRgbFunction(fnNode) {
  const slashIndex = fnNode.nodes.findIndex(isSlashDiv);
  if (slashIndex === -1) {
    return false;
  }
  const colorNodes = trimNodes(fnNode.nodes.slice(0, slashIndex));
  const alphaNodes = trimNodes(fnNode.nodes.slice(slashIndex + 1));
  if (!colorNodes.length || !alphaNodes.length) {
    return false;
  }
  const colorParts = splitColorComponents(colorNodes);
  if (colorParts.length !== 3) {
    return false;
  }
  const normalizedColors = colorParts.map((part) => _postcssvalueparser2.default.stringify(part).trim()).filter(Boolean);
  if (normalizedColors.length !== 3) {
    return false;
  }
  const alphaText = _postcssvalueparser2.default.stringify(alphaNodes).trim();
  if (!alphaText) {
    return false;
  }
  fnNode.value = "rgba";
  const nextValue = `${normalizedColors.join(", ")}, ${alphaText}`;
  fnNode.nodes = buildFunctionNodes(nextValue);
  return true;
}
function transformDeclaration(decl) {
  if (!decl.value || !decl.value.toLowerCase().includes(RGB_FUNCTION_NAME)) {
    return;
  }
  const parsed = _postcssvalueparser2.default.call(void 0, decl.value);
  let mutated = false;
  parsed.walk((node) => {
    if (node.type !== "function") {
      return;
    }
    if (node.value.toLowerCase() !== RGB_FUNCTION_NAME) {
      return;
    }
    if (convertRgbFunction(node)) {
      mutated = true;
    }
  });
  if (mutated) {
    decl.value = parsed.toString();
  }
}
function createColorFunctionalFallback() {
  return {
    postcssPlugin: "weapp-tailwindcss-color-functional-fallback",
    Declaration: transformDeclaration
  };
}
createColorFunctionalFallback.postcss = true;

// src/plugins/ctx.ts
function createContext() {
  const variablesScopeWeakMap = /* @__PURE__ */ new WeakMap();
  function isVariablesScope(rule) {
    return variablesScopeWeakMap.get(rule) === true;
  }
  function markVariablesScope(rule) {
    variablesScopeWeakMap.set(rule, true);
  }
  return {
    variablesScopeWeakMap,
    isVariablesScope,
    markVariablesScope
  };
}

// src/plugins/getCalcDuplicateCleaner.ts
function getCalcDuplicateCleaner(options) {
  if (!options.cssCalc) {
    return null;
  }
  return {
    postcssPlugin: "postcss-calc-duplicate-cleaner",
    Rule(rule) {
      rule.walkDecls((decl) => {
        const prev = decl.prev();
        if (!prev || prev.type !== "decl") {
          return;
        }
        if (prev.prop !== decl.prop) {
          return;
        }
        if (prev.important !== decl.important) {
          return;
        }
        if (prev.value !== decl.value) {
          return;
        }
        decl.remove();
      });
    }
  };
}

// src/plugins/getCalcPlugin.ts
var _postcsscalc = require('@weapp-tailwindcss/postcss-calc'); var _postcsscalc2 = _interopRequireDefault(_postcsscalc);

// ../../node_modules/.pnpm/es-toolkit@1.44.0/node_modules/es-toolkit/dist/object/omit.mjs
function omit(obj, keys) {
  const result = { ...obj };
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    delete result[key];
  }
  return result;
}

// src/plugins/getCalcPlugin.ts
function getCalcPlugin(options) {
  if (!options.cssCalc) {
    return null;
  }
  const calcOptions = Array.isArray(options.cssCalc) ? {} : typeof options.cssCalc === "object" ? omit(options.cssCalc, ["includeCustomProperties"]) : {};
  return _postcsscalc2.default.call(void 0, calcOptions);
}

// src/plugins/getCustomPropertyCleaner.ts


function getCustomPropertyCleaner(options) {
  const includeCustomProperties = Array.isArray(options.cssCalc) ? options.cssCalc : typeof options.cssCalc === "object" ? options.cssCalc.includeCustomProperties : [];
  const shouldMatchCustomProperties = Array.isArray(includeCustomProperties) && includeCustomProperties.length > 0;
  if (!shouldMatchCustomProperties) {
    return null;
  }
  return {
    postcssPlugin: "postcss-remove-include-custom-properties",
    OnceExit(root) {
      root.walkDecls((decl) => {
        const prevNode = decl.prev();
        if (!prevNode || prevNode.type !== "decl" || prevNode.prop !== decl.prop) {
          return;
        }
        if (prevNode.value === decl.value) {
          decl.remove();
          return;
        }
        if (!shouldMatchCustomProperties || !/--/.test(decl.value)) {
          return;
        }
        const parsed = _postcssvalueparser2.default.call(void 0, decl.value);
        let containsIncludedCustomProperty = false;
        parsed.walk((node) => {
          if (node.type !== "function" || node.value !== "var" || containsIncludedCustomProperty) {
            return;
          }
          const match = node.nodes.find((x) => {
            return x.type === "word" && _shared.regExpTest.call(void 0, includeCustomProperties, x.value);
          });
          if (match) {
            containsIncludedCustomProperty = true;
          }
        });
        if (containsIncludedCustomProperty) {
          decl.remove();
        }
      });
    }
  };
}

// src/plugins/getPxTransformPlugin.ts

var _postcsspxtrans = require('postcss-pxtrans'); var _postcsspxtrans2 = _interopRequireDefault(_postcsspxtrans);
var defaultPxTransformOptions = {
  platform: "weapp",
  targetUnit: "rpx",
  unitPrecision: 5,
  propList: ["*"],
  selectorBlackList: [],
  replace: true,
  mediaQuery: false,
  minPixelValue: 0,
  designWidth: 750,
  deviceRatio: {
    375: 2,
    640: 2.34 / 2,
    750: 1,
    828: 1.81 / 2
  }
};
function getPxTransformPlugin(options) {
  if (!options.px2rpx) {
    return null;
  }
  const userOptions = typeof options.px2rpx === "object" ? options.px2rpx : {};
  return _postcsspxtrans2.default.call(void 0, 
    _shared.defuOverrideArray.call(void 0, 
      userOptions,
      defaultPxTransformOptions
    )
  );
}

// src/plugins/getRemTransformPlugin.ts

var _postcssremtoresponsivepixel = require('postcss-rem-to-responsive-pixel'); var _postcssremtoresponsivepixel2 = _interopRequireDefault(_postcssremtoresponsivepixel);
var defaultRemOptions = {
  rootValue: 32,
  propList: ["*"],
  transformUnit: "rpx"
};
var defaultStage = {
  processorStage: "OnceExit"
};
function getRemTransformPlugin(options) {
  if (!options.rem2rpx) {
    return null;
  }
  const userOptions = typeof options.rem2rpx === "object" ? options.rem2rpx : defaultRemOptions;
  const merged = _shared.defuOverrideArray.call(void 0, 
    userOptions,
    defaultStage
  );
  return _postcssremtoresponsivepixel2.default.call(void 0, merged);
}

// src/plugins/getUnitsToPxPlugin.ts
var _postcssunitstopx = require('postcss-units-to-px'); var _postcssunitstopx2 = _interopRequireDefault(_postcssunitstopx);
function getUnitsToPxPlugin(options) {
  if (!options.unitsToPx) {
    return null;
  }
  const userOptions = typeof options.unitsToPx === "object" ? options.unitsToPx : void 0;
  return _postcssunitstopx2.default.call(void 0, userOptions);
}

// src/plugins/post.ts


// src/compat/tailwindcss-rpx.ts
var LENGTH_VALUE_REGEXP = /^[+-]?(?:\d+(?:\.\d+)?|\.\d+)(?:e[+-]?\d+)?rpx$/i;
function normalizeTailwindcssRpxDeclaration(decl, options) {
  const majorVersion = _optionalChain([options, 'optionalAccess', _12 => _12.majorVersion]);
  const normalizedValue = decl.value.trim();
  if (LENGTH_VALUE_REGEXP.test(normalizedValue) && (majorVersion === void 0 || majorVersion === 2 || majorVersion === 3 || majorVersion === 4)) {
    const lowerProp = decl.prop.toLowerCase();
    if (lowerProp === "color") {
      decl.prop = "font-size";
      return true;
    }
    if (lowerProp === "background-color") {
      decl.prop = "background-size";
      return true;
    }
    if (lowerProp === "outline-color") {
      decl.prop = "outline-width";
      return true;
    }
    if (lowerProp.startsWith("border") && lowerProp.endsWith("color")) {
      decl.prop = `${decl.prop.slice(0, -"color".length)}width`;
      return true;
    }
    if (lowerProp === "--tw-ring-color") {
      decl.prop = "--tw-ring-offset-width";
      return true;
    }
  }
  return false;
}

// src/cssVarsV4.ts
function property(ident, initialValue, _syntax) {
  return {
    prop: ident,
    value: initialValue || ""
  };
}
var nullShadow = "0 0 #0000";
var nodes = [
  // 参考：https://github.com/tailwindlabs/tailwindcss/blob/main/packages/tailwindcss/src/utilities.ts#L1137
  property("--tw-border-spacing-x", "0", "<length>"),
  property("--tw-border-spacing-y", "0", "<length>"),
  // 参考：https://github.com/tailwindlabs/tailwindcss/blob/main/packages/tailwindcss/src/utilities.ts#L1205
  property("--tw-translate-x", "0"),
  property("--tw-translate-y", "0"),
  property("--tw-translate-z", "0"),
  // 参考：https://github.com/tailwindlabs/tailwindcss/blob/main/packages/tailwindcss/src/utilities.ts#L1285
  property("--tw-scale-x", "1"),
  property("--tw-scale-y", "1"),
  property("--tw-scale-z", "1"),
  // 参考：https://github.com/tailwindlabs/tailwindcss/blob/main/packages/tailwindcss/src/utilities.ts#L1424
  property("--tw-rotate-x"),
  property("--tw-rotate-y"),
  property("--tw-rotate-z"),
  property("--tw-skew-x"),
  property("--tw-skew-y"),
  // 参考：https://github.com/tailwindlabs/tailwindcss/blob/main/packages/tailwindcss/src/utilities.ts#L1641C13-L1641C88
  property("--tw-pan-x"),
  property("--tw-pan-y"),
  property("--tw-pinch-zoom"),
  // 参考：https://github.com/tailwindlabs/tailwindcss/blob/main/packages/tailwindcss/src/utilities.ts#L1688C38-L1688C95
  property("--tw-scroll-snap-strictness", "proximity", "*"),
  // 参考：https://github.com/tailwindlabs/tailwindcss/blob/main/packages/tailwindcss/src/utilities.ts#L1688C38-L1688C95
  property("--tw-space-x-reverse", "0"),
  property("--tw-space-y-reverse", "0"),
  // 参考：https://github.com/tailwindlabs/tailwindcss/blob/main/packages/tailwindcss/src/utilities.ts#L2169C22-L2169C60
  property("--tw-border-style", "solid"),
  property("--tw-divide-x-reverse", "0"),
  property("--tw-divide-y-reverse", "0"),
  property("--tw-gradient-position", "initial"),
  property("--tw-gradient-from", "#0000", "<color>"),
  property("--tw-gradient-via", "#0000", "<color>"),
  property("--tw-gradient-to", "#0000", "<color>"),
  property("--tw-gradient-stops", "initial"),
  property("--tw-gradient-via-stops", "initial"),
  property("--tw-gradient-from-position", "0%", "<length-percentage>"),
  property("--tw-gradient-via-position", "50%", "<length-percentage>"),
  property("--tw-gradient-to-position", "100%", "<length-percentage>"),
  property("--tw-mask-linear", "linear-gradient(#fff, #fff)"),
  property("--tw-mask-radial", "linear-gradient(#fff, #fff)"),
  property("--tw-mask-conic", "linear-gradient(#fff, #fff)"),
  property("--tw-mask-left", "linear-gradient(#fff, #fff)"),
  property("--tw-mask-right", "linear-gradient(#fff, #fff)"),
  property("--tw-mask-bottom", "linear-gradient(#fff, #fff)"),
  property("--tw-mask-top", "linear-gradient(#fff, #fff)"),
  property("--tw-mask-linear-position", "0deg"),
  property("--tw-mask-linear-from-position", "0%"),
  property("--tw-mask-linear-to-position", "100%"),
  property("--tw-mask-linear-from-color", "black"),
  property("--tw-mask-linear-to-color", "transparent"),
  property("--tw-mask-radial-from-position", "0%"),
  property("--tw-mask-radial-to-position", "100%"),
  property("--tw-mask-radial-from-color", "black"),
  property("--tw-mask-radial-to-color", "transparent"),
  property("--tw-mask-radial-shape", "ellipse"),
  property("--tw-mask-radial-size", "farthest-corner"),
  property("--tw-mask-radial-position", "center"),
  property("--tw-mask-conic-position", "0deg"),
  property("--tw-mask-conic-from-position", "0%"),
  property("--tw-mask-conic-to-position", "100%"),
  property("--tw-mask-conic-from-color", "black"),
  property("--tw-mask-conic-to-color", "transparent"),
  property("--tw-font-weight"),
  property("--tw-blur"),
  property("--tw-brightness"),
  property("--tw-contrast"),
  property("--tw-grayscale"),
  property("--tw-hue-rotate"),
  property("--tw-invert"),
  property("--tw-opacity"),
  property("--tw-saturate"),
  property("--tw-sepia"),
  property("--tw-drop-shadow"),
  property("--tw-drop-shadow-color"),
  property("--tw-drop-shadow-alpha", "100%", "<percentage>"),
  property("--tw-drop-shadow-size"),
  property("--tw-backdrop-blur"),
  property("--tw-backdrop-brightness"),
  property("--tw-backdrop-contrast"),
  property("--tw-backdrop-grayscale"),
  property("--tw-backdrop-hue-rotate"),
  property("--tw-backdrop-invert"),
  property("--tw-backdrop-opacity"),
  property("--tw-backdrop-saturate"),
  property("--tw-backdrop-sepia"),
  property("--tw-duration", "initial"),
  property("--tw-ease", "initial"),
  property("--tw-content", '""'),
  property("--tw-contain-size"),
  property("--tw-contain-layout"),
  property("--tw-contain-paint"),
  property("--tw-contain-style"),
  property("--tw-leading"),
  property("--tw-tracking"),
  property("--tw-ordinal"),
  property("--tw-slashed-zero"),
  property("--tw-numeric-figure"),
  property("--tw-numeric-spacing"),
  property("--tw-numeric-fraction"),
  property("--tw-outline-style", "solid"),
  property("--tw-text-shadow-color", "initial"),
  property("--tw-text-shadow-alpha", "100%", "<percentage>"),
  property("--tw-shadow", nullShadow),
  property("--tw-shadow-color", "initial"),
  property("--tw-shadow-alpha", "100%", "<percentage>"),
  property("--tw-inset-shadow", nullShadow),
  property("--tw-inset-shadow-color", "initial"),
  property("--tw-inset-shadow-alpha", "100%", "<percentage>"),
  property("--tw-ring-color"),
  property("--tw-ring-shadow", nullShadow),
  property("--tw-inset-ring-color"),
  property("--tw-inset-ring-shadow", nullShadow),
  // Legacy 旧版变量兼容
  property("--tw-ring-inset"),
  property("--tw-ring-offset-width", "0px", "<length>"),
  property("--tw-ring-offset-color", "#fff"),
  property("--tw-ring-offset-shadow", nullShadow)
];
for (const edge of ["top", "right", "bottom", "left"]) {
  nodes.push(
    property(`--tw-mask-${edge}-from-position`, "0%"),
    property(`--tw-mask-${edge}-to-position`, "100%"),
    property(`--tw-mask-${edge}-from-color`, "black"),
    property(`--tw-mask-${edge}-to-color`, "transparent")
  );
}
var cssVarsV4_default = nodes;

// src/utils/css-vars.ts

function createCssVarNodes(definitions) {
  return definitions.map((def) => new (0, _postcss.Declaration)({
    prop: def.prop,
    value: def.value
  }));
}

// src/compat/tailwindcss-v4.ts
var OKLAB_SUFFIX = "in oklab";
var INFINITY_CALC_REGEXP = /calc\(\s*infinity\s*\*\s*(?:\d+(?:\.\d*)?|\.\d+)r?px/;
var RADIUS_THRESHOLD = 1e5;
var CLAMP_PX = 9999;
function isTailwindcssV4(options) {
  return _optionalChain([options, 'optionalAccess', _13 => _13.majorVersion]) === 4;
}
function testIfRootHostForV4(node) {
  return node.type === "rule" && node.selector.includes(":root") && node.selector.includes(":host");
}
var cssVarsV4Nodes = createCssVarNodes(cssVarsV4_default);
function isTailwindcssV4ModernCheck(atRule) {
  return atRule.name === "supports" && [
    /-webkit-hyphens\s*:\s*none/,
    /margin-trim\s*:\s*inline/,
    /-moz-orient\s*:\s*inline/,
    /color\s*:\s*rgb\(\s*from\s+red\s+r\s+g\s+b\s*\)/
  ].every((regex) => regex.test(atRule.params));
}
function normalizeTailwindcssV4Declaration(decl) {
  if (decl.prop === "--tw-gradient-position" && decl.value.endsWith(OKLAB_SUFFIX)) {
    decl.value = decl.value.slice(0, decl.value.length - OKLAB_SUFFIX.length);
    return true;
  }
  if (INFINITY_CALC_REGEXP.test(decl.value)) {
    decl.value = `${CLAMP_PX}px`;
    return true;
  }
  if (decl.prop.includes("radius")) {
    const next = decl.value.replace(
      /\b([+-]?(?:\d+(?:\.\d+)?|\.\d+)(?:e[+-]?\d+)?)\s*(r?px)\b/gi,
      (m, num) => {
        const n = Number(num);
        if (!Number.isFinite(n)) {
          return `${CLAMP_PX}px`;
        }
        if (/e/i.test(String(num)) || n > RADIUS_THRESHOLD) {
          return `${CLAMP_PX}px`;
        }
        return m;
      }
    );
    if (next !== decl.value) {
      decl.value = next;
      return true;
    }
  }
  return false;
}

// src/compat/uni-app-x.ts
function isUniAppXEnabled(options) {
  return Boolean(_optionalChain([options, 'optionalAccess', _14 => _14.uniAppX]));
}
function stripUnsupportedPseudoForUniAppX(node, enabled) {
  if (!enabled) {
    return;
  }
  node.remove();
}
function stripUnsupportedNodeForUniAppX(node, options) {
  if (!isUniAppXEnabled(options)) {
    return false;
  }
  if (node.type === "tag" || node.type === "attribute" || node.type === "pseudo") {
    node.remove();
    return true;
  }
  return false;
}
function shouldRemoveEmptyRuleForUniAppX(rule, options) {
  return isUniAppXEnabled(options) && rule.nodes.length === 0;
}

// src/constants.ts
var postcssPlugin = "postcss-weapp-tailwindcss-rename-plugin";

// src/selectorParser/before-after.ts
var _postcssselectorparser = require('postcss-selector-parser'); var _postcssselectorparser2 = _interopRequireDefault(_postcssselectorparser);
var beforeAfterStateRef = null;
var beforeAfterParser = _postcssselectorparser2.default.call(void 0, (selectors) => {
  const state = beforeAfterStateRef;
  if (!state) {
    return;
  }
  selectors.walkPseudos((s) => {
    if (_optionalChain([s, 'access', _15 => _15.parent, 'optionalAccess', _16 => _16.length]) === 1) {
      if (/^:?:before$/.test(s.value)) {
        state.before = true;
      }
      if (/^:?:after$/.test(s.value)) {
        state.after = true;
      }
    }
  });
});
function isOnlyBeforeAndAfterPseudoElement(node) {
  const state = {
    before: false,
    after: false
  };
  beforeAfterStateRef = state;
  beforeAfterParser.astSync(node);
  beforeAfterStateRef = null;
  return state.before && state.after;
}

// src/selectorParser/fallback.ts


// src/selectorParser/utils.ts

function normalizeTransformOptions(options) {
  return {
    lossless: false,
    updateSelector: true,
    ...options
  };
}
function mklist(node) {
  return [
    node,
    _postcssselectorparser2.default.combinator({
      value: "+"
    }),
    node.clone()
  ];
}
function composeIsPseudoAst(strs) {
  if (typeof strs === "string") {
    return mklist(_postcssselectorparser2.default.tag({
      value: strs
    }));
  }
  if (strs.length > 1) {
    return mklist(_postcssselectorparser2.default.pseudo({
      value: ":is",
      nodes: strs.map(
        (str) => _postcssselectorparser2.default.tag({
          value: str
        })
      )
    }));
  }
  return mklist(_postcssselectorparser2.default.tag({
    value: strs[0]
  }));
}
function getCombinatorSelectorAst(options) {
  let childCombinatorReplaceValue = mklist(_postcssselectorparser2.default.tag({ value: "view" }));
  const { cssChildCombinatorReplaceValue } = options;
  if (typeof cssChildCombinatorReplaceValue === "string" || Array.isArray(cssChildCombinatorReplaceValue) && cssChildCombinatorReplaceValue.length > 0) {
    childCombinatorReplaceValue = composeIsPseudoAst(cssChildCombinatorReplaceValue);
  }
  return childCombinatorReplaceValue;
}

// src/selectorParser/fallback.ts
var fallbackRemoveCache = /* @__PURE__ */ new WeakMap();
var fallbackDefaultKey = {};
var FALLBACK_TRANSFORM_OPTIONS = normalizeTransformOptions();
function getFallbackRemove(_rule, options) {
  const cacheKey = _nullishCoalesce(options, () => ( fallbackDefaultKey));
  let entry = fallbackRemoveCache.get(cacheKey);
  if (!entry) {
    const uniAppX = isUniAppXEnabled(options);
    let currentRule;
    const parser = _postcssselectorparser2.default.call(void 0, (selectors) => {
      const activeRule = currentRule;
      let maybeImportantId = false;
      selectors.walk((selector, idx) => {
        if (idx === 0 && (selector.type === "id" || selector.type === "class" || selector.type === "attribute")) {
          maybeImportantId = true;
        }
        if (selector.type === "universal") {
          _optionalChain([selector, 'access', _17 => _17.parent, 'optionalAccess', _18 => _18.remove, 'call', _19 => _19()]);
        } else if (selector.type === "pseudo") {
          if (selector.value === ":is") {
            if (maybeImportantId && _optionalChain([selector, 'access', _20 => _20.nodes, 'access', _21 => _21[0], 'optionalAccess', _22 => _22.type]) === "selector") {
              selector.replaceWith(selector.nodes[0]);
            } else {
              _optionalChain([selector, 'access', _23 => _23.parent, 'optionalAccess', _24 => _24.remove, 'call', _25 => _25()]);
            }
          } else if (selector.value === ":not") {
            for (const x of selector.nodes) {
              if (x.nodes.length === 1 && x.nodes[0].type === "id" && x.nodes[0].value === "#") {
                x.nodes = [
                  _postcssselectorparser2.default.tag({
                    value: "#n"
                  })
                ];
              }
            }
          } else if (selector.value === ":where") {
            for (const n of selector.nodes) {
              for (const node of n.nodes) {
                if (node.type === "attribute") {
                  node.remove();
                }
              }
            }
          }
        } else if (selector.type === "attribute") {
          if (selector.attribute === "hidden") {
            _optionalChain([activeRule, 'optionalAccess', _26 => _26.remove, 'call', _27 => _27()]);
          }
        }
      });
      selectors.walk((selector) => {
        if (selector.type === "pseudo") {
          if (selector.value === ":where") {
            const res = selector.nodes.every((x) => x.nodes.length === 0);
            if (res) {
              selector.remove();
            }
          } else if (selector.type === "pseudo") {
            stripUnsupportedPseudoForUniAppX(selector, uniAppX);
          }
        }
      });
    });
    const rawTransformSync = parser.transformSync.bind(parser);
    const transform = (targetRule) => {
      currentRule = targetRule;
      try {
        rawTransformSync(targetRule, FALLBACK_TRANSFORM_OPTIONS);
      } finally {
        currentRule = void 0;
      }
    };
    parser.transformSync = ((input, opts) => {
      const transformOptions = opts ? normalizeTransformOptions(opts) : FALLBACK_TRANSFORM_OPTIONS;
      if (input && typeof input === "object" && "type" in input) {
        const maybeRule = input;
        if (maybeRule.type === "rule") {
          currentRule = input;
          try {
            return rawTransformSync(input, transformOptions);
          } finally {
            currentRule = void 0;
          }
        }
      }
      return rawTransformSync(input, transformOptions);
    });
    entry = {
      parser,
      transform
    };
    fallbackRemoveCache.set(cacheKey, entry);
  }
  return entry.parser;
}

// src/selectorParser/rule-transformer.ts


// src/shared.ts
var _escape = require('@weapp-core/escape');
function internalCssSelectorReplacer(selectors, options = {
  escapeMap: _escape.MappingChars2String
}) {
  const { escapeMap } = options;
  const escapeOptions = {};
  if (escapeMap !== void 0) {
    escapeOptions["map"] = escapeMap;
  }
  return _escape.escape.call(void 0, selectors, escapeOptions);
}
function composeIsPseudo(strs) {
  if (typeof strs === "string") {
    return strs;
  }
  if (strs.length > 1) {
    return `:is(${strs.join(",")})`;
  }
  return strs.join("");
}

// src/utils/decl-order.ts
function reorderLiteralFirst(rule, declarations, isVariable) {
  if (declarations.length <= 1) {
    return;
  }
  const literals = [];
  const variables = [];
  for (const decl of declarations) {
    if (isVariable(decl)) {
      variables.push(decl);
    } else {
      literals.push(decl);
    }
  }
  if (literals.length === 0 || variables.length === 0) {
    return;
  }
  const desired = [...literals, ...variables];
  let alreadyOrdered = true;
  for (let index = 0; index < desired.length; index++) {
    if (desired[index] !== declarations[index]) {
      alreadyOrdered = false;
      break;
    }
  }
  if (alreadyOrdered) {
    return;
  }
  const anchor = _nullishCoalesce(_optionalChain([declarations, 'access', _28 => _28[declarations.length - 1], 'optionalAccess', _29 => _29.next, 'call', _30 => _30()]), () => ( void 0));
  for (const decl of declarations) {
    decl.remove();
  }
  for (const decl of desired) {
    if (anchor) {
      rule.insertBefore(anchor, decl);
    } else {
      rule.append(decl);
    }
  }
}

// src/selectorParser/spacing.ts
var MIRROR_PROP_PAIRS = [
  ["margin-top", "margin-bottom"],
  ["margin-left", "margin-right"],
  ["margin-inline-start", "margin-inline-end"],
  ["margin-block-start", "margin-block-end"],
  ["border-top-width", "border-bottom-width"],
  ["border-left-width", "border-right-width"],
  ["border-inline-start-width", "border-inline-end-width"],
  ["border-block-start-width", "border-block-end-width"]
];
var MIRROR_PROP_MAP = /* @__PURE__ */ new Map();
var SPACING_PROP_SET = /* @__PURE__ */ new Set();
for (const [a, b] of MIRROR_PROP_PAIRS) {
  MIRROR_PROP_MAP.set(a, b);
  MIRROR_PROP_MAP.set(b, a);
  SPACING_PROP_SET.add(a);
  SPACING_PROP_SET.add(b);
}
var LEGACY_WEBKIT_SPACING_PROPS = /* @__PURE__ */ new Set([
  "-webkit-margin-start",
  "-webkit-margin-end",
  "-webkit-margin-before",
  "-webkit-margin-after"
]);
var VAR_REFERENCE_PATTERN = /var\(/i;
function dedupeSpacingProps(rule) {
  const grouped = /* @__PURE__ */ new Map();
  for (const node of rule.nodes) {
    if (node.type !== "decl") {
      continue;
    }
    if (!SPACING_PROP_SET.has(node.prop)) {
      continue;
    }
    const list = grouped.get(node.prop);
    if (list) {
      list.push(node);
    } else {
      grouped.set(node.prop, [node]);
    }
  }
  for (const [, declarations] of grouped) {
    if (declarations.length <= 1) {
      continue;
    }
    const unique = [];
    const seenValues = /* @__PURE__ */ new Set();
    for (const decl of declarations) {
      if (decl.parent !== rule) {
        continue;
      }
      const key = `${decl.important ? "!important@@" : ""}${decl.value}`;
      if (seenValues.has(key)) {
        decl.remove();
        continue;
      }
      seenValues.add(key);
      unique.push(decl);
    }
    if (unique.length <= 1) {
      continue;
    }
    reorderLiteralFirst(
      rule,
      unique,
      (decl) => VAR_REFERENCE_PATTERN.test(decl.value)
    );
  }
}
function isNotLastChildPseudo(node) {
  if (!node || node.type !== "pseudo" || node.value !== ":not") {
    return false;
  }
  const selectors = node.nodes;
  if (!selectors || selectors.length !== 1) {
    return false;
  }
  const firstSelector = selectors[0];
  if (!firstSelector || firstSelector.type !== "selector") {
    return false;
  }
  const target = _optionalChain([firstSelector, 'access', _31 => _31.nodes, 'optionalAccess', _32 => _32[0]]);
  return Boolean(target && target.type === "pseudo" && target.value === ":last-child");
}
function transformSpacingSelector(nodes2, options) {
  if (!nodes2 || nodes2.length === 0) {
    return false;
  }
  for (let idx = 0; idx < nodes2.length; idx++) {
    const current = nodes2[idx];
    if (!current || current.type !== "class" && current.type !== "nesting") {
      continue;
    }
    const combinator = nodes2[idx + 1];
    if (!combinator || combinator.type !== "combinator" || combinator.value !== ">") {
      continue;
    }
    const candidate = nodes2[idx + 2];
    if (!isNotLastChildPseudo(candidate)) {
      continue;
    }
    const ast = getCombinatorSelectorAst(options);
    candidate.replaceWith(...ast);
    return true;
  }
  return false;
}
function normalizeSpacingDeclarations(rule) {
  for (const node of [...rule.nodes]) {
    if (node.type !== "decl") {
      continue;
    }
    if (LEGACY_WEBKIT_SPACING_PROPS.has(node.prop)) {
      node.remove();
      continue;
    }
    const mirror = MIRROR_PROP_MAP.get(node.prop);
    if (mirror) {
      node.prop = mirror;
    }
  }
  dedupeSpacingProps(rule);
}

// src/selectorParser/rule-transformer.ts
var ruleTransformCache = /* @__PURE__ */ new WeakMap();
var SELECTOR_TRANSFORM_OPTIONS = normalizeTransformOptions();
var SIMPLE_SELECTOR_FAST_PATH = /^[#.][\w-]+(?:\s+[#.][\w-]+)*$/;
var RTL_LANGUAGE_ANY_PSEUDO_SET = /* @__PURE__ */ new Set([
  ":-moz-any",
  ":-webkit-any",
  ":lang"
]);
var EMPTY_FUNCTIONAL_PSEUDO_CLEANUP_SET = /* @__PURE__ */ new Set([
  ":not",
  ":is",
  ":where",
  ":has",
  ":matches",
  ":-webkit-any",
  ":-moz-any",
  ":lang"
]);
function isRtlLanguageAnyPseudo(node) {
  return node.type === "pseudo" && RTL_LANGUAGE_ANY_PSEUDO_SET.has(node.value);
}
function getTopLevelSelector(node) {
  let current = node;
  let topLevel = node;
  while (current.parent) {
    const parent = current.parent;
    if (parent.type === "root") {
      break;
    }
    if (parent.type === "selector") {
      topLevel = parent;
    }
    current = parent;
  }
  return topLevel;
}
function stripUnsupportedRtlLanguagePseudo(node) {
  const selectorParent = node.parent;
  if (!selectorParent || selectorParent.type !== "selector") {
    return;
  }
  const maybeNot = selectorParent.parent;
  if (maybeNot && maybeNot.type === "pseudo" && maybeNot.value === ":not") {
    maybeNot.remove();
    return;
  }
  getTopLevelSelector(selectorParent).remove();
}
function shouldRemoveEmptyFunctionalPseudo(node) {
  return node.type === "pseudo" && EMPTY_FUNCTIONAL_PSEUDO_CLEANUP_SET.has(node.value) && Array.isArray(node.nodes) && node.nodes.length === 0;
}
function flattenWherePseudo(node, context, index, parent) {
  if (isUniAppXEnabled(context.options)) {
    node.value = ":is";
  }
  if (index === 0 && node.length === 1) {
    const targetSelector = _optionalChain([node, 'access', _33 => _33.nodes, 'optionalAccess', _34 => _34[0]]);
    if (targetSelector && targetSelector.type === "selector" && transformSpacingSelector(targetSelector.nodes, context.options)) {
      context.requiresSpacingNormalization = true;
    }
    node.replaceWith(...node.nodes);
    if (parent && parent.type === "selector" && parent.length === 0) {
      parent.remove();
    }
  }
}
function handleClassNode(node, context) {
  if (node.type !== "class") {
    return;
  }
  const { escapeMap } = context.options;
  node.value = escapeMap === void 0 ? internalCssSelectorReplacer(node.value, {}) : internalCssSelectorReplacer(node.value, { escapeMap });
}
function handleUniversalNode(node, context) {
  if (node.type !== "universal") {
    return;
  }
  const replacement = _optionalChain([context, 'access', _35 => _35.options, 'access', _36 => _36.cssSelectorReplacement, 'optionalAccess', _37 => _37.universal]);
  if (replacement) {
    node.value = composeIsPseudo(replacement);
  }
}
function shouldRemoveHoverSelector(selector, options) {
  if (!options.cssRemoveHoverPseudoClass) {
    return false;
  }
  return selector.nodes.some((node) => node.type === "pseudo" && node.value === ":hover");
}
var UNSUPPORTED_PSEUDO_ELEMENT_SELECTOR_SET = /* @__PURE__ */ new Set([
  "::backdrop",
  "::file-selector-button"
]);
function shouldRemoveUnsupportedPseudoElementSelector(selector) {
  return selector.nodes.some((node) => node.type === "pseudo" && UNSUPPORTED_PSEUDO_ELEMENT_SELECTOR_SET.has(node.value));
}
function isHiddenOrTemplateNotPseudo(node) {
  if (!node || node.type !== "pseudo" || node.value !== ":not") {
    return false;
  }
  const selector = node.first;
  if (!selector || selector.type !== "selector") {
    return false;
  }
  const first = selector.first;
  if (!first) {
    return false;
  }
  if (first.type === "attribute") {
    return first.attribute === "hidden";
  }
  if (first.type === "tag") {
    return first.value === "template";
  }
  return false;
}
function handleCombinatorNode(node, index, context) {
  if (node.type !== "combinator" || node.value !== ">") {
    return;
  }
  const nodes2 = _optionalChain([node, 'access', _38 => _38.parent, 'optionalAccess', _39 => _39.nodes]);
  if (!nodes2) {
    return;
  }
  const first = nodes2[index + 1];
  const second = nodes2[index + 2];
  const third = nodes2[index + 3];
  if (isHiddenOrTemplateNotPseudo(first) && second && second.type === "combinator" && (second.value === "~" || second.value === "+") && isHiddenOrTemplateNotPseudo(third)) {
    const ast = getCombinatorSelectorAst(context.options);
    nodes2.splice(index + 1, 3, ...ast);
  }
}
function handlePseudoNode(node, index, context, parent) {
  if (node.type !== "pseudo") {
    return;
  }
  if (isRtlLanguageAnyPseudo(node)) {
    stripUnsupportedRtlLanguagePseudo(node);
    return;
  }
  if (node.value === ":root" && _optionalChain([context, 'access', _40 => _40.options, 'access', _41 => _41.cssSelectorReplacement, 'optionalAccess', _42 => _42.root])) {
    node.value = composeIsPseudo(context.options.cssSelectorReplacement.root);
    return;
  }
  if (node.value === ":where") {
    flattenWherePseudo(node, context, index, parent);
  }
}
function handleTagOrAttribute(node, context) {
  stripUnsupportedNodeForUniAppX(node, context.options);
}
function handleSelectorNode(selector, context) {
  if (shouldRemoveUnsupportedPseudoElementSelector(selector)) {
    selector.remove();
    return;
  }
  if (shouldRemoveHoverSelector(selector, context.options)) {
    selector.remove();
    return;
  }
  if (transformSpacingSelector(selector.nodes, context.options)) {
    context.requiresSpacingNormalization = true;
  }
}
function canSkipRuleTransform(rule) {
  const selector = rule.selector.trim();
  if (!selector) {
    return false;
  }
  return SIMPLE_SELECTOR_FAST_PATH.test(selector);
}
function transformSelectors(selectors, context) {
  selectors.walk((node, index) => {
    const parent = _optionalChain([node, 'access', _43 => _43.parent, 'optionalAccess', _44 => _44.type]) === "selector" ? node.parent : void 0;
    switch (node.type) {
      case "class":
        handleClassNode(node, context);
        break;
      case "universal":
        handleUniversalNode(node, context);
        break;
      case "selector":
        handleSelectorNode(node, context);
        break;
      case "pseudo":
        if (!isNotLastChildPseudo(node)) {
          handlePseudoNode(node, index, context, parent);
        }
        break;
      case "combinator":
        handleCombinatorNode(node, index, context);
        break;
      case "tag":
      case "attribute":
        handleTagOrAttribute(node, context);
        break;
      default:
        break;
    }
  });
  if (context.requiresSpacingNormalization) {
    normalizeSpacingDeclarations(context.rule);
  }
  selectors.walk((node) => {
    if (shouldRemoveEmptyFunctionalPseudo(node)) {
      node.remove();
      return;
    }
    if (node.type === "selector" && node.length === 0) {
      node.remove();
    }
  });
  if (selectors.length === 0) {
    context.rule.remove();
  }
}
function createRuleTransformer(options) {
  let context;
  const selectorResultCache = /* @__PURE__ */ new Map();
  const selectorResultCacheLimit = 5e4;
  function writeSelectorResultCache(selector, result) {
    if (selectorResultCache.size >= selectorResultCacheLimit) {
      selectorResultCache.clear();
    }
    selectorResultCache.set(selector, result);
  }
  const transform = (selectors) => {
    if (!context) {
      return;
    }
    transformSelectors(selectors, context);
  };
  const parser = _postcssselectorparser2.default.call(void 0, transform);
  return (rule) => {
    const sourceSelector = rule.selector;
    if (!sourceSelector) {
      return;
    }
    const cached = selectorResultCache.get(sourceSelector);
    if (cached) {
      if (cached.action === "remove") {
        rule.remove();
      } else if (cached.action === "update" && cached.selector && cached.selector !== sourceSelector) {
        rule.selector = cached.selector;
      }
      return;
    }
    if (canSkipRuleTransform(rule)) {
      writeSelectorResultCache(sourceSelector, { action: "keep" });
      return;
    }
    context = {
      options,
      requiresSpacingNormalization: false,
      rule
    };
    let wasRemoved = false;
    let requiresSpacingNormalization = false;
    try {
      parser.transformSync(rule, SELECTOR_TRANSFORM_OPTIONS);
    } finally {
      const currentContext = context;
      wasRemoved = rule.parent == null;
      requiresSpacingNormalization = _optionalChain([currentContext, 'optionalAccess', _45 => _45.requiresSpacingNormalization]) === true;
      context = void 0;
    }
    if (wasRemoved) {
      writeSelectorResultCache(sourceSelector, { action: "remove" });
      return;
    }
    if (requiresSpacingNormalization) {
      return;
    }
    const transformedSelector = rule.selector;
    if (transformedSelector === sourceSelector) {
      writeSelectorResultCache(sourceSelector, { action: "keep" });
    } else {
      writeSelectorResultCache(sourceSelector, {
        action: "update",
        selector: transformedSelector
      });
    }
  };
}
function ruleTransformSync(rule, options) {
  let transformer = ruleTransformCache.get(options);
  if (!transformer) {
    transformer = createRuleTransformer(options);
    ruleTransformCache.set(options, transformer);
  }
  transformer(rule);
}

// src/plugins/post/decl-dedupe.ts
var logicalPropMap = /* @__PURE__ */ new Map([
  // margin 方向映射
  ["margin-inline-start", "margin-left"],
  ["margin-inline-end", "margin-right"],
  ["margin-block-start", "margin-top"],
  ["margin-block-end", "margin-bottom"],
  // padding 方向映射
  ["padding-inline-start", "padding-left"],
  ["padding-inline-end", "padding-right"],
  ["padding-block-start", "padding-top"],
  ["padding-block-end", "padding-bottom"],
  // border 方向映射
  ["border-inline-start", "border-left"],
  ["border-inline-end", "border-right"],
  ["border-block-start", "border-top"],
  ["border-block-end", "border-bottom"],
  ["border-inline-start-width", "border-left-width"],
  ["border-inline-end-width", "border-right-width"]
]);
var variablePriorityProps = /* @__PURE__ */ new Set([
  "margin-left",
  "margin-right",
  "margin-top",
  "margin-bottom",
  "border-left-width",
  "border-right-width",
  "border-top-width",
  "border-bottom-width"
]);
function getCanonicalProp(prop) {
  return _nullishCoalesce(logicalPropMap.get(prop), () => ( prop));
}
function normalizeCalcValue(value) {
  if (!value.includes("calc")) {
    return value;
  }
  let next = value;
  let prev;
  do {
    prev = next;
    next = prev.replace(/calc\(\s*calc\(/gi, "calc((");
  } while (next !== prev);
  return next.replace(/calc\(\s*(1\s*-\s*var\([^()]+\))\s*\)/gi, "($1)");
}
function hasVariableReference(value) {
  return value.includes("var(");
}
function dedupeDeclarations(rule) {
  const entries = [];
  for (const node of [...rule.nodes]) {
    if (node.type !== "decl") {
      continue;
    }
    const decl = node;
    const normalizedValue = normalizeCalcValue(decl.value);
    if (normalizedValue !== decl.value) {
      decl.value = normalizedValue;
    }
    const canonicalProp = getCanonicalProp(decl.prop);
    entries.push({
      decl,
      normalizedValue,
      canonicalProp,
      importantKey: decl.important ? "!important" : "",
      isLogical: canonicalProp !== decl.prop
    });
  }
  const seen = /* @__PURE__ */ new Map();
  for (const entry of entries) {
    const key = `${entry.canonicalProp}${entry.importantKey}@@${entry.normalizedValue}`;
    const existing = seen.get(key);
    if (!existing) {
      seen.set(key, entry);
      continue;
    }
    if (existing.isLogical && !entry.isLogical) {
      existing.decl.remove();
      seen.set(key, entry);
    } else {
      entry.decl.remove();
    }
  }
  const reorderGroups = /* @__PURE__ */ new Map();
  for (const node of rule.nodes) {
    if (node.type !== "decl") {
      continue;
    }
    const canonical = getCanonicalProp(node.prop);
    if (!variablePriorityProps.has(canonical)) {
      continue;
    }
    const existing = reorderGroups.get(canonical);
    if (existing) {
      existing.push(node);
    } else {
      reorderGroups.set(canonical, [node]);
    }
  }
  for (const declarations of reorderGroups.values()) {
    if (declarations.length <= 1) {
      continue;
    }
    reorderLiteralFirst(
      rule,
      declarations,
      (decl) => hasVariableReference(decl.value)
    );
  }
  const literalSeen = /* @__PURE__ */ new Map();
  for (const node of [...rule.nodes]) {
    if (node.type !== "decl") {
      continue;
    }
    const canonical = getCanonicalProp(node.prop);
    if (!variablePriorityProps.has(canonical)) {
      continue;
    }
    if (hasVariableReference(node.value)) {
      continue;
    }
    const existing = literalSeen.get(canonical);
    if (existing) {
      node.remove();
    } else {
      literalSeen.set(canonical, node);
    }
  }
}

// src/plugins/post/specificity-cleaner.ts
function normalizeSelectorList(value) {
  if (value === void 0 || value === false) {
    return [];
  }
  return Array.isArray(value) ? value.filter(Boolean) : [value];
}
function getSpecificityMatchingName(options) {
  const feature = _optionalChain([options, 'access', _46 => _46.cssPresetEnv, 'optionalAccess', _47 => _47.features, 'optionalAccess', _48 => _48["is-pseudo-class"]]);
  if (feature && typeof feature === "object" && "specificityMatchingName" in feature) {
    const specificityName = feature.specificityMatchingName;
    return typeof specificityName === "string" && specificityName.length > 0 ? specificityName : void 0;
  }
  return void 0;
}
function createRootSpecificityCleaner(options) {
  const specificityMatchingName = getSpecificityMatchingName(options);
  const selectors = normalizeSelectorList(_optionalChain([options, 'access', _49 => _49.cssSelectorReplacement, 'optionalAccess', _50 => _50.root]));
  if (!specificityMatchingName || selectors.length === 0) {
    return void 0;
  }
  const suffix = `:not(.${specificityMatchingName})`;
  const targets = selectors.map((selector) => _optionalChain([selector, 'optionalAccess', _51 => _51.trim, 'call', _52 => _52()])).filter((selector) => Boolean(_optionalChain([selector, 'optionalAccess', _53 => _53.length]))).map((selector) => ({
    match: `${selector}${suffix}`,
    spacedMatch: `${selector} ${suffix}`,
    replacement: selector
  }));
  if (!targets.length) {
    return void 0;
  }
  return (rule) => {
    if (!rule.selectors || rule.selectors.length === 0) {
      return;
    }
    const next = rule.selectors.map((selector) => {
      let updated = selector;
      for (const target of targets) {
        if (updated.includes(target.match)) {
          updated = updated.split(target.match).join(target.replacement);
        }
        if (updated.includes(target.spacedMatch)) {
          updated = updated.split(target.spacedMatch).join(target.replacement);
        }
      }
      return updated;
    });
    rule.selectors = next;
  };
}

// src/plugins/post.ts
var DEFAULT_ROOT_SELECTORS = ["page", ".tw-root", "wx-root-portal-content"];
function normalizeRootSelectors(value) {
  if (value === void 0 || value === false) {
    return [];
  }
  return Array.isArray(value) ? value.filter(Boolean) : [value];
}
function shouldAppendHostSelector(rule, options) {
  const selectors = _nullishCoalesce(rule.selectors, () => ( []));
  if (selectors.includes(":host")) {
    return false;
  }
  const rootSelectors = normalizeRootSelectors(_optionalChain([options, 'access', _54 => _54.cssSelectorReplacement, 'optionalAccess', _55 => _55.root]));
  if (rootSelectors.length !== DEFAULT_ROOT_SELECTORS.length || !rootSelectors.every((selector, index) => selector === DEFAULT_ROOT_SELECTORS[index])) {
    return false;
  }
  return DEFAULT_ROOT_SELECTORS.every((selector) => selectors.includes(selector));
}
var postcssWeappTailwindcssPostPlugin = (options) => {
  const opts = _shared.defu.call(void 0, options, {
    isMainChunk: true
  });
  const p = {
    postcssPlugin
  };
  const cleanRootSpecificity = createRootSpecificityCleaner(opts);
  const enableMainChunkTransforms = opts.isMainChunk !== false;
  if (enableMainChunkTransforms || cleanRootSpecificity) {
    const fallbackRemove = enableMainChunkTransforms ? getFallbackRemove(void 0, opts) : void 0;
    p.RuleExit = (rule) => {
      if (enableMainChunkTransforms) {
        _optionalChain([fallbackRemove, 'optionalAccess', _56 => _56.transformSync, 'call', _57 => _57(rule)]);
      }
      _optionalChain([cleanRootSpecificity, 'optionalCall', _58 => _58(rule)]);
      if (enableMainChunkTransforms) {
        if (shouldAppendHostSelector(rule, opts)) {
          rule.selectors = [...rule.selectors, ":host"];
        }
        dedupeDeclarations(rule);
        if (rule.selectors.length === 0 || rule.selectors.length === 1 && rule.selector.trim() === "") {
          rule.remove();
        }
        if (shouldRemoveEmptyRuleForUniAppX(rule, opts)) {
          rule.remove();
        }
      }
    };
  }
  if (enableMainChunkTransforms) {
    p.DeclarationExit = (decl) => {
      if (opts.majorVersion === void 0) {
        normalizeTailwindcssRpxDeclaration(decl);
      } else {
        normalizeTailwindcssRpxDeclaration(decl, { majorVersion: opts.majorVersion });
      }
      normalizeTailwindcssV4Declaration(decl);
    };
    p.AtRuleExit = (atRule) => {
      if (opts.cssRemoveProperty && atRule.name === "property") {
        atRule.remove();
      }
      _optionalChain([atRule, 'access', _59 => _59.nodes, 'optionalAccess', _60 => _60.length]) === 0 && atRule.remove();
    };
  }
  return p;
};
postcssWeappTailwindcssPostPlugin.postcss = true;

// src/plugins/pre.ts


// src/mp.ts


// src/cssVarsV3.ts
var cssVarsV3_default = [
  {
    prop: "--tw-border-spacing-x",
    value: "0"
  },
  {
    prop: "--tw-border-spacing-y",
    value: "0"
  },
  {
    prop: "--tw-translate-x",
    value: "0"
  },
  {
    prop: "--tw-translate-y",
    value: "0"
  },
  {
    prop: "--tw-rotate",
    value: "0"
  },
  {
    prop: "--tw-skew-x",
    value: "0"
  },
  {
    prop: "--tw-skew-y",
    value: "0"
  },
  {
    prop: "--tw-scale-x",
    value: "1"
  },
  {
    prop: "--tw-scale-y",
    value: "1"
  },
  {
    prop: "--tw-pan-x",
    value: " "
  },
  {
    prop: "--tw-pan-y",
    value: " "
  },
  {
    prop: "--tw-pinch-zoom",
    value: " "
  },
  {
    prop: "--tw-scroll-snap-strictness",
    value: "proximity"
  },
  {
    prop: "--tw-gradient-from-position",
    value: " "
  },
  {
    prop: "--tw-gradient-via-position",
    value: " "
  },
  {
    prop: "--tw-gradient-to-position",
    value: " "
  },
  {
    prop: "--tw-ordinal",
    value: " "
  },
  {
    prop: "--tw-slashed-zero",
    value: " "
  },
  {
    prop: "--tw-numeric-figure",
    value: " "
  },
  {
    prop: "--tw-numeric-spacing",
    value: " "
  },
  {
    prop: "--tw-numeric-fraction",
    value: " "
  },
  {
    prop: "--tw-ring-inset",
    value: " "
  },
  {
    prop: "--tw-ring-offset-width",
    value: "0px"
  },
  {
    prop: "--tw-ring-offset-color",
    value: "#fff"
  },
  {
    prop: "--tw-ring-color",
    value: "rgb(59 130 246 / 0.5)"
  },
  {
    prop: "--tw-ring-offset-shadow",
    value: "0 0 #0000"
  },
  {
    prop: "--tw-ring-shadow",
    value: "0 0 #0000"
  },
  {
    prop: "--tw-shadow",
    value: "0 0 #0000"
  },
  {
    prop: "--tw-shadow-colored",
    value: "0 0 #0000"
  },
  {
    prop: "--tw-blur",
    value: " "
  },
  {
    prop: "--tw-brightness",
    value: " "
  },
  {
    prop: "--tw-contrast",
    value: " "
  },
  {
    prop: "--tw-grayscale",
    value: " "
  },
  {
    prop: "--tw-hue-rotate",
    value: " "
  },
  {
    prop: "--tw-invert",
    value: " "
  },
  {
    prop: "--tw-saturate",
    value: " "
  },
  {
    prop: "--tw-sepia",
    value: " "
  },
  {
    prop: "--tw-drop-shadow",
    value: " "
  },
  {
    prop: "--tw-backdrop-blur",
    value: " "
  },
  {
    prop: "--tw-backdrop-brightness",
    value: " "
  },
  {
    prop: "--tw-backdrop-contrast",
    value: " "
  },
  {
    prop: "--tw-backdrop-grayscale",
    value: " "
  },
  {
    prop: "--tw-backdrop-hue-rotate",
    value: " "
  },
  {
    prop: "--tw-backdrop-invert",
    value: " "
  },
  {
    prop: "--tw-backdrop-opacity",
    value: " "
  },
  {
    prop: "--tw-backdrop-saturate",
    value: " "
  },
  {
    prop: "--tw-backdrop-sepia",
    value: " "
  },
  {
    prop: "--tw-contain-size",
    value: " "
  },
  {
    prop: "--tw-contain-layout",
    value: " "
  },
  {
    prop: "--tw-contain-paint",
    value: " "
  },
  {
    prop: "--tw-contain-style",
    value: " "
  }
];

// src/utils/tw-vars.ts
function hasTwVars(rule, count = 2) {
  let matched = 0;
  for (const node of _nullishCoalesce(rule.nodes, () => ( []))) {
    if (node.type === "decl" && node.prop.startsWith("--tw-")) {
      matched++;
      if (matched >= count) {
        return true;
      }
    }
  }
  return false;
}

// src/mp.ts
var cssVarsV3Nodes = createCssVarNodes(cssVarsV3_default);
var DEFAULT_ROOT_SELECTORS2 = ["page", ".tw-root", "wx-root-portal-content"];
function testIfVariablesScope(node, count = 2) {
  if (isOnlyBeforeAndAfterPseudoElement(node)) {
    return hasTwVars(node, count);
  }
  return false;
}
function testIfTwBackdrop(node, count = 2) {
  if (node.type === "rule" && node.selector === "::backdrop") {
    return hasTwVars(node, count);
  }
  return false;
}
function makePseudoVarRule() {
  const pseudoVarRule = new (0, _postcss.Rule)({
    // selectors: ['::before', '::after'],
    selector: "::before,::after"
  });
  pseudoVarRule.append(
    new (0, _postcss.Declaration)({
      prop: "--tw-content",
      value: '""'
    })
  );
  return pseudoVarRule;
}
function remakeCssVarSelector(selectors, options) {
  const { cssPreflightRange, cssSelectorReplacement } = options;
  if (cssPreflightRange === "all" && !selectors.includes(":not(not)")) {
    selectors.push(":not(not)");
  }
  if (cssSelectorReplacement) {
    if (Array.isArray(cssSelectorReplacement.universal)) {
      if (!cssSelectorReplacement.universal.every((x) => {
        return selectors.includes(x);
      }) && !selectors.includes("*")) {
        selectors.unshift("*");
      }
    } else if (typeof cssSelectorReplacement.universal === "string" && !selectors.includes(cssSelectorReplacement.universal) && !selectors.includes("*")) {
      selectors.unshift("*");
    }
  }
  return selectors;
}
function commonChunkPreflight(node, options) {
  const { ctx, cssInjectPreflight, injectAdditionalCssVarScope } = options;
  const rootOption = _optionalChain([options, 'access', _61 => _61.cssSelectorReplacement, 'optionalAccess', _62 => _62.root]);
  const rootSelectors = rootOption === false || rootOption === void 0 ? [] : Array.isArray(rootOption) ? rootOption.filter(Boolean) : [rootOption];
  const hasHostSelector = node.selectors.some((selector) => selector.includes(":host"));
  const hasRootPseudoSelector = node.selectors.some((selector) => selector.includes(":root"));
  const hasAllDefaultRootSelectors = DEFAULT_ROOT_SELECTORS2.every((selector) => node.selectors.includes(selector));
  if (!hasHostSelector && !rootSelectors.includes(":host") && (hasRootPseudoSelector || rootSelectors.length === DEFAULT_ROOT_SELECTORS2.length && rootSelectors.every((selector, index) => selector === DEFAULT_ROOT_SELECTORS2[index]) && hasAllDefaultRootSelectors)) {
    node.selectors = [...node.selectors, ":host"];
  }
  if (testIfVariablesScope(node)) {
    _optionalChain([ctx, 'optionalAccess', _63 => _63.markVariablesScope, 'call', _64 => _64(node)]);
    node.selectors = remakeCssVarSelector(node.selectors, options);
    node.before(makePseudoVarRule());
    if (typeof cssInjectPreflight === "function") {
      node.append(...cssInjectPreflight());
    }
  }
  const isTailwindcss4 = isTailwindcssV4(options);
  if (injectAdditionalCssVarScope && (isTailwindcss4 ? testIfRootHostForV4(node) : testIfTwBackdrop(node))) {
    const syntheticRule = new (0, _postcss.Rule)({
      selectors: ["*", "::after", "::before"],
      nodes: isTailwindcss4 ? cssVarsV4Nodes : cssVarsV3Nodes
    });
    syntheticRule.selectors = remakeCssVarSelector(syntheticRule.selectors, options);
    node.before(syntheticRule);
    node.before(makePseudoVarRule());
    if (typeof cssInjectPreflight === "function") {
      syntheticRule.append(...cssInjectPreflight());
    }
  }
}

// src/plugins/pre.ts
function isAtMediaHover(atRule) {
  return /media\(\s*hover\s*:\s*hover\s*\)/.test(atRule.name) || atRule.name === "media" && /\(\s*hover\s*:\s*hover\s*\)/.test(atRule.params);
}
var postcssWeappTailwindcssPrePlugin = (options) => {
  const opts = _shared.defu.call(void 0, options, { isMainChunk: true });
  const p = {
    postcssPlugin,
    Rule(rule) {
      ruleTransformSync(rule, opts);
    },
    AtRule(atRule) {
      if (isAtMediaHover(atRule)) {
        if (atRule.nodes) {
          atRule.replaceWith(atRule.nodes);
        } else {
          atRule.remove();
        }
      } else if (atRule.name === "supports") {
        if (/color-mix/.test(atRule.params)) {
          atRule.remove();
        }
      } else if (atRule.name === "layer") {
        if (atRule.nodes === void 0 || Array.isArray(atRule.nodes) && atRule.nodes.length === 0) {
          atRule.remove();
        }
      }
    }
  };
  if (opts.isMainChunk) {
    let layerProperties;
    p.Once = (root) => {
      root.walkAtRules((atRule) => {
        if (atRule.name === "layer") {
          if (atRule.params === "properties") {
            if (atRule.nodes === void 0 || _optionalChain([atRule, 'access', _65 => _65.nodes, 'optionalAccess', _66 => _66.length]) === 0) {
              layerProperties = atRule;
            } else if (_optionalChain([atRule, 'access', _67 => _67.first, 'optionalAccess', _68 => _68.type]) === "atrule" && isTailwindcssV4ModernCheck(atRule.first)) {
              if (layerProperties) {
                layerProperties.replaceWith(atRule.first.nodes);
                atRule.remove();
              } else {
                atRule.replaceWith(atRule.first.nodes);
              }
            }
          } else {
            atRule.replaceWith(atRule.nodes);
          }
        } else if (isTailwindcssV4ModernCheck(atRule)) {
          if (_optionalChain([atRule, 'access', _69 => _69.first, 'optionalAccess', _70 => _70.type]) === "atrule" && atRule.first.name === "layer") {
            atRule.replaceWith(atRule.first.nodes);
          }
        }
      });
      root.walkRules((rule) => {
        commonChunkPreflight(rule, opts);
      });
    };
  }
  return p;
};
postcssWeappTailwindcssPrePlugin.postcss = true;

// src/pipeline.ts
var STAGE_ORDER = ["pre", "normal", "post"];
function normalizeUserPlugins(plugins) {
  if (!plugins) {
    return [];
  }
  if (Array.isArray(plugins)) {
    return plugins.filter(Boolean);
  }
  if (typeof plugins === "object") {
    return Object.values(plugins).filter(Boolean);
  }
  return [];
}
function createStaticDefinition(id, stage, plugin) {
  return {
    id,
    stage,
    prepare: () => ({
      id,
      stage,
      createPlugin: () => plugin
    })
  };
}
function createPipelineDefinitions(options) {
  const stages = {
    pre: [],
    normal: [],
    post: []
  };
  const userPlugins = normalizeUserPlugins(_optionalChain([options, 'access', _71 => _71.postcssOptions, 'optionalAccess', _72 => _72.plugins]));
  userPlugins.forEach((plugin, index) => {
    stages.pre.push(createStaticDefinition(`pre:user-${index}`, "pre", plugin));
  });
  stages.pre.push({
    id: "pre:core",
    stage: "pre",
    prepare: () => ({
      id: "pre:core",
      stage: "pre",
      createPlugin: () => postcssWeappTailwindcssPrePlugin(options)
    })
  });
  stages.normal.push({
    id: "normal:preset-env",
    stage: "normal",
    prepare: () => ({
      id: "normal:preset-env",
      stage: "normal",
      createPlugin: () => _postcsspresetenv2.default.call(void 0, options.cssPresetEnv)
    })
  });
  stages.normal.push({
    id: "normal:color-functional-fallback",
    stage: "normal",
    prepare: () => ({
      id: "normal:color-functional-fallback",
      stage: "normal",
      createPlugin: () => createColorFunctionalFallback()
    })
  });
  stages.normal.push({
    id: "normal:units-to-px",
    stage: "normal",
    prepare: () => {
      const plugin = getUnitsToPxPlugin(options);
      return plugin ? {
        id: "normal:units-to-px",
        stage: "normal",
        createPlugin: () => plugin
      } : void 0;
    }
  });
  stages.normal.push({
    id: "normal:px-transform",
    stage: "normal",
    prepare: () => {
      const plugin = getPxTransformPlugin(options);
      return plugin ? {
        id: "normal:px-transform",
        stage: "normal",
        createPlugin: () => plugin
      } : void 0;
    }
  });
  stages.normal.push({
    id: "normal:rem-transform",
    stage: "normal",
    prepare: () => {
      const plugin = getRemTransformPlugin(options);
      return plugin ? {
        id: "normal:rem-transform",
        stage: "normal",
        createPlugin: () => plugin
      } : void 0;
    }
  });
  stages.normal.push({
    id: "normal:calc",
    stage: "normal",
    prepare: () => {
      const plugin = getCalcPlugin(options);
      return plugin ? {
        id: "normal:calc",
        stage: "normal",
        createPlugin: () => plugin
      } : void 0;
    }
  });
  stages.normal.push({
    id: "normal:calc-duplicate-cleaner",
    stage: "normal",
    prepare: () => {
      const plugin = getCalcDuplicateCleaner(options);
      return plugin ? {
        id: "normal:calc-duplicate-cleaner",
        stage: "normal",
        createPlugin: () => plugin
      } : void 0;
    }
  });
  stages.normal.push({
    id: "normal:custom-property-cleaner",
    stage: "normal",
    prepare: () => {
      const plugin = getCustomPropertyCleaner(options);
      return plugin ? {
        id: "normal:custom-property-cleaner",
        stage: "normal",
        createPlugin: () => plugin
      } : void 0;
    }
  });
  stages.post.push({
    id: "post:core",
    stage: "post",
    prepare: () => ({
      id: "post:core",
      stage: "post",
      createPlugin: () => postcssWeappTailwindcssPostPlugin(options)
    })
  });
  return STAGE_ORDER.flatMap((stage) => stages[stage]);
}
function createStylePipeline(options) {
  options.ctx = createContext();
  const preparedNodes = createPipelineDefinitions(options).map((definition) => definition.prepare(options)).filter(Boolean);
  if (preparedNodes.length === 0) {
    return {
      nodes: [],
      plugins: []
    };
  }
  const stageSizes = /* @__PURE__ */ new Map();
  preparedNodes.forEach((node) => {
    stageSizes.set(node.stage, (_nullishCoalesce(stageSizes.get(node.stage), () => ( 0))) + 1);
  });
  const stageIndices = /* @__PURE__ */ new Map();
  const nodes2 = [];
  const size = preparedNodes.length;
  preparedNodes.forEach((node, index) => {
    const stageIndex = _nullishCoalesce(stageIndices.get(node.stage), () => ( 0));
    const context = {
      stage: node.stage,
      index,
      size,
      stageIndex,
      stageSize: _nullishCoalesce(stageSizes.get(node.stage), () => ( 0))
    };
    if (index > 0) {
      const prevNode = preparedNodes[index - 1];
      if (prevNode) {
        context.previous = {
          id: prevNode.id,
          stage: prevNode.stage
        };
      }
    }
    if (index < size - 1) {
      const nextNode = preparedNodes[index + 1];
      if (nextNode) {
        context.next = {
          id: nextNode.id,
          stage: nextNode.stage
        };
      }
    }
    stageIndices.set(node.stage, stageIndex + 1);
    nodes2.push({
      id: node.id,
      stage: node.stage,
      plugin: node.createPlugin(context),
      context
    });
  });
  return {
    nodes: nodes2,
    plugins: nodes2.map((node) => node.plugin)
  };
}

// src/processor-cache.ts
function createProcessOptions(options) {
  return {
    from: void 0,
    ..._nullishCoalesce(_optionalChain([options, 'access', _73 => _73.postcssOptions, 'optionalAccess', _74 => _74.options]), () => ( {}))
  };
}
var StyleProcessorCache = (_class = class {constructor() { _class.prototype.__init.call(this);_class.prototype.__init2.call(this);_class.prototype.__init3.call(this);_class.prototype.__init4.call(this);_class.prototype.__init5.call(this); }
  __init() {this.pipelineCache = /* @__PURE__ */ new WeakMap()}
  __init2() {this.processOptionsCache = /* @__PURE__ */ new WeakMap()}
  __init3() {this.processorCache = /* @__PURE__ */ new WeakMap()}
  __init4() {this.processorCacheByKey = /* @__PURE__ */ new Map()}
  __init5() {this.processorKeyCache = /* @__PURE__ */ new WeakMap()}
  createProcessorCacheKey(options) {
    const from = _optionalChain([options, 'access', _75 => _75.postcssOptions, 'optionalAccess', _76 => _76.options, 'optionalAccess', _77 => _77.from]);
    if (from == null) {
      return fingerprintOptions(options);
    }
    return fingerprintOptions({
      ...options,
      postcssOptions: {
        ..._nullishCoalesce(options.postcssOptions, () => ( {})),
        options: {
          ..._nullishCoalesce(_optionalChain([options, 'access', _78 => _78.postcssOptions, 'optionalAccess', _79 => _79.options]), () => ( {})),
          from: void 0
        }
      }
    });
  }
  getPipeline(options) {
    let pipeline = this.pipelineCache.get(options);
    if (!pipeline) {
      pipeline = createStylePipeline(options);
      this.pipelineCache.set(options, pipeline);
    }
    return pipeline;
  }
  getProcessOptions(options) {
    const source = _optionalChain([options, 'access', _80 => _80.postcssOptions, 'optionalAccess', _81 => _81.options]);
    const fingerprint = source ? fingerprintOptions(source) : void 0;
    const cached = this.processOptionsCache.get(options);
    if (!cached || cached.fingerprint !== fingerprint) {
      const created = createProcessOptions(options);
      this.processOptionsCache.set(options, { value: created, fingerprint });
      return { ...created };
    }
    return { ...cached.value };
  }
  getProcessor(options) {
    let processor = this.processorCache.get(options);
    if (!processor) {
      let cacheKey = this.processorKeyCache.get(options);
      if (!cacheKey) {
        cacheKey = this.createProcessorCacheKey(options);
        this.processorKeyCache.set(options, cacheKey);
      }
      processor = this.processorCacheByKey.get(cacheKey);
      if (!processor) {
        const pipeline = this.getPipeline(options);
        processor = _postcss2.default.call(void 0, pipeline.plugins);
        this.processorCacheByKey.set(cacheKey, processor);
      }
      this.processorCache.set(options, processor);
    }
    return processor;
  }
}, _class);

// src/handler.ts
function createStyleHandler(options) {
  const cachedOptions = _shared.defuOverrideArray.call(void 0, 
    options,
    getDefaultOptions(options)
  );
  cachedOptions.cssInjectPreflight = createInjectPreflight(cachedOptions.cssPreflight);
  const resolver = createOptionsResolver(cachedOptions);
  const cache = new StyleProcessorCache();
  const base = resolver.resolve();
  cache.getProcessor(base);
  cache.getProcessOptions(base);
  const handler = ((rawSource, opt) => {
    const resolvedOptions = resolver.resolve(opt);
    const processor = cache.getProcessor(resolvedOptions);
    const processOptions = cache.getProcessOptions(resolvedOptions);
    return processor.process(
      rawSource,
      processOptions
    ).async();
  });
  handler.getPipeline = (opt) => {
    const resolvedOptions = resolver.resolve(opt);
    return cache.getPipeline(resolvedOptions);
  };
  return handler;
}





exports.createInjectPreflight = createInjectPreflight; exports.createStyleHandler = createStyleHandler; exports.createStylePipeline = createStylePipeline; exports.internalCssSelectorReplacer = internalCssSelectorReplacer;
