// tailwindcss-patch@8.7.3/dist/index.mjs
import {
  CacheStore,
  MIGRATION_REPORT_KIND,
  MIGRATION_REPORT_SCHEMA_VERSION,
  TailwindcssPatcher,
  VALIDATE_EXIT_CODES,
  VALIDATE_FAILURE_REASONS,
  ValidateCommandError,
  collectClassesFromContexts,
  collectClassesFromTailwindV4,
  createTailwindcssPatchCli,
  extractProjectCandidatesWithPositions,
  extractRawCandidates,
  extractRawCandidatesWithPositions,
  extractValidCandidates,
  getPatchStatusReport,
  groupTokensByFile,
  loadRuntimeContexts,
  logger_default,
  migrateConfigFiles,
  mountTailwindcssPatchCommands,
  normalizeOptions,
  restoreConfigFiles,
  runTailwindBuild,
  tailwindcssPatchCommands
} from "./chunk-3T6WSV7F.mjs";
import "./chunk-JHEI2MLC.mjs";

// src/index.ts
import { defineConfig } from "@tailwindcss-mangle/config";
export {
  CacheStore,
  MIGRATION_REPORT_KIND,
  MIGRATION_REPORT_SCHEMA_VERSION,
  TailwindcssPatcher,
  VALIDATE_EXIT_CODES,
  VALIDATE_FAILURE_REASONS,
  ValidateCommandError,
  collectClassesFromContexts,
  collectClassesFromTailwindV4,
  createTailwindcssPatchCli,
  defineConfig,
  extractProjectCandidatesWithPositions,
  extractRawCandidates,
  extractRawCandidatesWithPositions,
  extractValidCandidates,
  getPatchStatusReport,
  groupTokensByFile,
  loadRuntimeContexts,
  logger_default as logger,
  migrateConfigFiles,
  mountTailwindcssPatchCommands,
  normalizeOptions,
  restoreConfigFiles,
  runTailwindBuild,
  tailwindcssPatchCommands
};
