// tailwindcss-patch@8.7.3/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});
























var _chunkEFYAZO6Cjs = require('./chunk-EFYAZO6C.js');
require('./chunk-7JAOSSRO.js');

// src/index.ts
var _config = require('@tailwindcss-mangle/config');


























exports.CacheStore = _chunkEFYAZO6Cjs.CacheStore; exports.MIGRATION_REPORT_KIND = _chunkEFYAZO6Cjs.MIGRATION_REPORT_KIND; exports.MIGRATION_REPORT_SCHEMA_VERSION = _chunkEFYAZO6Cjs.MIGRATION_REPORT_SCHEMA_VERSION; exports.TailwindcssPatcher = _chunkEFYAZO6Cjs.TailwindcssPatcher; exports.VALIDATE_EXIT_CODES = _chunkEFYAZO6Cjs.VALIDATE_EXIT_CODES; exports.VALIDATE_FAILURE_REASONS = _chunkEFYAZO6Cjs.VALIDATE_FAILURE_REASONS; exports.ValidateCommandError = _chunkEFYAZO6Cjs.ValidateCommandError; exports.collectClassesFromContexts = _chunkEFYAZO6Cjs.collectClassesFromContexts; exports.collectClassesFromTailwindV4 = _chunkEFYAZO6Cjs.collectClassesFromTailwindV4; exports.createTailwindcssPatchCli = _chunkEFYAZO6Cjs.createTailwindcssPatchCli; exports.defineConfig = _config.defineConfig; exports.extractProjectCandidatesWithPositions = _chunkEFYAZO6Cjs.extractProjectCandidatesWithPositions; exports.extractRawCandidates = _chunkEFYAZO6Cjs.extractRawCandidates; exports.extractRawCandidatesWithPositions = _chunkEFYAZO6Cjs.extractRawCandidatesWithPositions; exports.extractValidCandidates = _chunkEFYAZO6Cjs.extractValidCandidates; exports.getPatchStatusReport = _chunkEFYAZO6Cjs.getPatchStatusReport; exports.groupTokensByFile = _chunkEFYAZO6Cjs.groupTokensByFile; exports.loadRuntimeContexts = _chunkEFYAZO6Cjs.loadRuntimeContexts; exports.logger = _chunkEFYAZO6Cjs.logger_default; exports.migrateConfigFiles = _chunkEFYAZO6Cjs.migrateConfigFiles; exports.mountTailwindcssPatchCommands = _chunkEFYAZO6Cjs.mountTailwindcssPatchCommands; exports.normalizeOptions = _chunkEFYAZO6Cjs.normalizeOptions; exports.restoreConfigFiles = _chunkEFYAZO6Cjs.restoreConfigFiles; exports.runTailwindBuild = _chunkEFYAZO6Cjs.runTailwindBuild; exports.tailwindcssPatchCommands = _chunkEFYAZO6Cjs.tailwindcssPatchCommands;
