// weapp-tailwindcss@4.10.3/dist/chunk-YJSFFRNZ.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }







var _chunk6Z4GEN2Yjs = require('./chunk-6Z4GEN2Y.js');


var _chunkLTJQUORKjs = require('./chunk-LTJQUORK.js');


var _chunkGCRL3ZYPjs = require('./chunk-GCRL3ZYP.js');











var _chunkDEIJXHGJjs = require('./chunk-DEIJXHGJ.js');


var _chunkDYLQ6UOIjs = require('./chunk-DYLQ6UOI.js');

// src/bundlers/vite/index.ts
var _path = require('path'); var _path2 = _interopRequireDefault(_path);
var _process = require('process'); var _process2 = _interopRequireDefault(_process);
var _htmltransform = require('@weapp-tailwindcss/postcss/html-transform'); var _htmltransform2 = _interopRequireDefault(_htmltransform);

// src/uni-app-x/transform.ts
var _compilerdom = require('@vue/compiler-dom');
var _compilersfc = require('@vue/compiler-sfc');
var _magicstring = require('magic-string'); var _magicstring2 = _interopRequireDefault(_magicstring);
function traverse(node, visitor) {
  visitor(node);
  if (Array.isArray(node.children)) {
    for (const child of node.children) {
      if (child && typeof child === "object" && "type" in child) {
        traverse(child, visitor);
      }
    }
  }
}
function updateStaticAttribute(ms, prop) {
  if (!prop.value) {
    return;
  }
  const start = prop.value.loc.start.offset + 1;
  const end = prop.value.loc.end.offset - 1;
  if (start < end) {
    ms.update(start, end, _chunkDEIJXHGJjs.replaceWxml.call(void 0, prop.value.content));
  }
}
function updateDirectiveExpression(ms, prop, jsHandler, runtimeSet) {
  if (_optionalChain([prop, 'access', _ => _.exp, 'optionalAccess', _2 => _2.type]) !== _compilerdom.NodeTypes.SIMPLE_EXPRESSION) {
    return;
  }
  const start = prop.exp.loc.start.offset;
  const end = prop.exp.loc.end.offset;
  if (start >= end) {
    return;
  }
  const generated = _chunkDEIJXHGJjs.generateCode.call(void 0, prop.exp.content, {
    jsHandler,
    runtimeSet,
    wrapExpression: true
  });
  ms.update(start, end, generated);
}
function shouldHandleAttribute(tag, attrName, disabledDefaultTemplateHandler, matchCustomAttribute) {
  const lowerName = attrName.toLowerCase();
  const shouldHandleDefault = !disabledDefaultTemplateHandler && lowerName === "class";
  const shouldHandleCustom = _nullishCoalesce(_optionalChain([matchCustomAttribute, 'optionalCall', _3 => _3(tag, attrName)]), () => ( false));
  return {
    shouldHandleDefault,
    shouldHandleCustom,
    shouldHandle: shouldHandleDefault || shouldHandleCustom
  };
}
var defaultCreateJsHandlerOptions = {
  babelParserOptions: {
    plugins: [
      "typescript"
    ]
  }
};
function transformUVue(code, id, jsHandler, runtimeSet, options = {}) {
  if (!/\.(?:uvue|nvue)(?:\?.*)?$/.test(id)) {
    return;
  }
  const { customAttributesEntities, disabledDefaultTemplateHandler = false } = options;
  const matchCustomAttribute = _chunkDEIJXHGJjs.createAttributeMatcher.call(void 0, customAttributesEntities);
  const ms = new (0, _magicstring2.default)(code);
  const { descriptor, errors } = _compilersfc.parse.call(void 0, code);
  if (errors.length === 0) {
    if (_optionalChain([descriptor, 'access', _4 => _4.template, 'optionalAccess', _5 => _5.ast])) {
      traverse(descriptor.template.ast, (node) => {
        if (node.type !== _compilerdom.NodeTypes.ELEMENT) {
          return;
        }
        const tag = node.tag;
        for (const prop of node.props) {
          if (prop.type === _compilerdom.NodeTypes.ATTRIBUTE) {
            const { shouldHandle, shouldHandleDefault } = shouldHandleAttribute(
              tag,
              prop.name,
              disabledDefaultTemplateHandler,
              matchCustomAttribute
            );
            if (!shouldHandle) {
              continue;
            }
            updateStaticAttribute(ms, prop);
            if (shouldHandleDefault) {
              continue;
            }
          } else if (prop.type === _compilerdom.NodeTypes.DIRECTIVE && prop.name === "bind" && _optionalChain([prop, 'access', _6 => _6.arg, 'optionalAccess', _7 => _7.type]) === _compilerdom.NodeTypes.SIMPLE_EXPRESSION && prop.arg.isStatic) {
            const attrName = prop.arg.content;
            const { shouldHandle } = shouldHandleAttribute(
              tag,
              attrName,
              disabledDefaultTemplateHandler,
              matchCustomAttribute
            );
            if (!shouldHandle) {
              continue;
            }
            updateDirectiveExpression(ms, prop, jsHandler, runtimeSet);
          }
        }
      });
    }
    if (descriptor.script) {
      const { code: code2 } = jsHandler(descriptor.script.content, _nullishCoalesce(runtimeSet, () => ( /* @__PURE__ */ new Set())), defaultCreateJsHandlerOptions);
      ms.update(descriptor.script.loc.start.offset, descriptor.script.loc.end.offset, code2);
    }
    if (descriptor.scriptSetup) {
      const { code: code2 } = jsHandler(descriptor.scriptSetup.content, _nullishCoalesce(runtimeSet, () => ( /* @__PURE__ */ new Set())), defaultCreateJsHandlerOptions);
      ms.update(descriptor.scriptSetup.loc.start.offset, descriptor.scriptSetup.loc.end.offset, code2);
    }
  }
  return {
    code: ms.toString(),
    // @ts-ignore
    get map() {
      return ms.generateMap();
    }
  };
}

// src/bundlers/vite/query.ts
function parseVueRequest(id) {
  const [filename, rawQuery] = id.split(`?`, 2);
  const searchParams = new URLSearchParams(rawQuery);
  const query = Object.fromEntries(searchParams);
  if (query.vue != null) {
    query.vue = true;
  }
  if (query.index != null) {
    query.index = Number(query.index);
  }
  if (query.raw != null) {
    query.raw = true;
  }
  if (query.url != null) {
    query.url = true;
  }
  if (query.scoped != null) {
    query.scoped = true;
  }
  const langTypeMatch = [...searchParams.keys()].find((key) => key.startsWith("lang."));
  const langType = query.lang || (langTypeMatch ? langTypeMatch.slice("lang.".length) : void 0);
  if (langType) {
    query.lang = langType;
  }
  return {
    filename,
    query
  };
}

// src/bundlers/vite/utils.ts


var _shared = require('@weapp-tailwindcss/shared');
function slash(p) {
  return _shared.ensurePosix.call(void 0, p);
}
var isWindows = _process2.default.platform === "win32";
var cssLangs = `\\.(css|less|sass|scss|styl|stylus|pcss|postcss)($|\\?)`;
var cssLangRE = new RegExp(cssLangs);
function isCSSRequest(request) {
  return cssLangRE.test(request);
}
function normalizePath(id) {
  return _path2.default.posix.normalize(isWindows ? _shared.ensurePosix.call(void 0, id) : id);
}
async function formatPostcssSourceMap(rawMap, file) {
  const inputFileDir = _path2.default.dirname(file);
  const sources = rawMap.sources.map((source) => {
    const cleanSource = _shared.cleanUrl.call(void 0, decodeURIComponent(source));
    if (cleanSource[0] === "<" && cleanSource.endsWith(">")) {
      return `\0${cleanSource}`;
    }
    return normalizePath(_path2.default.resolve(inputFileDir, cleanSource));
  });
  return {
    file,
    mappings: rawMap.mappings,
    names: rawMap.names,
    sources,
    sourcesContent: rawMap.sourcesContent,
    version: rawMap.version
  };
}

// src/uni-app-x/vite.ts
var preprocessorLangs = /* @__PURE__ */ new Set(["scss", "sass", "less", "styl", "stylus"]);
function isPreprocessorRequest(id, lang) {
  const normalizedLang = _optionalChain([lang, 'optionalAccess', _8 => _8.toLowerCase, 'call', _9 => _9()]);
  if (normalizedLang && preprocessorLangs.has(normalizedLang)) {
    return true;
  }
  const inlineLangMatch = id.match(/lang\.([a-z]+)/i);
  if (inlineLangMatch && preprocessorLangs.has(inlineLangMatch[1].toLowerCase())) {
    return true;
  }
  return /\.(?:scss|sass|less|styl|stylus)(?:\?|$)/i.test(id);
}
function createUniAppXPlugins(options) {
  const {
    appType,
    customAttributesEntities,
    disabledDefaultTemplateHandler,
    isIosPlatform: providedIosPlatform,
    mainCssChunkMatcher,
    runtimeState,
    styleHandler,
    jsHandler,
    ensureRuntimeClassSet,
    getResolvedConfig
  } = options;
  const isIosPlatform = _nullishCoalesce(providedIosPlatform, () => ( _chunkDYLQ6UOIjs.resolveUniUtsPlatform.call(void 0, ).isAppIos));
  async function transformStyle(code, id, query) {
    const parsed = _nullishCoalesce(query, () => ( parseVueRequest(id).query));
    if (isCSSRequest(id) || parsed.vue && parsed.type === "style") {
      const postcssResult = await styleHandler(code, {
        isMainChunk: mainCssChunkMatcher(id, appType),
        postcssOptions: {
          options: {
            from: id,
            map: {
              inline: false,
              annotation: false,
              // PostCSS 可能返回虚拟文件，因此需要启用这一项以获取源内容
              sourcesContent: true
              // 若上游预处理器已经生成 source map，sources 中可能出现重复条目
            }
          }
        }
      });
      const rawPostcssMap = postcssResult.map.toJSON();
      const postcssMap = await formatPostcssSourceMap(
        rawPostcssMap,
        _shared.cleanUrl.call(void 0, id)
      );
      return {
        code: postcssResult.css,
        map: postcssMap
      };
    }
  }
  const cssPrePlugin = {
    name: "weapp-tailwindcss:uni-app-x:css:pre",
    enforce: "pre",
    async transform(code, id) {
      await runtimeState.patchPromise;
      const { query } = parseVueRequest(id);
      const lang = query.lang;
      if (isIosPlatform && isPreprocessorRequest(id, lang)) {
        return;
      }
      return transformStyle(code, id, query);
    }
  };
  const cssPlugin = {
    name: "weapp-tailwindcss:uni-app-x:css",
    async transform(code, id) {
      await runtimeState.patchPromise;
      return transformStyle(code, id);
    }
  };
  const cssPlugins = [cssPlugin, cssPrePlugin];
  const nvuePlugin = {
    name: "weapp-tailwindcss:uni-app-x:nvue",
    enforce: "pre",
    async buildStart() {
      await ensureRuntimeClassSet(true);
    },
    async transform(code, id) {
      if (!/\.(?:uvue|nvue)(?:\?.*)?$/.test(id)) {
        return;
      }
      const resolvedConfig = getResolvedConfig();
      const isServeCommand = _optionalChain([resolvedConfig, 'optionalAccess', _10 => _10.command]) === "serve";
      const isWatchBuild = _optionalChain([resolvedConfig, 'optionalAccess', _11 => _11.command]) === "build" && !!_optionalChain([resolvedConfig, 'access', _12 => _12.build, 'optionalAccess', _13 => _13.watch]);
      const isNonWatchBuild = _optionalChain([resolvedConfig, 'optionalAccess', _14 => _14.command]) === "build" && !_optionalChain([resolvedConfig, 'access', _15 => _15.build, 'optionalAccess', _16 => _16.watch]);
      const shouldForceRefresh = isServeCommand || isWatchBuild || isNonWatchBuild;
      const currentRuntimeSet = shouldForceRefresh ? await ensureRuntimeClassSet(true) : await ensureRuntimeClassSet();
      const extraOptions = customAttributesEntities.length > 0 || disabledDefaultTemplateHandler ? {
        customAttributesEntities,
        disabledDefaultTemplateHandler
      } : void 0;
      if (extraOptions) {
        return transformUVue(code, id, jsHandler, currentRuntimeSet, extraOptions);
      }
      return transformUVue(code, id, jsHandler, currentRuntimeSet);
    },
    async handleHotUpdate(ctx) {
      const resolvedConfig = getResolvedConfig();
      if (_optionalChain([resolvedConfig, 'optionalAccess', _17 => _17.command]) !== "serve") {
        return;
      }
      if (!/\.(?:uvue|nvue)$/.test(ctx.file)) {
        return;
      }
      await ensureRuntimeClassSet(true);
    },
    async watchChange(id) {
      const resolvedConfig = getResolvedConfig();
      if (_optionalChain([resolvedConfig, 'optionalAccess', _18 => _18.command]) !== "build" || !_optionalChain([resolvedConfig, 'access', _19 => _19.build, 'optionalAccess', _20 => _20.watch])) {
        return;
      }
      if (!/\.(?:uvue|nvue)(?:\?.*)?$/.test(id)) {
        return;
      }
      await ensureRuntimeClassSet(true);
    }
  };
  return [
    ...cssPlugins,
    nvuePlugin
  ];
}
function createUniAppXAssetTask(file, originalSource, outDir, options) {
  return async () => {
    const {
      cache,
      hashKey,
      createHandlerOptions,
      debug: debug2,
      jsHandler,
      onUpdate,
      runtimeSet,
      applyLinkedResults: applyLinkedResults2
    } = options;
    const absoluteFile = _chunk6Z4GEN2Yjs.toAbsoluteOutputPath.call(void 0, file, outDir);
    const rawSource = originalSource.source.toString();
    const rawHashSource = options.hashSalt ? `${rawSource}
/*${options.hashSalt}*/` : rawSource;
    await _chunkLTJQUORKjs.processCachedTask.call(void 0, {
      cache,
      cacheKey: file,
      hashKey,
      rawSource: rawHashSource,
      applyResult(source) {
        originalSource.source = source;
      },
      onCacheHit() {
        debug2("js cache hit: %s", file);
      },
      async transform() {
        const currentSource = originalSource.source.toString();
        const { code, linked } = await jsHandler(currentSource, runtimeSet, createHandlerOptions(absoluteFile, {
          uniAppX: _nullishCoalesce(options.uniAppX, () => ( true)),
          babelParserOptions: {
            plugins: [
              "typescript"
            ],
            sourceType: "unambiguous"
          }
        }));
        onUpdate(file, currentSource, code);
        debug2("js handle: %s", file);
        applyLinkedResults2(linked);
        return {
          result: code
        };
      }
    });
  };
}

// src/bundlers/vite/generate-bundle.ts



// src/bundlers/vite/bundle-entries.ts
var _buffer = require('buffer');
function readOutputEntry(entry) {
  if (entry.output.type === "chunk") {
    return entry.output.code;
  }
  const source = entry.output.source;
  if (typeof source === "string") {
    return source;
  }
  if (source instanceof Uint8Array) {
    return _buffer.Buffer.from(source).toString();
  }
  const fallbackSource = source;
  if (fallbackSource == null) {
    return void 0;
  }
  if (typeof fallbackSource.toString === "function") {
    return fallbackSource.toString();
  }
  return void 0;
}
function isJavaScriptEntry(entry) {
  if (entry.output.type === "chunk") {
    return true;
  }
  return entry.fileName.endsWith(".js");
}
function createBundleModuleGraphOptions(outputDir, entries) {
  return {
    resolve(specifier, importer) {
      return _chunk6Z4GEN2Yjs.resolveOutputSpecifier.call(void 0, specifier, importer, outputDir, (candidate) => entries.has(candidate));
    },
    load(id) {
      const entry = entries.get(id);
      if (!entry) {
        return void 0;
      }
      return readOutputEntry(entry);
    },
    filter(id) {
      return entries.has(id);
    }
  };
}
function applyLinkedResults(linked, entries, onLinkedUpdate, onApplied) {
  if (!linked) {
    return;
  }
  for (const [id, { code }] of Object.entries(linked)) {
    const entry = entries.get(id);
    if (!entry) {
      continue;
    }
    const previous = readOutputEntry(entry);
    if (previous == null || previous === code) {
      continue;
    }
    if (entry.output.type === "chunk") {
      entry.output.code = code;
    } else {
      entry.output.source = code;
    }
    _optionalChain([onApplied, 'optionalCall', _21 => _21(entry, code)]);
    onLinkedUpdate(entry.fileName, previous, code);
  }
}

// src/bundlers/vite/js-precheck.ts
var FAST_JS_TRANSFORM_HINT_RE = /className\b|class\s*=|classList\.|\b(?:twMerge|clsx|classnames|cn|cva)\b|\[["'`]class["'`]\]|text-\[|bg-\[|\b(?:[whpm]|px|py|mx|my|rounded|flex|grid|gap)-/;
var DEPENDENCY_HINT_RE = /\bimport\s*[("'`{*]|\brequire\s*\(|\bexport\s+\*\s+from\s+["'`]|\bexport\s*\{[^}]*\}\s*from\s+["'`]/;
function shouldSkipViteJsTransform(rawSource, options) {
  if (!rawSource) {
    return true;
  }
  if (_optionalChain([options, 'optionalAccess', _22 => _22.alwaysEscape])) {
    return false;
  }
  if (_optionalChain([options, 'optionalAccess', _23 => _23.moduleSpecifierReplacements]) && Object.keys(options.moduleSpecifierReplacements).length > 0) {
    return false;
  }
  if (_optionalChain([options, 'optionalAccess', _24 => _24.wrapExpression])) {
    return false;
  }
  if (DEPENDENCY_HINT_RE.test(rawSource)) {
    return false;
  }
  return !FAST_JS_TRANSFORM_HINT_RE.test(rawSource);
}

// src/bundlers/vite/generate-bundle.ts
function formatDebugFileList(files, limit = 8) {
  if (files.size === 0) {
    return "-";
  }
  const sorted = [...files].sort();
  if (sorted.length <= limit) {
    return sorted.join(",");
  }
  return `${sorted.slice(0, limit).join(",")},...(+${sorted.length - limit})`;
}
function createEmptyMetric() {
  return {
    total: 0,
    transformed: 0,
    cacheHits: 0,
    elapsed: 0
  };
}
function createEmptyMetrics() {
  return {
    runtimeSet: 0,
    html: createEmptyMetric(),
    js: createEmptyMetric(),
    css: createEmptyMetric()
  };
}
function classifyEntry(file, opts) {
  if (opts.cssMatcher(file)) {
    return "css";
  }
  if (opts.htmlMatcher(file)) {
    return "html";
  }
  if (opts.jsMatcher(file) || opts.wxsMatcher(file)) {
    return "js";
  }
  return "other";
}
function readEntrySource(output) {
  if (output.type === "chunk") {
    return output.code;
  }
  return output.source.toString();
}
function toJsAbsoluteFilename(file, outDir) {
  return _chunk6Z4GEN2Yjs.toAbsoluteOutputPath.call(void 0, file, outDir);
}
function computeDirtyEntries(entries, opts, state) {
  const nextSourceHashByFile = /* @__PURE__ */ new Map();
  const changedByType = {
    html: /* @__PURE__ */ new Set(),
    js: /* @__PURE__ */ new Set(),
    css: /* @__PURE__ */ new Set(),
    other: /* @__PURE__ */ new Set()
  };
  for (const [file, output] of entries) {
    const type = classifyEntry(file, opts);
    const source = readEntrySource(output);
    const hash = opts.cache.computeHash(source);
    nextSourceHashByFile.set(file, hash);
    const previousHash = state.previousSourceHashByFile.get(file);
    if (previousHash == null || previousHash !== hash) {
      changedByType[type].add(file);
    }
  }
  return {
    sourceHashByFile: nextSourceHashByFile,
    changedByType
  };
}
function buildProcessSets(entries, opts, changedByType, previousLinkedByEntry, forceAll = false) {
  const processFiles = {
    html: /* @__PURE__ */ new Set(),
    js: /* @__PURE__ */ new Set(),
    css: /* @__PURE__ */ new Set()
  };
  const linkedImpactsByEntry = /* @__PURE__ */ new Map();
  if (forceAll) {
    for (const [file] of entries) {
      const type = classifyEntry(file, opts);
      if (type === "html" || type === "js" || type === "css") {
        processFiles[type].add(file);
      }
    }
    return {
      files: processFiles,
      linkedImpactsByEntry
    };
  }
  const firstRun = previousLinkedByEntry.size === 0;
  if (firstRun) {
    for (const [file] of entries) {
      const type = classifyEntry(file, opts);
      if (type === "html" || type === "js" || type === "css") {
        processFiles[type].add(file);
      }
    }
    return {
      files: processFiles,
      linkedImpactsByEntry
    };
  }
  for (const [file] of entries) {
    if (classifyEntry(file, opts) === "html") {
      processFiles.html.add(file);
    }
  }
  for (const file of changedByType.css) {
    processFiles.css.add(file);
  }
  for (const file of changedByType.js) {
    processFiles.js.add(file);
  }
  for (const changedFile of changedByType.js) {
    for (const [entryFile, linkedFiles] of previousLinkedByEntry.entries()) {
      if (linkedFiles.has(changedFile)) {
        processFiles.js.add(entryFile);
        let impacts = linkedImpactsByEntry.get(entryFile);
        if (!impacts) {
          impacts = /* @__PURE__ */ new Set();
          linkedImpactsByEntry.set(entryFile, impacts);
        }
        impacts.add(changedFile);
      }
    }
  }
  return {
    files: processFiles,
    linkedImpactsByEntry
  };
}
function measureElapsed(start) {
  return performance.now() - start;
}
function formatCacheHitRate(metric) {
  if (metric.total === 0) {
    return "0.00%";
  }
  return `${(metric.cacheHits / metric.total * 100).toFixed(2)}%`;
}
function formatMs(value) {
  return value.toFixed(2);
}
function summarizeStringDiff(previous, next) {
  if (previous === next) {
    return "same";
  }
  const previousLength = previous.length;
  const nextLength = next.length;
  const minLength = Math.min(previousLength, nextLength);
  let prefixLength = 0;
  while (prefixLength < minLength && previous.charCodeAt(prefixLength) === next.charCodeAt(prefixLength)) {
    prefixLength += 1;
  }
  let previousSuffixCursor = previousLength - 1;
  let nextSuffixCursor = nextLength - 1;
  while (previousSuffixCursor >= prefixLength && nextSuffixCursor >= prefixLength && previous.charCodeAt(previousSuffixCursor) === next.charCodeAt(nextSuffixCursor)) {
    previousSuffixCursor -= 1;
    nextSuffixCursor -= 1;
  }
  const previousChangedLength = previousSuffixCursor >= prefixLength ? previousSuffixCursor - prefixLength + 1 : 0;
  const nextChangedLength = nextSuffixCursor >= prefixLength ? nextSuffixCursor - prefixLength + 1 : 0;
  return `changed@${prefixLength} old=${previousChangedLength} new=${nextChangedLength} len=${previousLength}->${nextLength}`;
}
function createLinkedImpactSignature(entry, linkedImpactsByEntry, sourceHashByFile) {
  const changedLinkedFiles = linkedImpactsByEntry.get(entry);
  if (!changedLinkedFiles || changedLinkedFiles.size === 0) {
    return void 0;
  }
  const parts = [...changedLinkedFiles].sort().map((file) => {
    const hash = _nullishCoalesce(sourceHashByFile.get(file), () => ( "missing"));
    return `${file}:${hash}`;
  });
  return parts.join(",");
}
function createJsHashSalt(runtimeSignature, linkedImpactSignature) {
  if (!linkedImpactSignature) {
    return runtimeSignature;
  }
  return `${runtimeSignature}:linked:${linkedImpactSignature}`;
}
function hasRuntimeAffectingSourceChanges(changedByType) {
  return changedByType.html.size > 0 || changedByType.js.size > 0;
}
function createGenerateBundleHook(context) {
  const state = {
    iteration: 0,
    previousSourceHashByFile: /* @__PURE__ */ new Map(),
    previousLinkedByEntry: /* @__PURE__ */ new Map(),
    changedByType: {
      html: /* @__PURE__ */ new Set(),
      js: /* @__PURE__ */ new Set(),
      css: /* @__PURE__ */ new Set(),
      other: /* @__PURE__ */ new Set()
    }
  };
  return async function generateBundle(_opt, bundle) {
    const {
      opts,
      runtimeState,
      ensureRuntimeClassSet,
      debug: debug2,
      getResolvedConfig
    } = context;
    const {
      appType,
      cache,
      mainCssChunkMatcher,
      onEnd,
      onStart,
      onUpdate,
      styleHandler,
      templateHandler,
      jsHandler,
      uniAppX
    } = opts;
    await runtimeState.patchPromise;
    debug2("start");
    onStart();
    const metrics = createEmptyMetrics();
    const forceRuntimeRefreshByEnv = _process2.default.env.WEAPP_TW_VITE_FORCE_RUNTIME_REFRESH === "1";
    const disableDirtyOptimization = _process2.default.env.WEAPP_TW_VITE_DISABLE_DIRTY === "1";
    const disableJsPrecheck = _process2.default.env.WEAPP_TW_VITE_DISABLE_JS_PRECHECK === "1";
    const debugCssDiff = _process2.default.env.WEAPP_TW_VITE_DEBUG_CSS_DIFF === "1";
    const entries = Object.entries(bundle);
    const dirtyEntries = computeDirtyEntries(entries, opts, state);
    const forceRuntimeRefreshBySource = hasRuntimeAffectingSourceChanges(dirtyEntries.changedByType);
    const forceRuntimeRefresh = forceRuntimeRefreshByEnv || forceRuntimeRefreshBySource;
    const processSets = buildProcessSets(entries, opts, dirtyEntries.changedByType, state.previousLinkedByEntry, disableDirtyOptimization);
    const processFiles = processSets.files;
    debug2(
      "dirty iteration=%d html=%d[%s] js=%d[%s] css=%d[%s] other=%d[%s]",
      state.iteration + 1,
      dirtyEntries.changedByType.html.size,
      formatDebugFileList(dirtyEntries.changedByType.html),
      dirtyEntries.changedByType.js.size,
      formatDebugFileList(dirtyEntries.changedByType.js),
      dirtyEntries.changedByType.css.size,
      formatDebugFileList(dirtyEntries.changedByType.css),
      dirtyEntries.changedByType.other.size,
      formatDebugFileList(dirtyEntries.changedByType.other)
    );
    debug2(
      "process iteration=%d html=%d[%s] js=%d[%s] css=%d[%s]",
      state.iteration + 1,
      processFiles.html.size,
      formatDebugFileList(processFiles.html),
      processFiles.js.size,
      formatDebugFileList(processFiles.js),
      processFiles.css.size,
      formatDebugFileList(processFiles.css)
    );
    const resolvedConfig = getResolvedConfig();
    const rootDir = _optionalChain([resolvedConfig, 'optionalAccess', _25 => _25.root]) ? _path2.default.resolve(resolvedConfig.root) : _process2.default.cwd();
    const outDir = _optionalChain([resolvedConfig, 'optionalAccess', _26 => _26.build, 'optionalAccess', _27 => _27.outDir]) ? _path2.default.resolve(rootDir, resolvedConfig.build.outDir) : rootDir;
    const jsEntries = /* @__PURE__ */ new Map();
    for (const [fileName, output] of entries) {
      const entry = { fileName, output };
      if (isJavaScriptEntry(entry)) {
        const absolute = toJsAbsoluteFilename(fileName, outDir);
        jsEntries.set(absolute, entry);
      }
    }
    const moduleGraphOptions = createBundleModuleGraphOptions(outDir, jsEntries);
    const runtimeStart = performance.now();
    const runtime = await ensureRuntimeClassSet(forceRuntimeRefresh);
    metrics.runtimeSet = measureElapsed(runtimeStart);
    if (forceRuntimeRefreshBySource) {
      debug2(
        "runtimeSet forced refresh due to source changes: html=%d js=%d",
        dirtyEntries.changedByType.html.size,
        dirtyEntries.changedByType.js.size
      );
    }
    debug2("get runtimeSet, class count: %d", runtime.size);
    const runtimeSignature = _nullishCoalesce(_chunkDEIJXHGJjs.getRuntimeClassSetSignature.call(void 0, runtimeState.twPatcher), () => ( "runtime:missing"));
    const handleLinkedUpdate = (fileName, previous, next) => {
      onUpdate(fileName, previous, next);
      debug2("js linked handle: %s", fileName);
    };
    const pendingLinkedUpdates = [];
    const scheduleLinkedApply = (entry, code) => {
      pendingLinkedUpdates.push(() => {
        if (entry.output.type === "chunk") {
          entry.output.code = code;
        } else {
          entry.output.source = code;
        }
      });
    };
    const applyLinkedUpdates = (linked) => {
      applyLinkedResults(linked, jsEntries, handleLinkedUpdate, scheduleLinkedApply);
    };
    const createHandlerOptions = (absoluteFilename, extra) => ({
      ...extra,
      filename: absoluteFilename,
      tailwindcssMajorVersion: runtimeState.twPatcher.majorVersion,
      moduleGraph: moduleGraphOptions,
      babelParserOptions: {
        ..._nullishCoalesce(_optionalChain([extra, 'optionalAccess', _28 => _28.babelParserOptions]), () => ( {})),
        sourceFilename: absoluteFilename
      }
    });
    const linkedByEntry = /* @__PURE__ */ new Map();
    const tasks = [];
    const jsTaskFactories = [];
    for (const [file, originalSource] of entries) {
      const type = classifyEntry(file, opts);
      if (type === "html" && originalSource.type === "asset") {
        metrics.html.total++;
        if (!processFiles.html.has(file)) {
          continue;
        }
        const rawSource = originalSource.source.toString();
        tasks.push(
          _chunkLTJQUORKjs.processCachedTask.call(void 0, {
            cache,
            cacheKey: file,
            rawSource,
            hashKey: `${file}:html:${runtimeSignature}`,
            applyResult(source) {
              originalSource.source = source;
            },
            onCacheHit() {
              metrics.html.cacheHits++;
              debug2("html cache hit: %s", file);
            },
            async transform() {
              const start = performance.now();
              const transformed = await templateHandler(rawSource, {
                runtimeSet: runtime
              });
              metrics.html.elapsed += measureElapsed(start);
              metrics.html.transformed++;
              onUpdate(file, rawSource, transformed);
              debug2("html handle: %s", file);
              return {
                result: transformed
              };
            }
          })
        );
        continue;
      }
      if (type === "css" && originalSource.type === "asset") {
        metrics.css.total++;
        const rawSource = originalSource.source.toString();
        tasks.push(
          _chunkLTJQUORKjs.processCachedTask.call(void 0, {
            cache,
            cacheKey: file,
            rawSource,
            hashKey: `${file}:css:${runtimeSignature}:${_nullishCoalesce(runtimeState.twPatcher.majorVersion, () => ( "unknown"))}`,
            applyResult(source) {
              originalSource.source = source;
            },
            onCacheHit() {
              metrics.css.cacheHits++;
              debug2("css cache hit: %s", file);
            },
            async transform() {
              const start = performance.now();
              await runtimeState.patchPromise;
              const { css } = await styleHandler(rawSource, {
                isMainChunk: mainCssChunkMatcher(originalSource.fileName, appType),
                postcssOptions: {
                  options: {
                    from: file
                  }
                },
                majorVersion: runtimeState.twPatcher.majorVersion
              });
              if (debugCssDiff) {
                debug2("css diff %s: %s", file, summarizeStringDiff(rawSource, css));
              }
              metrics.css.elapsed += measureElapsed(start);
              metrics.css.transformed++;
              onUpdate(file, rawSource, css);
              debug2("css handle: %s", file);
              return {
                result: css
              };
            }
          })
        );
        continue;
      }
      if (type !== "js") {
        continue;
      }
      metrics.js.total++;
      const shouldTransformJs = processFiles.js.has(file);
      if (!shouldTransformJs) {
        debug2("js skip transform (clean), replay cache: %s", file);
      }
      if (originalSource.type === "chunk") {
        const absoluteFile = toJsAbsoluteFilename(file, outDir);
        const initialRawSource = originalSource.code;
        const linkedSet = /* @__PURE__ */ new Set();
        linkedByEntry.set(file, linkedSet);
        jsTaskFactories.push(async () => {
          const linkedImpactSignature = createLinkedImpactSignature(
            file,
            processSets.linkedImpactsByEntry,
            dirtyEntries.sourceHashByFile
          );
          const hashSalt = createJsHashSalt(runtimeSignature, linkedImpactSignature);
          await _chunkLTJQUORKjs.processCachedTask.call(void 0, {
            cache,
            cacheKey: file,
            hashKey: `${file}:js`,
            rawSource: `${initialRawSource}
/*${hashSalt}*/`,
            applyResult(source) {
              originalSource.code = source;
            },
            onCacheHit() {
              metrics.js.cacheHits++;
              debug2("js cache hit: %s", file);
            },
            async transform() {
              const start = performance.now();
              const rawSource = originalSource.code;
              if (!shouldTransformJs) {
                debug2("js cache replay miss, fallback transform: %s", file);
              }
              const handlerOptions = createHandlerOptions(absoluteFile);
              if (!disableJsPrecheck && shouldSkipViteJsTransform(rawSource, handlerOptions)) {
                metrics.js.elapsed += measureElapsed(start);
                metrics.js.transformed++;
                return {
                  result: rawSource
                };
              }
              const { code, linked } = await jsHandler(rawSource, runtime, handlerOptions);
              metrics.js.elapsed += measureElapsed(start);
              metrics.js.transformed++;
              onUpdate(file, rawSource, code);
              debug2("js handle: %s", file);
              if (linked) {
                for (const id of Object.keys(linked)) {
                  const linkedEntry = jsEntries.get(id);
                  if (linkedEntry) {
                    linkedSet.add(linkedEntry.fileName);
                  }
                }
              }
              applyLinkedUpdates(linked);
              return {
                result: code
              };
            }
          });
        });
      } else if (uniAppX && originalSource.type === "asset") {
        const linkedSet = /* @__PURE__ */ new Set();
        linkedByEntry.set(file, linkedSet);
        const baseApplyLinkedUpdates = applyLinkedUpdates;
        const wrappedApplyLinkedUpdates = (linked) => {
          if (linked) {
            for (const id of Object.keys(linked)) {
              const linkedEntry = jsEntries.get(id);
              if (linkedEntry) {
                linkedSet.add(linkedEntry.fileName);
              }
            }
          }
          baseApplyLinkedUpdates(linked);
        };
        const factory = createUniAppXAssetTask(
          file,
          originalSource,
          outDir,
          {
            cache,
            hashKey: `${file}:js`,
            hashSalt: createJsHashSalt(
              runtimeSignature,
              createLinkedImpactSignature(
                file,
                processSets.linkedImpactsByEntry,
                dirtyEntries.sourceHashByFile
              )
            ),
            createHandlerOptions,
            debug: debug2,
            jsHandler,
            onUpdate,
            runtimeSet: runtime,
            applyLinkedResults: wrappedApplyLinkedUpdates,
            uniAppX
          }
        );
        jsTaskFactories.push(async () => {
          const start = performance.now();
          if (!shouldTransformJs) {
            debug2("js skip transform (clean, uni-app-x), replay cache: %s", file);
            await factory();
            metrics.js.elapsed += measureElapsed(start);
            metrics.js.transformed++;
            return;
          }
          const currentSource = originalSource.source.toString();
          const absoluteFile = toJsAbsoluteFilename(file, outDir);
          const precheckOptions = createHandlerOptions(absoluteFile, {
            uniAppX: _nullishCoalesce(uniAppX, () => ( true)),
            babelParserOptions: {
              plugins: ["typescript"],
              sourceType: "unambiguous"
            }
          });
          if (!disableJsPrecheck && shouldSkipViteJsTransform(currentSource, precheckOptions)) {
            metrics.js.elapsed += measureElapsed(start);
            metrics.js.transformed++;
            return;
          }
          await factory();
          metrics.js.elapsed += measureElapsed(start);
          metrics.js.transformed++;
        });
      }
    }
    _chunk6Z4GEN2Yjs.pushConcurrentTaskFactories.call(void 0, tasks, jsTaskFactories);
    await Promise.all(tasks);
    for (const apply of pendingLinkedUpdates) {
      apply();
    }
    state.iteration += 1;
    state.previousSourceHashByFile = dirtyEntries.sourceHashByFile;
    state.changedByType = dirtyEntries.changedByType;
    const nextLinkedByEntry = new Map(state.previousLinkedByEntry);
    for (const [entryFile, linkedFiles] of linkedByEntry.entries()) {
      nextLinkedByEntry.set(entryFile, linkedFiles);
    }
    for (const entryFile of [...nextLinkedByEntry.keys()]) {
      const exists = entries.some(([fileName]) => fileName === entryFile);
      if (!exists) {
        nextLinkedByEntry.delete(entryFile);
      }
    }
    state.previousLinkedByEntry = nextLinkedByEntry;
    debug2(
      "metrics iteration=%d runtime=%sms html(total=%d transform=%d hit=%d rate=%s elapsed=%sms) js(total=%d transform=%d hit=%d rate=%s elapsed=%sms) css(total=%d transform=%d hit=%d rate=%s elapsed=%sms)",
      state.iteration,
      formatMs(metrics.runtimeSet),
      metrics.html.total,
      metrics.html.transformed,
      metrics.html.cacheHits,
      formatCacheHitRate(metrics.html),
      formatMs(metrics.html.elapsed),
      metrics.js.total,
      metrics.js.transformed,
      metrics.js.cacheHits,
      formatCacheHitRate(metrics.js),
      formatMs(metrics.js.elapsed),
      metrics.css.total,
      metrics.css.transformed,
      metrics.css.cacheHits,
      formatCacheHitRate(metrics.css),
      formatMs(metrics.css.elapsed)
    );
    onEnd();
    debug2("end");
  };
}

// src/bundlers/vite/rewrite-css-imports.ts
function joinPosixPath(base, subpath) {
  if (base.endsWith("/")) {
    return `${base}${subpath}`;
  }
  return `${base}/${subpath}`;
}
function isCssLikeImporter(importer) {
  if (!importer) {
    return false;
  }
  const normalized = _shared.cleanUrl.call(void 0, importer);
  return isCSSRequest(normalized);
}
function createRewriteCssImportsPlugins(options) {
  if (!options.shouldRewrite) {
    return [];
  }
  const { appType, weappTailwindcssDirPosix: weappTailwindcssDirPosix2 } = options;
  return [
    {
      name: `${_chunkDEIJXHGJjs.vitePluginName}:rewrite-css-imports`,
      enforce: "pre",
      resolveId: {
        order: "pre",
        handler(id, importer) {
          const replacement = _chunk6Z4GEN2Yjs.resolveTailwindcssImport.call(void 0, id, weappTailwindcssDirPosix2, {
            join: joinPosixPath,
            appType
          });
          if (!replacement) {
            return null;
          }
          if (importer && !isCssLikeImporter(importer)) {
            return null;
          }
          return replacement;
        }
      },
      transform: {
        order: "pre",
        handler(code, id) {
          if (!isCSSRequest(id)) {
            return null;
          }
          const rewritten = _chunk6Z4GEN2Yjs.rewriteTailwindcssImportsInCode.call(void 0, code, weappTailwindcssDirPosix2, {
            join: joinPosixPath,
            appType
          });
          if (!rewritten) {
            return null;
          }
          return {
            code: rewritten,
            map: null
          };
        }
      }
    }
  ];
}

// src/bundlers/vite/index.ts
var debug = _chunkDEIJXHGJjs.createDebug.call(void 0, );
var weappTailwindcssPackageDir = _chunk6Z4GEN2Yjs.resolvePackageDir.call(void 0, "weapp-tailwindcss");
var weappTailwindcssDirPosix = slash(weappTailwindcssPackageDir);
function UnifiedViteWeappTailwindcssPlugin(options = {}) {
  const rewriteCssImportsSpecified = Object.prototype.hasOwnProperty.call(options, "rewriteCssImports");
  const hasExplicitTailwindcssBasedir = typeof options.tailwindcssBasedir === "string" && options.tailwindcssBasedir.trim().length > 0;
  const opts = _chunkDEIJXHGJjs.getCompilerContext.call(void 0, options);
  const {
    disabled,
    customAttributes,
    onLoad,
    mainCssChunkMatcher,
    appType,
    styleHandler,
    jsHandler,
    twPatcher: initialTwPatcher,
    refreshTailwindcssPatcher,
    uniAppX,
    disabledDefaultTemplateHandler
  } = opts;
  const disabledOptions = _chunk6Z4GEN2Yjs.resolveDisabledOptions.call(void 0, disabled);
  const tailwindcssMajorVersion = _nullishCoalesce(initialTwPatcher.majorVersion, () => ( 0));
  const shouldRewriteCssImports = opts.rewriteCssImports !== false && !disabledOptions.rewriteCssImports && (rewriteCssImportsSpecified || tailwindcssMajorVersion >= 4);
  const rewritePlugins = createRewriteCssImportsPlugins({
    appType,
    shouldRewrite: shouldRewriteCssImports,
    weappTailwindcssDirPosix
  });
  if (disabledOptions.plugin) {
    return rewritePlugins.length ? rewritePlugins : void 0;
  }
  const customAttributesEntities = _chunkDEIJXHGJjs.toCustomAttributesEntities.call(void 0, customAttributes);
  const patchRecorderState = _chunkGCRL3ZYPjs.setupPatchRecorder.call(void 0, initialTwPatcher, opts.tailwindcssBasedir, {
    source: "runtime",
    cwd: _nullishCoalesce(opts.tailwindcssBasedir, () => ( _process2.default.cwd()))
  });
  const runtimeState = {
    twPatcher: initialTwPatcher,
    patchPromise: patchRecorderState.patchPromise,
    refreshTailwindcssPatcher,
    onPatchCompleted: patchRecorderState.onPatchCompleted
  };
  let runtimeSet;
  let runtimeSetPromise;
  let resolvedConfig;
  let runtimeRefreshSignature;
  let runtimeRefreshOptionsKey;
  function resolveRuntimeRefreshOptions() {
    const configPath = _optionalChain([runtimeState, 'access', _29 => _29.twPatcher, 'access', _30 => _30.options, 'optionalAccess', _31 => _31.tailwind, 'optionalAccess', _32 => _32.config]);
    const signature = _chunkDEIJXHGJjs.getRuntimeClassSetSignature.call(void 0, runtimeState.twPatcher);
    const optionsKey = JSON.stringify({
      appType,
      uniAppX: Boolean(uniAppX),
      customAttributesEntities,
      disabledDefaultTemplateHandler,
      configPath,
      rewriteCssImports: shouldRewriteCssImports
    });
    const changed = signature !== runtimeRefreshSignature || optionsKey !== runtimeRefreshOptionsKey;
    runtimeRefreshSignature = signature;
    runtimeRefreshOptionsKey = optionsKey;
    return {
      changed,
      signature,
      optionsKey
    };
  }
  async function refreshRuntimeState(force) {
    const invalidation = resolveRuntimeRefreshOptions();
    const shouldRefresh = force || invalidation.changed;
    const refreshed = await _chunkDEIJXHGJjs.refreshTailwindRuntimeState.call(void 0, runtimeState, {
      force: shouldRefresh,
      clearCache: force || invalidation.changed
    });
    if (invalidation.changed) {
      debug("runtime signature changed, refresh triggered. signature: %s", invalidation.signature);
    }
    if (refreshed) {
      runtimeSet = void 0;
      runtimeSetPromise = void 0;
    }
  }
  async function ensureRuntimeClassSet(force = false) {
    const forceRuntimeRefresh = force || _process2.default.env.WEAPP_TW_VITE_FORCE_RUNTIME_REFRESH === "1";
    await refreshRuntimeState(force);
    await runtimeState.patchPromise;
    if (!forceRuntimeRefresh && runtimeSet) {
      return runtimeSet;
    }
    if (forceRuntimeRefresh || !runtimeSetPromise) {
      const invalidation = resolveRuntimeRefreshOptions();
      const task2 = _chunkDEIJXHGJjs.collectRuntimeClassSet.call(void 0, runtimeState.twPatcher, {
        force: forceRuntimeRefresh || invalidation.changed,
        skipRefresh: forceRuntimeRefresh,
        clearCache: forceRuntimeRefresh || invalidation.changed
      });
      runtimeSetPromise = task2;
    }
    const task = runtimeSetPromise;
    try {
      runtimeSet = await task;
      return runtimeSet;
    } finally {
      if (runtimeSetPromise === task) {
        runtimeSetPromise = void 0;
      }
    }
  }
  onLoad();
  const getResolvedConfig = () => resolvedConfig;
  const utsPlatform = _chunkDYLQ6UOIjs.resolveUniUtsPlatform.call(void 0, );
  const isIosPlatform = utsPlatform.isAppIos;
  const uniAppXPlugins = uniAppX ? createUniAppXPlugins({
    appType,
    customAttributesEntities,
    disabledDefaultTemplateHandler,
    isIosPlatform,
    mainCssChunkMatcher,
    runtimeState,
    styleHandler,
    jsHandler,
    ensureRuntimeClassSet,
    getResolvedConfig
  }) : void 0;
  const plugins = [
    ...rewritePlugins,
    {
      name: `${_chunkDEIJXHGJjs.vitePluginName}:post`,
      enforce: "post",
      async configResolved(config) {
        resolvedConfig = config;
        const resolvedRoot = config.root ? _path2.default.resolve(config.root) : void 0;
        if (!hasExplicitTailwindcssBasedir && resolvedRoot && opts.tailwindcssBasedir !== resolvedRoot) {
          const previousBasedir = opts.tailwindcssBasedir;
          opts.tailwindcssBasedir = resolvedRoot;
          debug("align tailwindcss basedir with vite root: %s -> %s", _nullishCoalesce(previousBasedir, () => ( "undefined")), resolvedRoot);
          await refreshRuntimeState(true);
        }
        if (typeof config.css.postcss === "object" && Array.isArray(config.css.postcss.plugins)) {
          const idx = config.css.postcss.plugins.findIndex((x) => (
            // @ts-ignore
            x.postcssPlugin === "postcss-html-transform"
          ));
          if (idx > -1) {
            config.css.postcss.plugins.splice(idx, 1, _htmltransform2.default.call(void 0, ));
            debug("remove postcss-html-transform plugin from vite config");
          }
        }
      },
      generateBundle: createGenerateBundleHook({
        opts,
        runtimeState,
        ensureRuntimeClassSet,
        debug,
        getResolvedConfig
      })
    }
  ];
  if (uniAppXPlugins) {
    plugins.push(...uniAppXPlugins);
  }
  return plugins;
}



exports.UnifiedViteWeappTailwindcssPlugin = UnifiedViteWeappTailwindcssPlugin;
