// weapp-tailwindcss@4.10.3/dist/chunk-W2N6G2QQ.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }











var _chunkWF636Q5Ejs = require('./chunk-WF636Q5E.js');






var _chunk6Z4GEN2Yjs = require('./chunk-6Z4GEN2Y.js');


var _chunkLTJQUORKjs = require('./chunk-LTJQUORK.js');


var _chunkGCRL3ZYPjs = require('./chunk-GCRL3ZYP.js');






var _chunkDEIJXHGJjs = require('./chunk-DEIJXHGJ.js');


var _chunkDYLQ6UOIjs = require('./chunk-DYLQ6UOI.js');

// src/bundlers/webpack/BaseUnifiedPlugin/v5.ts
var _process = require('process'); var _process2 = _interopRequireDefault(_process);

// src/bundlers/webpack/BaseUnifiedPlugin/v5-assets.ts
var _path = require('path'); var _path2 = _interopRequireDefault(_path);

function resolveWebpackStaleClassNameFallback(option, _compiler) {
  if (typeof option === "boolean") {
    return option;
  }
  return false;
}
function setupWebpackV5ProcessAssetsHook(options) {
  const {
    compiler,
    options: compilerOptions,
    appType,
    runtimeState,
    debug: debug2
  } = options;
  const { Compilation, sources } = compiler.webpack;
  const { ConcatSource } = sources;
  compiler.hooks.compilation.tap(_chunkDEIJXHGJjs.pluginName, (compilation) => {
    compilation.hooks.processAssets.tapPromise(
      {
        name: _chunkDEIJXHGJjs.pluginName,
        stage: Compilation.PROCESS_ASSETS_STAGE_SUMMARIZE
      },
      async (assets) => {
        compilerOptions.onStart();
        debug2("start");
        await runtimeState.patchPromise;
        for (const chunk of compilation.chunks) {
          if (chunk.id && chunk.hash) {
            compilerOptions.cache.calcHashValueChanged(chunk.id, chunk.hash);
          }
        }
        const assetHashByChunk = _chunkWF636Q5Ejs.createAssetHashByChunkMap.call(void 0, compilation.chunks);
        const entries = Object.entries(assets);
        const compilerOutputPath = _nullishCoalesce(_optionalChain([compilation, 'access', _ => _.compiler, 'optionalAccess', _2 => _2.outputPath]), () => ( compiler.outputPath));
        const outputDir = compilerOutputPath ? _path2.default.resolve(compilerOutputPath) : _nullishCoalesce(_optionalChain([compilation, 'access', _3 => _3.outputOptions, 'optionalAccess', _4 => _4.path]), () => ( _process2.default.cwd()));
        const jsAssets = /* @__PURE__ */ new Map();
        for (const [file] of entries) {
          if (compilerOptions.jsMatcher(file) || compilerOptions.wxsMatcher(file)) {
            const absolute = _chunk6Z4GEN2Yjs.toAbsoluteOutputPath.call(void 0, file, outputDir);
            jsAssets.set(absolute, file);
          }
        }
        const moduleGraphOptions = {
          resolve(specifier, importer) {
            return _chunk6Z4GEN2Yjs.resolveOutputSpecifier.call(void 0, specifier, importer, outputDir, (candidate) => jsAssets.has(candidate));
          },
          load: (id) => {
            const assetName = jsAssets.get(id);
            if (!assetName) {
              return void 0;
            }
            const asset = compilation.getAsset(assetName);
            if (!asset) {
              return void 0;
            }
            const source = asset.source.source();
            return typeof source === "string" ? source : source.toString();
          },
          filter(id) {
            return jsAssets.has(id);
          }
        };
        const applyLinkedResults = (linked) => {
          if (!linked) {
            return;
          }
          for (const [id, { code }] of Object.entries(linked)) {
            const assetName = jsAssets.get(id);
            if (!assetName) {
              continue;
            }
            const asset = compilation.getAsset(assetName);
            if (!asset) {
              continue;
            }
            const previousSource = asset.source.source();
            const previous = typeof previousSource === "string" ? previousSource : previousSource.toString();
            if (previous === code) {
              continue;
            }
            const source = new ConcatSource(code);
            compilation.updateAsset(assetName, source);
            compilerOptions.onUpdate(assetName, previous, code);
            debug2("js linked handle: %s", assetName);
          }
        };
        const groupedEntries = _chunkDYLQ6UOIjs.getGroupedEntries.call(void 0, entries, compilerOptions);
        const staleClassNameFallback = resolveWebpackStaleClassNameFallback(compilerOptions.staleClassNameFallback, compiler);
        const runtimeSet = await _chunkDEIJXHGJjs.ensureRuntimeClassSet.call(void 0, runtimeState, {
          // webpack 的 script-only 热更新可能不会触发 runtime classset loader，
          // 这里强制收集可避免沿用上轮 class set，保证 JS 仅按最新集合精确命中。
          forceCollect: true,
          allowEmpty: false
        });
        debug2("get runtimeSet, class count: %d", runtimeSet.size);
        const tasks = [];
        if (Array.isArray(groupedEntries.html)) {
          for (const element of groupedEntries.html) {
            const [file, originalSource] = element;
            const rawSource = originalSource.source().toString();
            const cacheKey = file;
            const chunkHash = assetHashByChunk.get(file);
            tasks.push(
              _chunkLTJQUORKjs.processCachedTask.call(void 0, {
                cache: compilerOptions.cache,
                cacheKey,
                hashKey: `${file}:asset`,
                rawSource,
                hash: chunkHash,
                applyResult(source) {
                  compilation.updateAsset(file, source);
                },
                onCacheHit() {
                  debug2("html cache hit: %s", file);
                },
                transform: async () => {
                  const wxml = await compilerOptions.templateHandler(rawSource, {
                    runtimeSet
                  });
                  const source = new ConcatSource(wxml);
                  compilerOptions.onUpdate(file, rawSource, wxml);
                  debug2("html handle: %s", file);
                  return {
                    result: source
                  };
                }
              })
            );
          }
        }
        const jsTaskFactories = [];
        if (Array.isArray(groupedEntries.js)) {
          for (const [file] of groupedEntries.js) {
            const cacheKey = _chunkWF636Q5Ejs.getCacheKey.call(void 0, file);
            const asset = compilation.getAsset(file);
            if (!asset) {
              continue;
            }
            const absoluteFile = _chunk6Z4GEN2Yjs.toAbsoluteOutputPath.call(void 0, file, outputDir);
            const initialSource = asset.source.source();
            const initialRawSource = typeof initialSource === "string" ? initialSource : initialSource.toString();
            const chunkHash = assetHashByChunk.get(file);
            jsTaskFactories.push(async () => {
              await _chunkLTJQUORKjs.processCachedTask.call(void 0, {
                cache: compilerOptions.cache,
                cacheKey,
                hashKey: `${file}:asset`,
                rawSource: initialRawSource,
                hash: chunkHash,
                applyResult(source) {
                  compilation.updateAsset(file, source);
                },
                onCacheHit() {
                  debug2("js cache hit: %s", file);
                },
                transform: async () => {
                  const currentAsset = compilation.getAsset(file);
                  const currentSourceValue = _optionalChain([currentAsset, 'optionalAccess', _5 => _5.source, 'access', _6 => _6.source, 'call', _7 => _7()]);
                  const currentSource = typeof currentSourceValue === "string" ? currentSourceValue : _nullishCoalesce(_optionalChain([currentSourceValue, 'optionalAccess', _8 => _8.toString, 'call', _9 => _9()]), () => ( ""));
                  const { code, linked } = await compilerOptions.jsHandler(currentSource, runtimeSet, {
                    staleClassNameFallback,
                    tailwindcssMajorVersion: runtimeState.twPatcher.majorVersion,
                    filename: absoluteFile,
                    moduleGraph: moduleGraphOptions,
                    babelParserOptions: {
                      sourceFilename: absoluteFile
                    }
                  });
                  const source = new ConcatSource(code);
                  compilerOptions.onUpdate(file, currentSource, code);
                  debug2("js handle: %s", file);
                  applyLinkedResults(linked);
                  return {
                    result: source
                  };
                }
              });
            });
          }
        }
        if (Array.isArray(groupedEntries.css)) {
          for (const element of groupedEntries.css) {
            const [file, originalSource] = element;
            const rawSource = originalSource.source().toString();
            const cacheKey = file;
            const chunkHash = assetHashByChunk.get(file);
            tasks.push(
              _chunkLTJQUORKjs.processCachedTask.call(void 0, {
                cache: compilerOptions.cache,
                cacheKey,
                hashKey: `${file}:asset`,
                rawSource,
                hash: chunkHash,
                applyResult(source) {
                  compilation.updateAsset(file, source);
                },
                onCacheHit() {
                  debug2("css cache hit: %s", file);
                },
                transform: async () => {
                  await runtimeState.patchPromise;
                  const { css } = await compilerOptions.styleHandler(rawSource, {
                    isMainChunk: compilerOptions.mainCssChunkMatcher(file, appType),
                    postcssOptions: {
                      options: {
                        from: file
                      }
                    },
                    majorVersion: runtimeState.twPatcher.majorVersion
                  });
                  const source = new ConcatSource(css);
                  compilerOptions.onUpdate(file, rawSource, css);
                  debug2("css handle: %s", file);
                  return {
                    result: source
                  };
                }
              })
            );
          }
        }
        _chunk6Z4GEN2Yjs.pushConcurrentTaskFactories.call(void 0, tasks, jsTaskFactories);
        await Promise.all(tasks);
        debug2("end");
        compilerOptions.onEnd();
      }
    );
  });
}

// src/bundlers/webpack/BaseUnifiedPlugin/v5-loaders.ts
var _fs = require('fs'); var _fs2 = _interopRequireDefault(_fs);


function setupWebpackV5Loaders(options) {
  const {
    compiler,
    options: compilerOptions,
    appType,
    weappTailwindcssPackageDir: weappTailwindcssPackageDir2,
    shouldRewriteCssImports,
    runtimeLoaderPath,
    runtimeCssImportRewriteLoaderPath,
    getClassSetInLoader,
    debug: debug2
  } = options;
  const isMpxApp = _chunkWF636Q5Ejs.isMpx.call(void 0, appType);
  if (shouldRewriteCssImports && isMpxApp) {
    _chunkWF636Q5Ejs.ensureMpxTailwindcssAliases.call(void 0, compiler, weappTailwindcssPackageDir2);
  }
  const runtimeClassSetLoader = _nullishCoalesce(runtimeLoaderPath, () => ( _path2.default.resolve(__dirname, "./weapp-tw-runtime-classset-loader.js")));
  const runtimeCssImportRewriteLoader = shouldRewriteCssImports ? _nullishCoalesce(runtimeCssImportRewriteLoaderPath, () => ( _path2.default.resolve(__dirname, "./weapp-tw-css-import-rewrite-loader.js"))) : void 0;
  const runtimeClassSetLoaderExists = _fs2.default.existsSync(runtimeClassSetLoader);
  const runtimeCssImportRewriteLoaderExists = runtimeCssImportRewriteLoader ? _fs2.default.existsSync(runtimeCssImportRewriteLoader) : false;
  const runtimeLoaderRewriteOptions = shouldRewriteCssImports ? {
    pkgDir: weappTailwindcssPackageDir2,
    appType
  } : void 0;
  const classSetLoaderOptions = {
    getClassSet: getClassSetInLoader
  };
  const { findRewriteAnchor, findClassSetAnchor } = _chunkWF636Q5Ejs.createLoaderAnchorFinders.call(void 0, appType);
  const cssImportRewriteLoaderOptions = runtimeLoaderRewriteOptions ? {
    rewriteCssImports: runtimeLoaderRewriteOptions
  } : void 0;
  if (runtimeCssImportRewriteLoader && shouldRewriteCssImports && cssImportRewriteLoaderOptions && isMpxApp) {
    _chunkWF636Q5Ejs.injectMpxCssRewritePreRules.call(void 0, compiler, runtimeCssImportRewriteLoader, cssImportRewriteLoaderOptions);
  }
  const createRuntimeClassSetLoaderEntry = () => ({
    loader: runtimeClassSetLoader,
    options: classSetLoaderOptions,
    ident: null,
    type: null
  });
  const createCssImportRewriteLoaderEntry = () => {
    if (!runtimeCssImportRewriteLoader) {
      return null;
    }
    return {
      loader: runtimeCssImportRewriteLoader,
      options: cssImportRewriteLoaderOptions,
      ident: null,
      type: null
    };
  };
  const { NormalModule } = compiler.webpack;
  compiler.hooks.compilation.tap(_chunkDEIJXHGJjs.pluginName, (compilation) => {
    NormalModule.getCompilationHooks(compilation).loader.tap(_chunkDEIJXHGJjs.pluginName, (_loaderContext, module) => {
      const hasRuntimeLoader = runtimeClassSetLoaderExists || runtimeCssImportRewriteLoaderExists;
      if (!hasRuntimeLoader) {
        return;
      }
      _chunkWF636Q5Ejs.patchMpxLoaderResolve.call(void 0, _loaderContext, weappTailwindcssPackageDir2, shouldRewriteCssImports && isMpxApp);
      const loaderEntries = module.loaders || [];
      let rewriteAnchorIdx = findRewriteAnchor(loaderEntries);
      const classSetAnchorIdx = findClassSetAnchor(loaderEntries);
      const isCssModule = _chunkWF636Q5Ejs.isCssLikeModuleResource.call(void 0, module.resource, compilerOptions.cssMatcher, appType);
      if (_process2.default.env.WEAPP_TW_LOADER_DEBUG && isCssModule) {
        debug2("loader hook css module: %s loaders=%o anchors=%o", module.resource, loaderEntries.map((x) => x.loader), { rewriteAnchorIdx, classSetAnchorIdx });
      }
      if (_process2.default.env.WEAPP_TW_LOADER_DEBUG && typeof module.resource === "string" && module.resource.includes("app.css")) {
        debug2("app.css module loaders=%o anchors=%o", loaderEntries.map((x) => x.loader), { rewriteAnchorIdx, classSetAnchorIdx });
      } else if (_process2.default.env.WEAPP_TW_LOADER_DEBUG && typeof module.resource === "string" && module.resource.endsWith(".css")) {
        debug2("css module seen: %s loaders=%o anchors=%o", module.resource, loaderEntries.map((x) => x.loader), { rewriteAnchorIdx, classSetAnchorIdx });
      }
      if (rewriteAnchorIdx === -1 && classSetAnchorIdx === -1 && !isCssModule) {
        return;
      }
      const anchorlessInsert = (entry, position) => {
        if (position === "after") {
          loaderEntries.push(entry);
        } else {
          loaderEntries.unshift(entry);
        }
      };
      if (cssImportRewriteLoaderOptions && runtimeCssImportRewriteLoaderExists && runtimeCssImportRewriteLoader) {
        const existingIndex = loaderEntries.findIndex((entry) => _optionalChain([entry, 'access', _10 => _10.loader, 'optionalAccess', _11 => _11.includes, 'optionalCall', _12 => _12(runtimeCssImportRewriteLoader)]));
        const rewriteLoaderEntry = existingIndex !== -1 ? loaderEntries.splice(existingIndex, 1)[0] : createCssImportRewriteLoaderEntry();
        if (rewriteLoaderEntry) {
          const anchorIndex = findRewriteAnchor(loaderEntries);
          if (anchorIndex === -1) {
            anchorlessInsert(rewriteLoaderEntry, "after");
          } else {
            loaderEntries.splice(anchorIndex + 1, 0, rewriteLoaderEntry);
          }
          rewriteAnchorIdx = findRewriteAnchor(loaderEntries);
        }
      }
      if (runtimeClassSetLoaderExists && !_chunkWF636Q5Ejs.hasLoaderEntry.call(void 0, loaderEntries, runtimeClassSetLoader)) {
        const classSetLoaderEntry = createRuntimeClassSetLoaderEntry();
        const anchorIndex = findClassSetAnchor(loaderEntries);
        if (anchorIndex === -1) {
          anchorlessInsert(classSetLoaderEntry, "before");
        } else {
          const insertIndex = anchorIndex === -1 ? rewriteAnchorIdx : anchorIndex;
          loaderEntries.splice(insertIndex, 0, classSetLoaderEntry);
        }
      }
    });
  });
}

// src/bundlers/webpack/BaseUnifiedPlugin/v5.ts
var debug = _chunkDEIJXHGJjs.createDebug.call(void 0, );
var weappTailwindcssPackageDir = _chunk6Z4GEN2Yjs.resolvePackageDir.call(void 0, "weapp-tailwindcss");
var UnifiedWebpackPluginV5 = class {
  constructor(options = {}) {
    this.options = _chunkDEIJXHGJjs.getCompilerContext.call(void 0, options);
    this.appType = this.options.appType;
  }
  apply(compiler) {
    compiler.options = compiler.options || {};
    const {
      disabled,
      onLoad,
      runtimeLoaderPath,
      runtimeCssImportRewriteLoaderPath,
      twPatcher: initialTwPatcher,
      refreshTailwindcssPatcher
    } = this.options;
    const disabledOptions = _chunk6Z4GEN2Yjs.resolveDisabledOptions.call(void 0, disabled);
    const isTailwindcssV4 = (_nullishCoalesce(initialTwPatcher.majorVersion, () => ( 0))) >= 4;
    const shouldRewriteCssImports = isTailwindcssV4 && this.options.rewriteCssImports !== false && !disabledOptions.rewriteCssImports;
    const isMpxApp = _chunkWF636Q5Ejs.isMpx.call(void 0, this.appType);
    if (shouldRewriteCssImports) {
      _chunkWF636Q5Ejs.applyTailwindcssCssImportRewrite.call(void 0, compiler, {
        pkgDir: weappTailwindcssPackageDir,
        enabled: true,
        appType: this.appType
      });
      _chunkWF636Q5Ejs.setupMpxTailwindcssRedirect.call(void 0, weappTailwindcssPackageDir, isMpxApp);
    }
    if (disabledOptions.plugin) {
      return;
    }
    const patchRecorderState = _chunkGCRL3ZYPjs.setupPatchRecorder.call(void 0, initialTwPatcher, this.options.tailwindcssBasedir, {
      source: "runtime",
      cwd: _nullishCoalesce(this.options.tailwindcssBasedir, () => ( _process2.default.cwd()))
    });
    const runtimeState = {
      twPatcher: initialTwPatcher,
      patchPromise: patchRecorderState.patchPromise,
      refreshTailwindcssPatcher,
      onPatchCompleted: patchRecorderState.onPatchCompleted
    };
    let runtimeSetPrepared = false;
    let runtimeSetSignature;
    const resetRuntimePreparation = () => {
      runtimeSetPrepared = false;
    };
    if (_optionalChain([compiler, 'access', _13 => _13.hooks, 'access', _14 => _14.thisCompilation, 'optionalAccess', _15 => _15.tap])) {
      compiler.hooks.thisCompilation.tap(_chunkDEIJXHGJjs.pluginName, resetRuntimePreparation);
    } else if (_optionalChain([compiler, 'access', _16 => _16.hooks, 'access', _17 => _17.compilation, 'optionalAccess', _18 => _18.tap])) {
      compiler.hooks.compilation.tap(_chunkDEIJXHGJjs.pluginName, resetRuntimePreparation);
    }
    async function getClassSetInLoader() {
      if (runtimeSetPrepared) {
        return;
      }
      const signature = _chunkDEIJXHGJjs.getRuntimeClassSetSignature.call(void 0, runtimeState.twPatcher);
      const forceRefresh = signature !== runtimeSetSignature;
      runtimeSetPrepared = true;
      await _chunkDEIJXHGJjs.ensureRuntimeClassSet.call(void 0, runtimeState, {
        forceRefresh,
        forceCollect: true,
        clearCache: forceRefresh,
        allowEmpty: true
      });
      runtimeSetSignature = signature;
    }
    onLoad();
    setupWebpackV5Loaders({
      compiler,
      options: this.options,
      appType: this.appType,
      weappTailwindcssPackageDir,
      shouldRewriteCssImports,
      runtimeLoaderPath,
      runtimeCssImportRewriteLoaderPath,
      getClassSetInLoader,
      debug
    });
    setupWebpackV5ProcessAssetsHook({
      compiler,
      options: this.options,
      appType: this.appType,
      runtimeState,
      debug
    });
  }
};




exports.weappTailwindcssPackageDir = weappTailwindcssPackageDir; exports.UnifiedWebpackPluginV5 = UnifiedWebpackPluginV5;
