// weapp-tailwindcss@4.10.3/dist/chunk-RYC23C3K.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }

var _chunkDYLQ6UOIjs = require('./chunk-DYLQ6UOI.js');


var _chunkYUTKX7JZjs = require('./chunk-YUTKX7JZ.js');

// src/logger/index.ts
var _logger = require('@weapp-tailwindcss/logger');

// src/tailwindcss/v4/config.ts

var DEFAULT_CSS_CALC_CUSTOM_PROPERTIES = [];
function includesToken(list, token) {
  return list.some((candidate) => {
    if (typeof token === "string") {
      if (typeof candidate === "string") {
        return candidate === token;
      }
      candidate.lastIndex = 0;
      return candidate.test(token);
    }
    if (typeof candidate === "string") {
      token.lastIndex = 0;
      return token.test(candidate);
    }
    return candidate.source === token.source && candidate.flags === token.flags;
  });
}
function ensureDefaultsIncluded(value) {
  if (value === true) {
    return {
      includeCustomProperties: [...DEFAULT_CSS_CALC_CUSTOM_PROPERTIES]
    };
  }
  if (Array.isArray(value)) {
    if (!DEFAULT_CSS_CALC_CUSTOM_PROPERTIES.length) {
      return value;
    }
    const missing = DEFAULT_CSS_CALC_CUSTOM_PROPERTIES.filter((token) => !includesToken(value, token));
    return missing.length > 0 ? [...value, ...missing] : value;
  }
  if (value && typeof value === "object") {
    const include = value.includeCustomProperties;
    if (!Array.isArray(include)) {
      return {
        ...value,
        includeCustomProperties: [...DEFAULT_CSS_CALC_CUSTOM_PROPERTIES]
      };
    }
    if (!DEFAULT_CSS_CALC_CUSTOM_PROPERTIES.length) {
      return value;
    }
    const missing = DEFAULT_CSS_CALC_CUSTOM_PROPERTIES.filter((token) => !includesToken(include, token));
    return missing.length > 0 ? {
      ...value,
      includeCustomProperties: [...include, ...missing]
    } : value;
  }
  return value;
}
function normalizeCssEntriesConfig(entries) {
  if (!entries) {
    return void 0;
  }
  if (typeof entries === "string") {
    const trimmed = entries.trim();
    return trimmed ? [trimmed] : void 0;
  }
  if (!Array.isArray(entries)) {
    return void 0;
  }
  const normalized = entries.map((entry) => typeof entry === "string" ? entry.trim() : "").filter((entry) => entry.length > 0);
  return normalized.length > 0 ? normalized : void 0;
}
function hasConfiguredCssEntries(ctx) {
  if (normalizeCssEntriesConfig(ctx.cssEntries)) {
    return true;
  }
  if (normalizeCssEntriesConfig(_optionalChain([ctx, 'access', _ => _.tailwindcss, 'optionalAccess', _2 => _2.v4, 'optionalAccess', _3 => _3.cssEntries]))) {
    return true;
  }
  const patcherOptions = ctx.tailwindcssPatcherOptions;
  if (patcherOptions) {
    if (normalizeCssEntriesConfig(_optionalChain([patcherOptions, 'access', _4 => _4.tailwindcss, 'optionalAccess', _5 => _5.v4, 'optionalAccess', _6 => _6.cssEntries]))) {
      return true;
    }
    if (normalizeCssEntriesConfig(_optionalChain([patcherOptions, 'access', _7 => _7.tailwind, 'optionalAccess', _8 => _8.v4, 'optionalAccess', _9 => _9.cssEntries]))) {
      return true;
    }
    if (normalizeCssEntriesConfig(_optionalChain([patcherOptions, 'access', _10 => _10.patch, 'optionalAccess', _11 => _11.tailwindcss, 'optionalAccess', _12 => _12.v4, 'optionalAccess', _13 => _13.cssEntries]))) {
      return true;
    }
  }
  return false;
}
var hasWarnedMissingCssEntries = false;
function warnMissingCssEntries(ctx, patcher) {
  if (hasWarnedMissingCssEntries) {
    return;
  }
  if (_optionalChain([patcher, 'optionalAccess', _14 => _14.majorVersion]) !== 4) {
    return;
  }
  if (hasConfiguredCssEntries(ctx)) {
    return;
  }
  hasWarnedMissingCssEntries = true;
  _logger.logger.warn(
    '[tailwindcss@4] \u672A\u68C0\u6D4B\u5230 cssEntries \u914D\u7F6E\u3002\u8BF7\u4F20\u5165\u5305\u542B tailwindcss \u5F15\u7528\u7684 CSS \u7EDD\u5BF9\u8DEF\u5F84\uFF0C\u4F8B\u5982 cssEntries: ["/absolute/path/to/src/app.css"]\uFF0C\u5426\u5219 tailwindcss \u751F\u6210\u7684\u7C7B\u540D\u4E0D\u4F1A\u53C2\u4E0E\u8F6C\u8BD1\u3002'
  );
}
function applyV4CssCalcDefaults(cssCalc, patcher) {
  const cssCalcOptions = _nullishCoalesce(cssCalc, () => ( _optionalChain([patcher, 'optionalAccess', _15 => _15.majorVersion]) === 4));
  if (_optionalChain([patcher, 'optionalAccess', _16 => _16.majorVersion]) === 4 && cssCalcOptions) {
    return ensureDefaultsIncluded(cssCalcOptions);
  }
  return cssCalcOptions;
}

// src/tailwindcss/v4/patcher.ts


// src/tailwindcss/patcher.ts
var _path = require('path'); var _path2 = _interopRequireDefault(_path);
var _process = require('process'); var _process2 = _interopRequireDefault(_process);

var _shared = require('@weapp-tailwindcss/shared');
var _tailwindcsspatch = require('tailwindcss-patch');

// src/context/workspace.ts
var _fs = require('fs');

var IGNORED_WORKSPACE_DIRS = /* @__PURE__ */ new Set([
  "node_modules",
  ".git",
  ".hg",
  ".svn",
  ".turbo",
  ".output",
  ".next",
  "dist",
  "build"
]);
function findWorkspaceRoot(startDir) {
  if (!startDir) {
    return void 0;
  }
  let current = _path2.default.resolve(startDir);
  while (true) {
    const workspaceFile = _path2.default.join(current, "pnpm-workspace.yaml");
    if (_fs.existsSync.call(void 0, workspaceFile)) {
      return current;
    }
    const parent = _path2.default.dirname(current);
    if (parent === current) {
      return void 0;
    }
    current = parent;
  }
}
function findNearestPackageRoot(startDir) {
  if (!startDir) {
    return void 0;
  }
  let current = _path2.default.resolve(startDir);
  while (true) {
    const pkgPath = _path2.default.join(current, "package.json");
    if (_fs.existsSync.call(void 0, pkgPath)) {
      return current;
    }
    const parent = _path2.default.dirname(current);
    if (parent === current) {
      return void 0;
    }
    current = parent;
  }
}
function findWorkspacePackageDir(rootDir, packageName) {
  const visited = /* @__PURE__ */ new Set();
  const queue = [_path2.default.resolve(rootDir)];
  while (queue.length > 0) {
    const current = queue.shift();
    const normalized = _path2.default.normalize(current);
    if (visited.has(normalized)) {
      continue;
    }
    visited.add(normalized);
    try {
      const pkgPath = _path2.default.join(normalized, "package.json");
      if (_fs.existsSync.call(void 0, pkgPath)) {
        const pkg = JSON.parse(_fs.readFileSync.call(void 0, pkgPath, "utf8"));
        if (_optionalChain([pkg, 'optionalAccess', _17 => _17.name]) === packageName) {
          return normalized;
        }
      }
    } catch (e) {
    }
    let entries;
    try {
      entries = _fs.readdirSync.call(void 0, normalized, { withFileTypes: true });
    } catch (e2) {
      continue;
    }
    for (const entry of entries) {
      if (!entry.isDirectory() || IGNORED_WORKSPACE_DIRS.has(entry.name) || _optionalChain([entry, 'access', _18 => _18.isSymbolicLink, 'optionalCall', _19 => _19()])) {
        continue;
      }
      queue.push(_path2.default.join(normalized, entry.name));
    }
  }
  return void 0;
}

// src/tailwindcss/patcher-options.ts
function normalizeExtendLengthUnits(value) {
  if (value === false) {
    return false;
  }
  if (value === true) {
    return { enabled: true };
  }
  if (value && typeof value === "object") {
    return {
      enabled: true,
      ...value
    };
  }
  return void 0;
}
function normalizeTailwindcssPatcherOptions(options) {
  if (!options) {
    return void 0;
  }
  if ("patch" in options) {
    const { cache, patch } = options;
    const normalized = {};
    if (cache !== void 0) {
      normalized.cache = cache;
    }
    const apply = {};
    const extract = {};
    if (_optionalChain([patch, 'optionalAccess', _20 => _20.overwrite]) !== void 0) {
      apply.overwrite = patch.overwrite;
    }
    if (_optionalChain([patch, 'optionalAccess', _21 => _21.filter])) {
      normalized.filter = patch.filter;
    }
    const extendLengthUnits = normalizeExtendLengthUnits(_optionalChain([patch, 'optionalAccess', _22 => _22.applyPatches, 'optionalAccess', _23 => _23.extendLengthUnits]));
    const exposeContext = _optionalChain([patch, 'optionalAccess', _24 => _24.applyPatches, 'optionalAccess', _25 => _25.exportContext]);
    if (extendLengthUnits !== void 0 || exposeContext !== void 0) {
      apply.exposeContext = exposeContext;
      apply.extendLengthUnits = extendLengthUnits;
    }
    const cwd = _nullishCoalesce(_optionalChain([patch, 'optionalAccess', _26 => _26.cwd]), () => ( _optionalChain([patch, 'optionalAccess', _27 => _27.basedir])));
    if (cwd) {
      normalized.projectRoot = cwd;
    }
    const tailwindOptions = _optionalChain([patch, 'optionalAccess', _28 => _28.tailwindcss]) ? { ...patch.tailwindcss } : void 0;
    const legacyResolve = _optionalChain([patch, 'optionalAccess', _29 => _29.resolve]);
    let nextTailwindOptions = tailwindOptions;
    if (_optionalChain([nextTailwindOptions, 'optionalAccess', _30 => _30.version]) === 2 && !nextTailwindOptions.packageName) {
      nextTailwindOptions = {
        ...nextTailwindOptions,
        packageName: "@tailwindcss/postcss7-compat",
        postcssPlugin: nextTailwindOptions.postcssPlugin
      };
      if (!nextTailwindOptions.postcssPlugin) {
        nextTailwindOptions.postcssPlugin = "@tailwindcss/postcss7-compat";
      }
    }
    if (nextTailwindOptions || legacyResolve) {
      const resolveOptions = _optionalChain([nextTailwindOptions, 'optionalAccess', _31 => _31.resolve]);
      const mergedResolve = legacyResolve || resolveOptions ? {
        ..._nullishCoalesce(resolveOptions, () => ( {})),
        ..._nullishCoalesce(legacyResolve, () => ( {}))
      } : void 0;
      normalized.tailwindcss = {
        ..._nullishCoalesce(nextTailwindOptions, () => ( {})),
        ...mergedResolve ? { resolve: mergedResolve } : {}
      };
    }
    if (_optionalChain([patch, 'optionalAccess', _32 => _32.output])) {
      if (patch.output.filename !== void 0) {
        extract.file = patch.output.filename;
      }
      if (patch.output.loose !== void 0) {
        extract.pretty = patch.output.loose ? 2 : false;
      }
      if (patch.output.removeUniversalSelector !== void 0) {
        extract.removeUniversalSelector = patch.output.removeUniversalSelector;
      }
    }
    if (Object.keys(apply).length > 0) {
      normalized.apply = apply;
    }
    if (Object.keys(extract).length > 0) {
      normalized.extract = extract;
    }
    return toModernTailwindcssPatchOptions(normalized);
  }
  return toModernTailwindcssPatchOptions(options);
}
function toModernTailwindcssPatchOptions(options) {
  if (!options) {
    return void 0;
  }
  const normalized = {};
  if (options.cache !== void 0) {
    normalized.cache = options.cache;
  }
  if (options.filter) {
    normalized.filter = options.filter;
  }
  const projectRoot = _nullishCoalesce(options.projectRoot, () => ( options.cwd));
  if (projectRoot) {
    normalized.projectRoot = projectRoot;
  }
  const tailwindcss = _nullishCoalesce(options.tailwindcss, () => ( options.tailwind));
  if (tailwindcss) {
    normalized.tailwindcss = {
      ...tailwindcss
    };
  }
  const apply = {
    ..._nullishCoalesce(options.apply, () => ( {}))
  };
  if (apply.overwrite === void 0 && options.overwrite !== void 0) {
    apply.overwrite = options.overwrite;
  }
  if (apply.exposeContext === void 0 && _optionalChain([options, 'access', _33 => _33.features, 'optionalAccess', _34 => _34.exposeContext]) !== void 0) {
    apply.exposeContext = options.features.exposeContext;
  }
  if (apply.extendLengthUnits === void 0 && _optionalChain([options, 'access', _35 => _35.features, 'optionalAccess', _36 => _36.extendLengthUnits]) !== void 0) {
    apply.extendLengthUnits = options.features.extendLengthUnits;
  }
  if (Object.keys(apply).length > 0) {
    normalized.apply = apply;
  }
  const extract = {
    ..._nullishCoalesce(options.extract, () => ( {}))
  };
  if (options.output) {
    if (extract.write === void 0 && options.output.enabled !== void 0) {
      extract.write = options.output.enabled;
    }
    if (extract.file === void 0 && options.output.file !== void 0) {
      extract.file = options.output.file;
    }
    if (extract.format === void 0 && options.output.format !== void 0) {
      extract.format = options.output.format;
    }
    if (extract.pretty === void 0 && options.output.pretty !== void 0) {
      extract.pretty = options.output.pretty;
    }
    if (extract.removeUniversalSelector === void 0 && options.output.removeUniversalSelector !== void 0) {
      extract.removeUniversalSelector = options.output.removeUniversalSelector;
    }
  }
  if (Object.keys(extract).length > 0) {
    normalized.extract = extract;
  }
  return normalized;
}

// src/tailwindcss/patcher-resolve.ts

var _module = require('module');


var _url = require('url');
var GENERIC_RELATIVE_SPECIFIERS = [".", ".."];
var DEFAULT_TAILWIND_CONFIG_SPECIFIERS = [
  "stubs/config.full.js",
  "defaultConfig.js"
];
var TAILWIND_CONFIG_FILES = [
  "tailwind.config.js",
  "tailwind.config.cjs",
  "tailwind.config.mjs",
  "tailwind.config.ts",
  "tailwind.config.cts",
  "tailwind.config.mts"
];
function isPathSpecifier(specifier) {
  if (!specifier) {
    return false;
  }
  if (specifier.startsWith("file://")) {
    return true;
  }
  if (_path2.default.isAbsolute(specifier)) {
    return true;
  }
  return GENERIC_RELATIVE_SPECIFIERS.some((prefix) => specifier.startsWith(`${prefix}/`) || specifier.startsWith(`${prefix}\\`));
}
function resolveModuleFromPaths(specifier, paths) {
  if (!specifier || isPathSpecifier(specifier) || paths.length === 0) {
    return void 0;
  }
  try {
    const req = _module.createRequire.call(void 0, _chunkYUTKX7JZjs.importMetaUrl);
    return req.resolve(specifier, { paths });
  } catch (e3) {
    return void 0;
  }
}
function resolveTailwindConfigFallback(packageName, paths) {
  if (!packageName) {
    return void 0;
  }
  for (const suffix of DEFAULT_TAILWIND_CONFIG_SPECIFIERS) {
    const candidate = `${packageName}/${suffix}`;
    const resolved = resolveModuleFromPaths(candidate, paths);
    if (resolved) {
      return resolved;
    }
  }
  return void 0;
}
function appendNodeModules(paths, dir) {
  if (!dir) {
    return;
  }
  const nodeModulesDir = _path2.default.join(dir, "node_modules");
  if (_fs.existsSync.call(void 0, nodeModulesDir)) {
    paths.add(nodeModulesDir);
  }
}
function findTailwindConfig(searchRoots) {
  for (const root of searchRoots) {
    for (const file of TAILWIND_CONFIG_FILES) {
      const candidate = _path2.default.resolve(root, file);
      if (_fs.existsSync.call(void 0, candidate)) {
        return candidate;
      }
    }
  }
  return void 0;
}
function createDefaultResolvePaths(basedir) {
  const paths = /* @__PURE__ */ new Set();
  let fallbackCandidates = [];
  if (basedir) {
    const resolvedBase = _path2.default.resolve(basedir);
    appendNodeModules(paths, resolvedBase);
    fallbackCandidates.push(resolvedBase);
    const packageRoot = findNearestPackageRoot(resolvedBase);
    if (packageRoot) {
      appendNodeModules(paths, packageRoot);
      fallbackCandidates.push(packageRoot);
    }
  }
  const cwd = _process2.default.cwd();
  appendNodeModules(paths, cwd);
  try {
    const modulePath = _url.fileURLToPath.call(void 0, _chunkYUTKX7JZjs.importMetaUrl);
    const candidate = _fs.existsSync.call(void 0, modulePath) && !_path2.default.extname(modulePath) ? modulePath : _path2.default.dirname(modulePath);
    paths.add(candidate);
  } catch (e4) {
    paths.add(_chunkYUTKX7JZjs.importMetaUrl);
  }
  if (paths.size === 0) {
    fallbackCandidates = fallbackCandidates.filter(Boolean);
    if (fallbackCandidates.length === 0) {
      fallbackCandidates.push(cwd);
    }
    for (const candidate of fallbackCandidates) {
      paths.add(candidate);
    }
  }
  return [...paths];
}

// src/tailwindcss/patcher.ts
function createFallbackTailwindcssPatcher() {
  const packageInfo = {
    name: "tailwindcss",
    version: void 0,
    rootPath: "",
    packageJsonPath: "",
    packageJson: {}
  };
  return {
    packageInfo,
    async patch() {
      return {
        exposeContext: void 0,
        extendLengthUnits: void 0
      };
    },
    async getClassSet() {
      return /* @__PURE__ */ new Set();
    },
    async extract(_options) {
      const classSet = /* @__PURE__ */ new Set();
      return {
        classList: [],
        classSet
      };
    },
    async collectContentTokens() {
      return {
        entries: [],
        filesScanned: 0,
        sources: [],
        skippedFiles: []
      };
    }
  };
}
var hasLoggedMissingTailwind = false;
function createTailwindcssPatcher(options) {
  const { basedir, cacheDir, supportCustomLengthUnitsPatch, tailwindcss, tailwindcssPatcherOptions } = options || {};
  const cache = {
    driver: "memory"
  };
  const normalizedBasedir = basedir ? _path2.default.resolve(basedir) : void 0;
  const cacheRoot = _nullishCoalesce(_nullishCoalesce(findNearestPackageRoot(normalizedBasedir), () => ( normalizedBasedir)), () => ( _process2.default.cwd()));
  if (cacheDir) {
    if (_path2.default.isAbsolute(cacheDir)) {
      cache.dir = cacheDir;
    } else if (normalizedBasedir) {
      cache.dir = _path2.default.resolve(normalizedBasedir, cacheDir);
    } else {
      cache.dir = _path2.default.resolve(_process2.default.cwd(), cacheDir);
    }
  } else {
    cache.dir = _path2.default.join(cacheRoot, "node_modules", ".cache", "tailwindcss-patch");
  }
  if (normalizedBasedir) {
    cache.cwd = normalizedBasedir;
  }
  const resolvePaths = createDefaultResolvePaths(_nullishCoalesce(_nullishCoalesce(cache.cwd, () => ( normalizedBasedir)), () => ( _process2.default.cwd())));
  const normalizedUserOptions = normalizeTailwindcssPatcherOptions(tailwindcssPatcherOptions);
  const extendLengthUnits = normalizeExtendLengthUnits(_nullishCoalesce(supportCustomLengthUnitsPatch, () => ( true)));
  const baseTailwindOptions = _shared.defuOverrideArray.call(void 0, 
    _nullishCoalesce(tailwindcss, () => ( {})),
    {
      cwd: normalizedBasedir,
      resolve: {
        paths: resolvePaths
      }
    }
  );
  if (baseTailwindOptions.version === 2) {
    if (!baseTailwindOptions.packageName) {
      baseTailwindOptions.packageName = "@tailwindcss/postcss7-compat";
    }
    if (!baseTailwindOptions.postcssPlugin) {
      baseTailwindOptions.postcssPlugin = "@tailwindcss/postcss7-compat";
    }
  } else if (!baseTailwindOptions.packageName) {
    baseTailwindOptions.packageName = "tailwindcss";
  }
  if (!baseTailwindOptions.postcssPlugin) {
    baseTailwindOptions.postcssPlugin = baseTailwindOptions.version === 4 ? "@tailwindcss/postcss" : "tailwindcss";
  }
  if (typeof baseTailwindOptions.postcssPlugin === "string") {
    const resolvedPlugin = resolveModuleFromPaths(baseTailwindOptions.postcssPlugin, resolvePaths);
    if (resolvedPlugin) {
      baseTailwindOptions.postcssPlugin = resolvedPlugin;
    }
  }
  const baseOptions = {
    projectRoot: normalizedBasedir,
    cache,
    tailwindcss: baseTailwindOptions,
    apply: {
      exposeContext: true,
      extendLengthUnits
    }
  };
  const mergedOptions = _shared.defuOverrideArray.call(void 0, 
    _nullishCoalesce(normalizedUserOptions, () => ( {})),
    baseOptions
  );
  const resolvedOptions = _nullishCoalesce(toModernTailwindcssPatchOptions(mergedOptions), () => ( {}));
  const resolvedTailwindOptions = resolvedOptions.tailwindcss;
  if (resolvedTailwindOptions) {
    const existingResolve = _nullishCoalesce(resolvedTailwindOptions.resolve, () => ( {}));
    const customPaths = Array.isArray(existingResolve.paths) && existingResolve.paths.length > 0;
    const sourcePaths = customPaths ? existingResolve.paths : resolvePaths;
    resolvedTailwindOptions.resolve = {
      ...existingResolve,
      paths: Array.from(new Set(sourcePaths))
    };
    _logger.logger.debug("Tailwind resolve config %O", {
      packageName: resolvedTailwindOptions.packageName,
      version: resolvedTailwindOptions.version,
      resolve: resolvedTailwindOptions.resolve,
      cwd: resolvedTailwindOptions.cwd
    });
    if (typeof resolvedTailwindOptions.postcssPlugin === "string") {
      const resolvedPlugin = resolveModuleFromPaths(
        resolvedTailwindOptions.postcssPlugin,
        _nullishCoalesce(_optionalChain([resolvedTailwindOptions, 'access', _37 => _37.resolve, 'optionalAccess', _38 => _38.paths]), () => ( resolvePaths))
      );
      if (resolvedPlugin) {
        resolvedTailwindOptions.postcssPlugin = resolvedPlugin;
      }
    }
    const searchRoots = /* @__PURE__ */ new Set();
    if (resolvedTailwindOptions.cwd) {
      searchRoots.add(resolvedTailwindOptions.cwd);
    }
    for (const resolvePath of _nullishCoalesce(_optionalChain([resolvedTailwindOptions, 'access', _39 => _39.resolve, 'optionalAccess', _40 => _40.paths]), () => ( []))) {
      const parentDir = _path2.default.dirname(resolvePath);
      searchRoots.add(parentDir);
    }
    const configPath = findTailwindConfig(searchRoots);
    if (!resolvedTailwindOptions.config) {
      if (configPath) {
        resolvedTailwindOptions.config = configPath;
      } else {
        const fallbackConfig = resolveTailwindConfigFallback(
          resolvedTailwindOptions.packageName,
          _nullishCoalesce(resolvedTailwindOptions.resolve.paths, () => ( resolvePaths))
        );
        if (fallbackConfig) {
          resolvedTailwindOptions.config = fallbackConfig;
        }
      }
    }
    if (!resolvedTailwindOptions.cwd && configPath) {
      resolvedTailwindOptions.cwd = _path2.default.dirname(configPath);
    }
    resolvedOptions.tailwindcss = resolvedTailwindOptions;
  }
  try {
    return new (0, _tailwindcsspatch.TailwindcssPatcher)(resolvedOptions);
  } catch (error) {
    const searchPaths = _optionalChain([resolvedOptions, 'access', _41 => _41.tailwindcss, 'optionalAccess', _42 => _42.resolve, 'optionalAccess', _43 => _43.paths]);
    if (error instanceof Error && /tailwindcss not found/i.test(error.message)) {
      if (!hasLoggedMissingTailwind) {
        _logger.logger.warn("Tailwind CSS \u672A\u5B89\u88C5\uFF0C\u5DF2\u8DF3\u8FC7 Tailwind \u76F8\u5173\u8865\u4E01\u3002\u82E5\u9700\u4F7F\u7528 Tailwind \u80FD\u529B\uFF0C\u8BF7\u5B89\u88C5 tailwindcss\u3002");
        hasLoggedMissingTailwind = true;
      }
      return createFallbackTailwindcssPatcher();
    }
    if (error instanceof Error && /unable to locate tailwind css package/i.test(error.message)) {
      _logger.logger.error('\u65E0\u6CD5\u5B9A\u4F4D Tailwind CSS \u5305 "%s"\uFF0C\u5DF2\u5C1D\u8BD5\u8DEF\u5F84: %O', _optionalChain([resolvedOptions, 'access', _44 => _44.tailwindcss, 'optionalAccess', _45 => _45.packageName]), searchPaths);
    }
    throw error;
  }
}

// src/tailwindcss/v4/multi-patcher.ts
function createMultiTailwindcssPatcher(patchers) {
  if (patchers.length <= 1) {
    return patchers[0];
  }
  const [first] = patchers;
  const multiPatcher = {
    ...first,
    packageInfo: _optionalChain([first, 'optionalAccess', _46 => _46.packageInfo]),
    majorVersion: _optionalChain([first, 'optionalAccess', _47 => _47.majorVersion]),
    options: _optionalChain([first, 'optionalAccess', _48 => _48.options]),
    async patch() {
      let exposeContext;
      let extendLengthUnits;
      for (const patcher of patchers) {
        const result = await patcher.patch();
        if (_optionalChain([result, 'optionalAccess', _49 => _49.exposeContext]) && exposeContext == null) {
          exposeContext = result.exposeContext;
        }
        if (_optionalChain([result, 'optionalAccess', _50 => _50.extendLengthUnits]) && extendLengthUnits == null) {
          extendLengthUnits = result.extendLengthUnits;
        }
      }
      return {
        exposeContext,
        extendLengthUnits
      };
    },
    async getClassSet() {
      const aggregated = /* @__PURE__ */ new Set();
      for (const patcher of patchers) {
        const current = await patcher.getClassSet();
        for (const className of current) {
          aggregated.add(className);
        }
      }
      return aggregated;
    },
    async extract(options) {
      const aggregatedSet = /* @__PURE__ */ new Set();
      const aggregatedList = [];
      let filename;
      for (const patcher of patchers) {
        const result = await patcher.extract(options);
        if (!result) {
          continue;
        }
        if (filename === void 0 && result.filename) {
          filename = result.filename;
        }
        if (result.classList) {
          for (const className of result.classList) {
            if (!aggregatedSet.has(className)) {
              aggregatedList.push(className);
            }
            aggregatedSet.add(className);
          }
        }
        if (result.classSet) {
          for (const className of result.classSet) {
            aggregatedSet.add(className);
          }
        }
      }
      return {
        classList: aggregatedList,
        classSet: aggregatedSet,
        filename
      };
    }
  };
  if (patchers.every((patcher) => typeof patcher.getClassSetSync === "function")) {
    multiPatcher.getClassSetSync = () => {
      const aggregated = /* @__PURE__ */ new Set();
      for (const patcher of patchers) {
        const current = _optionalChain([patcher, 'access', _51 => _51.getClassSetSync, 'optionalCall', _52 => _52()]);
        if (!current) {
          continue;
        }
        for (const className of current) {
          aggregated.add(className);
        }
      }
      return aggregated;
    };
  }
  return multiPatcher;
}

// src/tailwindcss/v4/patcher-options.ts
function isLegacyTailwindcssPatcherOptions(options) {
  return typeof options === "object" && options !== null && "patch" in options;
}
function isModernTailwindcssPatchOptions(options) {
  return typeof options === "object" && options !== null && !("patch" in options);
}
function overrideTailwindcssPatcherOptionsForBase(options, baseDir, cssEntries) {
  const hasCssEntries = cssEntries.length > 0;
  if (!options) {
    return options;
  }
  if (isLegacyTailwindcssPatcherOptions(options)) {
    const patchOptions = options.patch;
    if (!patchOptions) {
      return options;
    }
    const nextPatch = {
      ...patchOptions,
      basedir: baseDir,
      cwd: _nullishCoalesce(patchOptions.cwd, () => ( baseDir))
    };
    if (patchOptions.tailwindcss) {
      const nextV4 = {
        ..._nullishCoalesce(patchOptions.tailwindcss.v4, () => ( {}))
      };
      if (hasCssEntries) {
        nextV4.cssEntries = cssEntries;
      } else {
        nextV4.cssEntries = _nullishCoalesce(nextV4.cssEntries, () => ( cssEntries));
        if (nextV4.base === void 0) {
          nextV4.base = baseDir;
        }
      }
      nextPatch.tailwindcss = {
        ...patchOptions.tailwindcss,
        v4: nextV4
      };
    }
    return {
      ...options,
      patch: nextPatch
    };
  }
  if (!isModernTailwindcssPatchOptions(options)) {
    return options;
  }
  const modernTailwind = _nullishCoalesce(options.tailwindcss, () => ( options.tailwind));
  if (!modernTailwind) {
    return options;
  }
  const { tailwind: _legacyTailwind, ...rest } = options;
  return {
    ...rest,
    tailwindcss: {
      ...modernTailwind,
      v4: {
        ..._nullishCoalesce(modernTailwind.v4, () => ( {})),
        ...hasCssEntries ? {} : { base: _nullishCoalesce(_optionalChain([modernTailwind, 'access', _53 => _53.v4, 'optionalAccess', _54 => _54.base]), () => ( baseDir)) },
        cssEntries: hasCssEntries ? cssEntries : _nullishCoalesce(_optionalChain([modernTailwind, 'access', _55 => _55.v4, 'optionalAccess', _56 => _56.cssEntries]), () => ( cssEntries))
      }
    }
  };
}

// src/tailwindcss/v4/css-entries.ts

function guessBasedirFromEntries(entries) {
  if (!entries) {
    return void 0;
  }
  for (const entry of entries) {
    if (typeof entry !== "string") {
      continue;
    }
    const trimmed = entry.trim();
    if (!trimmed || !_path2.default.isAbsolute(trimmed)) {
      continue;
    }
    const entryDir = _path2.default.dirname(trimmed);
    const resolved = _nullishCoalesce(findNearestPackageRoot(entryDir), () => ( entryDir));
    if (resolved) {
      return resolved;
    }
  }
  return void 0;
}
function normalizeCssEntries(entries, anchor) {
  if (!entries || entries.length === 0) {
    return void 0;
  }
  const normalized = /* @__PURE__ */ new Set();
  for (const entry of entries) {
    if (typeof entry !== "string") {
      continue;
    }
    const trimmed = entry.trim();
    if (trimmed.length === 0) {
      continue;
    }
    const resolved = _path2.default.isAbsolute(trimmed) ? _path2.default.normalize(trimmed) : _path2.default.normalize(_path2.default.resolve(anchor, trimmed));
    normalized.add(resolved);
  }
  return normalized.size > 0 ? [...normalized] : void 0;
}
function isSubPath(parent, child) {
  if (!parent || !child) {
    return false;
  }
  const relative = _path2.default.relative(parent, child);
  return relative === "" || !relative.startsWith("..") && !_path2.default.isAbsolute(relative);
}
function resolveCssEntryBase(entryDir, options) {
  const normalizedDir = _path2.default.normalize(entryDir);
  const { preferredBaseDir, workspaceRoot } = options;
  if (preferredBaseDir && isSubPath(preferredBaseDir, normalizedDir)) {
    return preferredBaseDir;
  }
  if (workspaceRoot && isSubPath(workspaceRoot, normalizedDir)) {
    return workspaceRoot;
  }
  const packageRoot = findNearestPackageRoot(normalizedDir);
  if (packageRoot) {
    return _path2.default.normalize(packageRoot);
  }
  return normalizedDir;
}
function groupCssEntriesByBase(entries, options = {}) {
  const normalizedOptions = {
    preferredBaseDir: options.preferredBaseDir ? _path2.default.normalize(options.preferredBaseDir) : void 0,
    workspaceRoot: options.workspaceRoot ? _path2.default.normalize(options.workspaceRoot) : void 0
  };
  const groups = /* @__PURE__ */ new Map();
  for (const entry of entries) {
    const entryDir = _path2.default.dirname(entry);
    const baseDir = resolveCssEntryBase(entryDir, normalizedOptions);
    const bucket = groups.get(baseDir);
    if (bucket) {
      bucket.push(entry);
    } else {
      groups.set(baseDir, [entry]);
    }
  }
  return groups;
}

// src/tailwindcss/v4/patcher.ts
function createPatcherForBase(baseDir, cssEntries, options) {
  const {
    tailwindcss,
    tailwindcssPatcherOptions,
    supportCustomLengthUnitsPatch
  } = options;
  const hasCssEntries = Boolean(_optionalChain([cssEntries, 'optionalAccess', _57 => _57.length]));
  const defaultTailwindcssConfig = {
    cwd: baseDir,
    v2: {
      cwd: baseDir
    },
    v3: {
      cwd: baseDir
    },
    v4: hasCssEntries ? { cssEntries } : {
      base: baseDir,
      cssEntries
    }
  };
  if (hasCssEntries && (tailwindcss == null || tailwindcss.version == null)) {
    defaultTailwindcssConfig.version = 4;
  }
  const mergedTailwindOptions = _chunkDYLQ6UOIjs.defuOverrideArray.call(void 0, 
    _nullishCoalesce(tailwindcss, () => ( {})),
    defaultTailwindcssConfig
  );
  if (!mergedTailwindOptions.v4) {
    mergedTailwindOptions.v4 = hasCssEntries ? { cssEntries: _nullishCoalesce(cssEntries, () => ( [])) } : {
      base: baseDir,
      cssEntries: _nullishCoalesce(cssEntries, () => ( []))
    };
  } else {
    if (!hasCssEntries && !mergedTailwindOptions.v4.base) {
      mergedTailwindOptions.v4.base = baseDir;
    }
    if (hasCssEntries) {
      if (_optionalChain([cssEntries, 'optionalAccess', _58 => _58.length])) {
        mergedTailwindOptions.v4.cssEntries = cssEntries;
      } else if (!mergedTailwindOptions.v4.cssEntries) {
        mergedTailwindOptions.v4.cssEntries = [];
      }
    } else if (!mergedTailwindOptions.v4.cssEntries) {
      mergedTailwindOptions.v4.cssEntries = _nullishCoalesce(cssEntries, () => ( []));
    }
  }
  const patchedOptions = overrideTailwindcssPatcherOptionsForBase(
    tailwindcssPatcherOptions,
    baseDir,
    _nullishCoalesce(cssEntries, () => ( []))
  );
  const configuredPackageName = _optionalChain([tailwindcss, 'optionalAccess', _59 => _59.packageName]) || _optionalChain([tailwindcssPatcherOptions, 'optionalAccess', _60 => _60.tailwindcss, 'optionalAccess', _61 => _61.packageName]) || _optionalChain([tailwindcssPatcherOptions, 'optionalAccess', _62 => _62.tailwind, 'optionalAccess', _63 => _63.packageName]) || _optionalChain([tailwindcssPatcherOptions, 'optionalAccess', _64 => _64.patch, 'optionalAccess', _65 => _65.tailwindcss, 'optionalAccess', _66 => _66.packageName]);
  const configuredVersion = _optionalChain([tailwindcss, 'optionalAccess', _67 => _67.version]) || _optionalChain([tailwindcssPatcherOptions, 'optionalAccess', _68 => _68.tailwindcss, 'optionalAccess', _69 => _69.version]) || _optionalChain([tailwindcssPatcherOptions, 'optionalAccess', _70 => _70.tailwind, 'optionalAccess', _71 => _71.version]) || _optionalChain([tailwindcssPatcherOptions, 'optionalAccess', _72 => _72.patch, 'optionalAccess', _73 => _73.tailwindcss, 'optionalAccess', _74 => _74.version]) || mergedTailwindOptions.version;
  const isTailwindcss4Package = (packageName) => Boolean(
    packageName && (packageName === "tailwindcss4" || packageName === "@tailwindcss/postcss" || packageName.includes("tailwindcss4"))
  );
  const isV4 = configuredVersion === 4 || mergedTailwindOptions.version === 4 || isTailwindcss4Package(_nullishCoalesce(configuredPackageName, () => ( mergedTailwindOptions.packageName)));
  const tailwindPackageConfigured = Boolean(configuredPackageName);
  const shouldPatchV4PostcssPackage = isV4 && !tailwindPackageConfigured;
  const packageCandidates = /* @__PURE__ */ new Set();
  if (shouldPatchV4PostcssPackage) {
    packageCandidates.add("@tailwindcss/postcss");
  }
  packageCandidates.add(
    _nullishCoalesce(_nullishCoalesce(mergedTailwindOptions.packageName, () => ( configuredPackageName)), () => ( "tailwindcss"))
  );
  const patchers = Array.from(packageCandidates).map((packageName) => {
    const tailwindOptionsForPackage = {
      ...mergedTailwindOptions,
      packageName
    };
    return createTailwindcssPatcher({
      basedir: baseDir,
      supportCustomLengthUnitsPatch: _nullishCoalesce(supportCustomLengthUnitsPatch, () => ( true)),
      tailwindcss: tailwindOptionsForPackage,
      tailwindcssPatcherOptions: patchedOptions
    });
  });
  return patchers.length === 1 ? patchers[0] : createMultiTailwindcssPatcher(patchers);
}
function tryCreateMultiTailwindcssPatcher(groups, options) {
  if (groups.size <= 1) {
    return void 0;
  }
  _logger.logger.debug("detected multiple Tailwind CSS entry bases: %O", [...groups.keys()]);
  const patchers = [];
  for (const [baseDir, entries] of groups) {
    patchers.push(createPatcherForBase(baseDir, entries, options));
  }
  return createMultiTailwindcssPatcher(patchers);
}

// src/context/tailwindcss/basedir.ts





var ENV_BASEDIR_KEYS = [
  "WEAPP_TAILWINDCSS_BASEDIR",
  "WEAPP_TAILWINDCSS_BASE_DIR",
  "TAILWINDCSS_BASEDIR",
  "TAILWINDCSS_BASE_DIR",
  "UNI_INPUT_DIR",
  "UNI_INPUT_ROOT",
  "UNI_CLI_ROOT",
  "UNI_APP_INPUT_DIR",
  "INIT_CWD",
  "PWD"
];
var GENERIC_ENV_BASEDIR_KEYS = /* @__PURE__ */ new Set(["INIT_CWD", "PWD"]);
function pickEnvBasedir() {
  for (const key of ENV_BASEDIR_KEYS) {
    const value = _process2.default.env[key];
    if (value && _path2.default.isAbsolute(value)) {
      return { key, value };
    }
  }
  return void 0;
}
function pickPackageEnvBasedir() {
  const packageJsonPath = _process2.default.env.npm_package_json;
  if (packageJsonPath) {
    const packageDir = _path2.default.dirname(packageJsonPath);
    if (packageDir && _path2.default.isAbsolute(packageDir)) {
      return packageDir;
    }
  }
  const localPrefix = _process2.default.env.npm_config_local_prefix;
  if (localPrefix && _path2.default.isAbsolute(localPrefix)) {
    return localPrefix;
  }
  return void 0;
}
function detectCallerBasedir() {
  const stack = new Error("resolveTailwindcssBasedir stack probe").stack;
  if (!stack) {
    return void 0;
  }
  if (_process2.default.env.WEAPP_TW_DEBUG_STACK === "1") {
    _logger.logger.debug("caller stack: %s", stack);
  }
  const lines = stack.split("\n");
  for (const line of lines) {
    const match = _nullishCoalesce(line.match(/\(([^)]+)\)/u), () => ( line.match(/at\s+(\S.*)$/u)));
    const location = _optionalChain([match, 'optionalAccess', _75 => _75[1]]);
    if (!location) {
      continue;
    }
    let filePath = location;
    if (filePath.startsWith("file://")) {
      try {
        filePath = _url.fileURLToPath.call(void 0, filePath);
      } catch (e5) {
        continue;
      }
    }
    const [candidate] = filePath.split(":");
    const resolvedPath = _path2.default.isAbsolute(filePath) ? filePath : candidate;
    if (!_path2.default.isAbsolute(resolvedPath)) {
      continue;
    }
    if (resolvedPath.includes("node_modules") && resolvedPath.includes("weapp-tailwindcss")) {
      continue;
    }
    try {
      return _path2.default.dirname(resolvedPath);
    } catch (e6) {
      continue;
    }
  }
  return void 0;
}
function resolveTailwindcssBasedir(basedir, fallback) {
  const envBasedirResult = pickEnvBasedir();
  const envBasedir = _optionalChain([envBasedirResult, 'optionalAccess', _76 => _76.value]);
  const envBasedirKey = _optionalChain([envBasedirResult, 'optionalAccess', _77 => _77.key]);
  const envBasedirIsGeneric = envBasedirKey ? GENERIC_ENV_BASEDIR_KEYS.has(envBasedirKey) : false;
  const packageEnvBasedir = pickPackageEnvBasedir();
  const shouldDetectCaller = !envBasedir || envBasedirIsGeneric;
  const callerBasedir = shouldDetectCaller ? detectCallerBasedir() : void 0;
  const cwd = _process2.default.cwd();
  const anchor = _nullishCoalesce(_nullishCoalesce(_nullishCoalesce(_nullishCoalesce(envBasedir, () => ( packageEnvBasedir)), () => ( fallback)), () => ( callerBasedir)), () => ( cwd));
  const resolveRelative = (value) => _path2.default.isAbsolute(value) ? _path2.default.normalize(value) : _path2.default.normalize(_path2.default.resolve(anchor, value));
  if (_process2.default.env.WEAPP_TW_DEBUG_STACK === "1") {
    _logger.logger.debug("resolveTailwindcssBasedir anchor %O", {
      basedir,
      envBasedir,
      envBasedirKey,
      envBasedirIsGeneric,
      packageEnvBasedir,
      fallback,
      callerBasedir,
      npm_package_json: _process2.default.env.npm_package_json,
      cwd,
      anchor
    });
  }
  if (basedir && basedir.trim().length > 0) {
    return resolveRelative(basedir);
  }
  if (envBasedir && !envBasedirIsGeneric) {
    return _path2.default.normalize(envBasedir);
  }
  if (fallback && fallback.trim().length > 0) {
    return resolveRelative(fallback);
  }
  if (packageEnvBasedir) {
    return _path2.default.normalize(packageEnvBasedir);
  }
  if (callerBasedir) {
    const normalizedCaller = _path2.default.normalize(callerBasedir);
    const librarySegment = `${_path2.default.sep}weapp-tailwindcss${_path2.default.sep}`;
    if (!normalizedCaller.includes(librarySegment)) {
      return normalizedCaller;
    }
  }
  const packageName = _process2.default.env.PNPM_PACKAGE_NAME;
  if (packageName) {
    try {
      const anchorRequire = _module.createRequire.call(void 0, _path2.default.join(anchor, "__resolve_tailwindcss_basedir__.cjs"));
      const packageJsonPath = anchorRequire.resolve(`${packageName}/package.json`);
      if (_process2.default.env.WEAPP_TW_DEBUG_STACK === "1") {
        _logger.logger.debug("package basedir resolved from PNPM_PACKAGE_NAME: %s", packageJsonPath);
      }
      return _path2.default.normalize(_path2.default.dirname(packageJsonPath));
    } catch (e7) {
      if (_process2.default.env.WEAPP_TW_DEBUG_STACK === "1") {
        _logger.logger.debug("failed to resolve package json for %s", packageName);
      }
      const workspaceRoot = findWorkspaceRoot(anchor);
      if (workspaceRoot) {
        const packageDir = findWorkspacePackageDir(workspaceRoot, packageName);
        if (packageDir) {
          return packageDir;
        }
      }
    }
  }
  if (envBasedir) {
    return _path2.default.normalize(envBasedir);
  }
  return _path2.default.normalize(cwd);
}

// src/context/tailwindcss.ts


// src/context/tailwindcss/rax.ts



function isRaxWorkspace(appType, baseDir) {
  if (appType === "rax") {
    return true;
  }
  try {
    const pkgPath = _path2.default.join(baseDir, "package.json");
    if (!_fs.existsSync.call(void 0, pkgPath)) {
      return false;
    }
    const pkg = JSON.parse(_fs.readFileSync.call(void 0, pkgPath, "utf8"));
    const deps = {
      ..._nullishCoalesce(pkg.dependencies, () => ( {})),
      ..._nullishCoalesce(pkg.devDependencies, () => ( {}))
    };
    if (deps["rax-app"] || deps.rax) {
      return true;
    }
  } catch (e8) {
    return false;
  }
  return false;
}
function collectRaxStyleEntries(baseDir) {
  const STYLE_CANDIDATES = [
    "src/global.css",
    "src/global.scss",
    "src/global.less",
    "src/global.sass",
    "src/global.styl",
    "src/global.stylus"
  ];
  const discovered = [];
  for (const relative of STYLE_CANDIDATES) {
    const candidate = _path2.default.resolve(baseDir, relative);
    if (_fs.existsSync.call(void 0, candidate)) {
      discovered.push(_path2.default.normalize(candidate));
    }
  }
  return discovered;
}
function detectImplicitCssEntries(appType, baseDir) {
  const baseCandidates = /* @__PURE__ */ new Set();
  baseCandidates.add(_path2.default.normalize(baseDir));
  const envCandidates = [_process2.default.cwd(), _process2.default.env.INIT_CWD, _process2.default.env.PWD];
  for (const candidate of envCandidates) {
    if (candidate) {
      baseCandidates.add(_path2.default.normalize(candidate));
    }
  }
  for (const candidateBase of baseCandidates) {
    if (!isRaxWorkspace(appType, candidateBase)) {
      continue;
    }
    const entries = collectRaxStyleEntries(candidateBase);
    if (entries.length) {
      return entries;
    }
  }
  return void 0;
}

// src/context/tailwindcss.ts
function createTailwindcssPatcherFromContext(ctx) {
  const {
    tailwindcssBasedir,
    supportCustomLengthUnitsPatch,
    tailwindcss,
    tailwindcssPatcherOptions,
    cssEntries: rawCssEntries,
    appType
  } = ctx;
  const absoluteCssEntryBasedir = guessBasedirFromEntries(rawCssEntries);
  const resolvedTailwindcssBasedir = resolveTailwindcssBasedir(tailwindcssBasedir, absoluteCssEntryBasedir);
  ctx.tailwindcssBasedir = resolvedTailwindcssBasedir;
  _logger.logger.debug("tailwindcss basedir resolved: %s", resolvedTailwindcssBasedir);
  let normalizedCssEntries = normalizeCssEntries(rawCssEntries, resolvedTailwindcssBasedir);
  if (!normalizedCssEntries) {
    normalizedCssEntries = detectImplicitCssEntries(ctx.appType, resolvedTailwindcssBasedir);
  }
  if (normalizedCssEntries) {
    ctx.cssEntries = normalizedCssEntries;
  }
  const shouldAttachBase = Boolean(ctx.tailwindcssBasedir && _optionalChain([normalizedCssEntries, 'optionalAccess', _78 => _78.length]));
  const tailwindcssWithBase = shouldAttachBase && _optionalChain([tailwindcss, 'optionalAccess', _79 => _79.v4]) !== null ? {
    ..._nullishCoalesce(tailwindcss, () => ( {})),
    v4: {
      ..._nullishCoalesce(_optionalChain([tailwindcss, 'optionalAccess', _80 => _80.v4]), () => ( {})),
      base: _nullishCoalesce(_optionalChain([tailwindcss, 'optionalAccess', _81 => _81.v4, 'optionalAccess', _82 => _82.base]), () => ( resolvedTailwindcssBasedir))
    }
  } : tailwindcss;
  const patcherOptions = {
    tailwindcss: tailwindcssWithBase,
    tailwindcssPatcherOptions,
    supportCustomLengthUnitsPatch,
    appType
  };
  const workspaceRoot = _nullishCoalesce(findWorkspaceRoot(resolvedTailwindcssBasedir), () => ( (absoluteCssEntryBasedir ? findWorkspaceRoot(absoluteCssEntryBasedir) : void 0)));
  const groupedCssEntries = normalizedCssEntries ? groupCssEntriesByBase(normalizedCssEntries, {
    preferredBaseDir: resolvedTailwindcssBasedir,
    workspaceRoot
  }) : void 0;
  const multiPatcher = groupedCssEntries ? tryCreateMultiTailwindcssPatcher(groupedCssEntries, patcherOptions) : void 0;
  if (multiPatcher) {
    return multiPatcher;
  }
  if (_optionalChain([groupedCssEntries, 'optionalAccess', _83 => _83.size]) === 1) {
    const firstGroup = groupedCssEntries.entries().next().value;
    if (firstGroup) {
      const [baseDir, entries] = firstGroup;
      return createPatcherForBase(baseDir, entries, patcherOptions);
    }
  }
  const effectiveCssEntries = _nullishCoalesce(normalizedCssEntries, () => ( rawCssEntries));
  return createPatcherForBase(
    resolvedTailwindcssBasedir,
    effectiveCssEntries,
    patcherOptions
  );
}









exports.findWorkspaceRoot = findWorkspaceRoot; exports.findNearestPackageRoot = findNearestPackageRoot; exports.warnMissingCssEntries = warnMissingCssEntries; exports.applyV4CssCalcDefaults = applyV4CssCalcDefaults; exports.resolveTailwindcssBasedir = resolveTailwindcssBasedir; exports.createTailwindcssPatcherFromContext = createTailwindcssPatcherFromContext; exports.logger = _logger.logger;
