// weapp-tailwindcss@4.10.3/dist/chunk-E7775SFS.mjs
import {
  applyV4CssCalcDefaults,
  createTailwindcssPatcherFromContext,
  findNearestPackageRoot,
  warnMissingCssEntries
} from "./chunk-QOTLDKI4.mjs";
import {
  getDefaultOptions
} from "./chunk-3WUHHFLF.mjs";
import {
  defuOverrideArray,
  isMap
} from "./chunk-OOHJLO5M.mjs";

// src/debug/index.ts
import _createDebug from "debug";
var _debug = _createDebug("weapp-tw");
function createDebug(prefix) {
  function debug3(formatter, ...args) {
    return _debug((prefix ?? "") + formatter, ...args);
  }
  return debug3;
}

// src/tailwindcss/runtime/cache.ts
import { statSync } from "fs";
var runtimeClassSetCache = /* @__PURE__ */ new WeakMap();
function getCacheEntry(twPatcher) {
  let entry = runtimeClassSetCache.get(twPatcher);
  if (!entry) {
    entry = {};
    runtimeClassSetCache.set(twPatcher, entry);
  }
  return entry;
}
function getTailwindConfigSignature(twPatcher) {
  const configPath = twPatcher.options?.tailwind?.config;
  if (typeof configPath !== "string" || configPath.length === 0) {
    return void 0;
  }
  try {
    const stats = statSync(configPath);
    return `${configPath}:${stats.size}:${stats.mtimeMs}`;
  } catch {
    return `${configPath}:missing`;
  }
}
function getPatchTargetSignature(twPatcher) {
  const packageInfo = twPatcher.packageInfo;
  return `${packageInfo?.rootPath ?? "missing"}:${packageInfo?.version ?? "unknown"}:${twPatcher.majorVersion ?? "unknown"}`;
}
function invalidateRuntimeClassSet(twPatcher) {
  if (!twPatcher) {
    return;
  }
  runtimeClassSetCache.delete(twPatcher);
}
function getRuntimeClassSetCacheEntry(twPatcher) {
  return getCacheEntry(twPatcher);
}
function getRuntimeClassSetSignature(twPatcher) {
  const configSignature = getTailwindConfigSignature(twPatcher) ?? "config:missing";
  const patchTargetSignature = getPatchTargetSignature(twPatcher);
  return `${configSignature}|patch:${patchTargetSignature}`;
}

// src/tailwindcss/runtime.ts
var debug = createDebug("[tailwindcss:runtime] ");
var refreshTailwindcssPatcherSymbol = /* @__PURE__ */ Symbol.for("weapp-tailwindcss.refreshTailwindcssPatcher");
function createTailwindPatchPromise(twPatcher, onPatched) {
  return Promise.resolve(twPatcher.patch()).then(async () => {
    invalidateRuntimeClassSet(twPatcher);
    if (onPatched) {
      try {
        await onPatched();
      } catch (error) {
        debug("failed to persist patch target after patch(): %O", error);
      }
    }
  });
}
var runtimeClassSetStateCache = /* @__PURE__ */ new WeakMap();
function getRuntimeClassSetStateEntry(state) {
  let entry = runtimeClassSetStateCache.get(state);
  if (!entry) {
    entry = {};
    runtimeClassSetStateCache.set(state, entry);
  }
  return entry;
}
async function refreshTailwindRuntimeState(state, forceOrOptions) {
  const normalizedOptions = typeof forceOrOptions === "boolean" ? { force: forceOrOptions } : forceOrOptions;
  const force = normalizedOptions.force;
  const clearCache = normalizedOptions.clearCache === true;
  if (!force) {
    return false;
  }
  await state.patchPromise;
  let refreshed = false;
  if (typeof state.refreshTailwindcssPatcher === "function") {
    const next = await state.refreshTailwindcssPatcher({ clearCache });
    if (next !== state.twPatcher) {
      state.twPatcher = next;
    }
    refreshed = true;
  }
  if (refreshed) {
    state.patchPromise = createTailwindPatchPromise(state.twPatcher, state.onPatchCompleted);
  }
  return refreshed;
}
async function ensureRuntimeClassSet(state, options = {}) {
  const forceRefresh = options.forceRefresh === true;
  const forceCollect = options.forceCollect === true;
  const clearCache = options.clearCache === true;
  const allowEmpty = options.allowEmpty === true;
  if (forceRefresh) {
    await refreshTailwindRuntimeState(state, {
      force: true,
      clearCache
    });
  }
  await state.patchPromise;
  const entry = getRuntimeClassSetStateEntry(state);
  const signature = getRuntimeClassSetSignature(state.twPatcher);
  const signatureChanged = entry.signature !== signature;
  const shouldForceCollect = forceCollect || forceRefresh || signatureChanged;
  if (!shouldForceCollect) {
    if (entry.value && (allowEmpty || entry.value.size > 0)) {
      return entry.value;
    }
    if (entry.promise) {
      return entry.promise;
    }
  }
  const task = (async () => {
    const collected = await collectRuntimeClassSet(state.twPatcher, {
      force: shouldForceCollect,
      skipRefresh: true,
      clearCache
    });
    if (allowEmpty || collected.size > 0) {
      return collected;
    }
    await refreshTailwindRuntimeState(state, {
      force: true,
      clearCache: true
    });
    await state.patchPromise;
    return collectRuntimeClassSet(state.twPatcher, {
      force: true,
      skipRefresh: true,
      clearCache: true
    });
  })();
  entry.promise = task;
  try {
    const runtimeSet = await task;
    entry.value = runtimeSet;
    entry.signature = getRuntimeClassSetSignature(state.twPatcher);
    return runtimeSet;
  } finally {
    if (entry.promise === task) {
      entry.promise = void 0;
    }
  }
}
function shouldPreferSync(majorVersion) {
  if (majorVersion == null) {
    return true;
  }
  if (majorVersion === 3) {
    return true;
  }
  if (majorVersion === 4) {
    return true;
  }
  return false;
}
function tryGetRuntimeClassSetSync(twPatcher) {
  if (typeof twPatcher.getClassSetSync !== "function") {
    return void 0;
  }
  if (!shouldPreferSync(twPatcher.majorVersion)) {
    return void 0;
  }
  try {
    const set = twPatcher.getClassSetSync();
    if (set && set.size === 0) {
      return void 0;
    }
    return set;
  } catch (error) {
    if (twPatcher.majorVersion === 4) {
      debug("getClassSetSync() unavailable for tailwindcss v4, fallback to async getClassSet(): %O", error);
    } else {
      debug("getClassSetSync() failed, fallback to async getClassSet(): %O", error);
    }
    return void 0;
  }
}
async function collectRuntimeClassSet(twPatcher, options = {}) {
  let activePatcher = twPatcher;
  if (options.force && !options.skipRefresh) {
    const refresh = activePatcher[refreshTailwindcssPatcherSymbol];
    if (typeof refresh === "function") {
      try {
        const refreshed = await refresh({ clearCache: options.clearCache === true });
        if (refreshed) {
          activePatcher = refreshed;
        }
      } catch (error) {
        debug("refreshTailwindcssPatcher failed, continuing with existing patcher: %O", error);
      }
    }
  }
  const entry = getRuntimeClassSetCacheEntry(activePatcher);
  const signature = getRuntimeClassSetSignature(activePatcher);
  if (!options.force) {
    if (entry.value && entry.signature === signature) {
      return entry.value;
    }
    if (entry.promise) {
      return entry.promise;
    }
  } else {
    entry.value = void 0;
  }
  const task = (async () => {
    const preExtractSyncSet = options.force ? tryGetRuntimeClassSetSync(activePatcher) : void 0;
    if (preExtractSyncSet) {
      debug("runtime class set snapshot via getClassSetSync() before extract(), size=%d", preExtractSyncSet.size);
    }
    const preferExtract = options.force === true;
    try {
      const result = await activePatcher.extract({ write: false });
      if (result?.classSet) {
        if (result.classSet.size > 0) {
          debug("runtime class set resolved via extract(), size=%d", result.classSet.size);
          return result.classSet;
        }
        if (preferExtract) {
          debug("runtime class set from extract() is empty on force collect, fallback to sync/async class set");
        } else {
          debug("runtime class set from extract() is empty, fallback to sync/async class set");
        }
      }
    } catch (error) {
      debug("extract() failed, fallback to getClassSet(): %O", error);
    }
    if (preExtractSyncSet) {
      debug("runtime class set fallback to pre-extract sync snapshot, size=%d", preExtractSyncSet.size);
      return preExtractSyncSet;
    }
    const syncSet = tryGetRuntimeClassSetSync(activePatcher);
    if (syncSet) {
      debug("runtime class set resolved via getClassSetSync(), size=%d", syncSet.size);
      return syncSet;
    }
    try {
      const fallbackSet = await Promise.resolve(activePatcher.getClassSet());
      if (fallbackSet) {
        debug("runtime class set resolved via getClassSet(), size=%d", fallbackSet.size);
        return fallbackSet;
      }
    } catch (error) {
      debug("getClassSet() failed, returning empty set: %O", error);
    }
    return /* @__PURE__ */ new Set();
  })();
  entry.promise = task;
  entry.signature = signature;
  try {
    const resolved = await task;
    entry.value = resolved;
    entry.promise = void 0;
    entry.signature = signature;
    return resolved;
  } catch (error) {
    entry.promise = void 0;
    throw error;
  }
}

// package.json
var package_default = {
  name: "weapp-tailwindcss",
  version: "4.10.3",
  description: "\u628A tailwindcss \u539F\u5B50\u5316\u6837\u5F0F\u601D\u60F3\uFF0C\u5E26\u7ED9\u5C0F\u7A0B\u5E8F\u5F00\u53D1\u8005\u4EEC! bring tailwindcss to miniprogram developers!",
  author: "ice breaker <1324318532@qq.com>",
  license: "MIT",
  homepage: "https://tw.icebreaker.top",
  repository: {
    type: "git",
    url: "git+https://github.com/sonofmagic/weapp-tailwindcss.git",
    directory: "packages/weapp-tailwindcss"
  },
  bugs: {
    url: "https://github.com/sonofmagic/weapp-tailwindcss/issues"
  },
  keywords: [
    "tailwindcss",
    "weapp",
    "wechat",
    "mini",
    "miniprogram",
    "mini app",
    "weapp-tw",
    "weapp-tailwindcss",
    "taro",
    "uni-app",
    "remax",
    "rax",
    "mpx",
    "jit",
    "mp",
    "android",
    "ios",
    "\u5C0F\u7A0B\u5E8F",
    "vite",
    "postcss",
    "webpack",
    "webpack-plugin",
    "gulp",
    "gulp-plugin"
  ],
  exports: {
    ".": {
      style: "./index.css",
      types: "./dist/index.d.ts",
      import: "./dist/index.mjs",
      require: "./dist/index.js"
    },
    "./escape": {
      types: "./dist/escape.d.ts",
      import: "./dist/escape.mjs",
      require: "./dist/escape.js"
    },
    "./vite": {
      types: "./dist/vite.d.ts",
      import: "./dist/vite.mjs",
      require: "./dist/vite.js"
    },
    "./webpack": {
      types: "./dist/webpack.d.ts",
      import: "./dist/webpack.mjs",
      require: "./dist/webpack.js"
    },
    "./webpack4": {
      types: "./dist/webpack4.d.ts",
      import: "./dist/webpack4.mjs",
      require: "./dist/webpack4.js"
    },
    "./core": {
      types: "./dist/core.d.ts",
      import: "./dist/core.mjs",
      require: "./dist/core.js"
    },
    "./gulp": {
      types: "./dist/gulp.d.ts",
      import: "./dist/gulp.mjs",
      require: "./dist/gulp.js"
    },
    "./defaults": {
      types: "./dist/defaults.d.ts",
      import: "./dist/defaults.mjs",
      require: "./dist/defaults.js"
    },
    "./presets": {
      types: "./dist/presets.d.ts",
      import: "./dist/presets.mjs",
      require: "./dist/presets.js"
    },
    "./reset": {
      types: "./dist/reset.d.ts",
      import: "./dist/reset.mjs",
      require: "./dist/reset.js"
    },
    "./css-macro/postcss": {
      types: "./dist/css-macro/postcss.d.ts",
      import: "./dist/css-macro/postcss.mjs",
      require: "./dist/css-macro/postcss.js"
    },
    "./css-macro": {
      types: "./dist/css-macro.d.ts",
      import: "./dist/css-macro.mjs",
      require: "./dist/css-macro.js"
    },
    "./types": {
      types: "./dist/types.d.ts",
      import: "./dist/types.mjs",
      require: "./dist/types.js"
    },
    "./postcss-html-transform": {
      types: "./dist/postcss-html-transform.d.ts",
      import: "./dist/postcss-html-transform.mjs",
      require: "./dist/postcss-html-transform.js"
    },
    "./package.json": "./package.json",
    "./index.css": "./index.css",
    "./index": "./index.css",
    "./preflight.css": "./preflight.css",
    "./preflight": "./preflight.css",
    "./theme.css": "./theme.css",
    "./theme": "./theme.css",
    "./utilities.css": "./utilities.css",
    "./utilities": "./utilities.css",
    "./with-layer.css": "./with-layer.css",
    "./with-layer": "./with-layer.css",
    "./uni-app-x": "./uni-app-x.css",
    "./uni-app-x.css": "./uni-app-x.css",
    "./css": "./css/index.css",
    "./*": "./*"
  },
  main: "./dist/index.js",
  module: "./dist/index.mjs",
  types: "./dist/index.d.ts",
  style: "index.css",
  typesVersions: {
    "*": {
      "*": [
        "./dist/*",
        "./dist/index.d.ts"
      ]
    }
  },
  bin: {
    "weapp-tailwindcss-webpack-plugin": "bin/weapp-tailwindcss.js",
    "weapp-tailwindcss": "bin/weapp-tailwindcss.js",
    "weapp-tw": "bin/weapp-tailwindcss.js"
  },
  files: [
    "bin",
    "css",
    "dist",
    "index.css",
    "preflight.css",
    "theme.css",
    "uni-app-x.css",
    "utilities.css",
    "with-layer.css"
  ],
  engines: {
    node: "^20.19.0 || >=22.12.0"
  },
  scripts: {
    dev: "tsup --watch --sourcemap",
    build: "tsup && node scripts/ensure-escape-dts.mjs",
    "build:tsc": "cross-env NODE_ENV=development tsc --build tsconfig.json",
    "build:cli": "cd plugins/cli && pnpm run build",
    "build:css": "tsx scripts/build-css.ts",
    "build:weapp-theme": "tsx scripts/build-weapp-theme.ts",
    test: "npm run postinstall && vitest run",
    "test:dev": "vitest",
    "test:ui": "vitest --ui",
    bench: "vitest bench --config ./vitest.config.ts",
    "bench:vite-perf": "tsx scripts/vite-perf-bench.ts",
    "bench:vite-perf:summary": "tsx scripts/vite-perf-summary.ts",
    "bench:js-handlers": "tsx scripts/js-bench.ts",
    "bench:js-diff": "tsx scripts/js-bench-diff.ts",
    tsd: "tsd",
    clean: "tsx scripts/clean.ts",
    "get-decl": "tsx scripts/get-decl.ts",
    "ls:pack": "npm pack --dry-run",
    "cli:patch": "node bin/weapp-tailwindcss.js patch",
    colors: "tsx scripts/colors.ts",
    release: "tsx scripts/release.ts",
    lint: "eslint .",
    "lint:fix": "eslint ./src --fix",
    postinstall: "node bin/weapp-tailwindcss.js patch",
    "bench:vite-dev-hmr": "tsx scripts/vite-dev-hmr-bench.ts",
    "test:watch-hmr": "tsx scripts/watch-hmr-regression/index.ts"
  },
  publishConfig: {
    access: "public",
    registry: "https://registry.npmjs.org"
  },
  dependencies: {
    "@ast-core/escape": "~1.0.1",
    "@babel/parser": "~7.29.0",
    "@babel/traverse": "~7.29.0",
    "@babel/types": "~7.29.0",
    "@tailwindcss-mangle/config": "^6.1.3",
    "@vue/compiler-dom": "catalog:vue3",
    "@vue/compiler-sfc": "catalog:vue3",
    "@weapp-core/escape": "~7.0.0",
    "@weapp-core/regex": "~1.0.1",
    "@weapp-tailwindcss/logger": "workspace:*",
    "@weapp-tailwindcss/postcss": "workspace:*",
    "@weapp-tailwindcss/shared": "workspace:*",
    cac: "^6.7.14",
    debug: "~4.4.3",
    "fast-glob": "^3.3.3",
    htmlparser2: "10.1.0",
    "loader-utils": "2.0.4",
    "local-pkg": "^1.1.2",
    "lru-cache": "10.4.3",
    "magic-string": "0.30.21",
    semver: "~7.7.4",
    "tailwindcss-patch": "catalog:tailwindcssPatch",
    "webpack-sources": "3.3.4",
    yaml: "^2.8.2"
  }
};

// src/constants.ts
var pluginName = "weapp-tailwindcss-webpack-plugin";
var vitePluginName = "weapp-tailwindcss:adaptor";
var WEAPP_TW_REQUIRED_NODE_VERSION = "18.17.0";
var WEAPP_TW_VERSION = package_default.version;
var DEFAULT_RUNTIME_PACKAGE_REPLACEMENTS = {
  "tailwind-merge": "@weapp-tailwindcss/merge",
  "class-variance-authority": "@weapp-tailwindcss/cva",
  "tailwind-variants": "@weapp-tailwindcss/variants"
};

// src/context/index.ts
import { rm } from "fs/promises";
import { logger as logger5, pc } from "@weapp-tailwindcss/logger";

// src/cache/index.ts
import { LRUCache } from "lru-cache";

// src/cache/md5.ts
import { md5 } from "@weapp-tailwindcss/shared/node";

// src/cache/index.ts
function isProcessResult(value) {
  return typeof value === "object" && value !== null && "result" in value;
}
function createCache(options) {
  const disabled = options === false;
  const hashMap = /* @__PURE__ */ new Map();
  const instance = new LRUCache({
    // 可能会添加和删除一些页面和组件, 先设定 1024 吧
    max: 1024,
    ttl: 0,
    ttlAutopurge: false
  });
  const cache = {
    hashMap,
    instance,
    hasHashKey(key) {
      return hashMap.has(key);
    },
    getHashValue(key) {
      return hashMap.get(key);
    },
    setHashValue(key, value) {
      return hashMap.set(key, value);
    },
    get(key) {
      return instance.get(key);
    },
    set(key, value) {
      return instance.set(key, value);
    },
    computeHash(message) {
      return md5(message);
    },
    calcHashValueChanged(key, hash) {
      const hit = hashMap.get(key);
      if (hit) {
        hashMap.set(key, {
          changed: hash !== hit.hash,
          hash
        });
      } else {
        hashMap.set(key, {
          changed: true,
          hash
        });
      }
      return cache;
    },
    has(key) {
      return instance.has(key);
    },
    async process({
      key,
      hashKey,
      rawSource,
      hash,
      resolveCache,
      transform,
      onCacheHit
    }) {
      if (disabled) {
        const value2 = await transform();
        return isProcessResult(value2) ? value2.result : value2;
      }
      const cacheHashKey = hashKey ?? key;
      let hasChanged = true;
      if (hash != null || rawSource != null) {
        const nextHash = hash ?? cache.computeHash(rawSource);
        cache.calcHashValueChanged(cacheHashKey, nextHash);
        const entry = cache.getHashValue(cacheHashKey);
        hasChanged = entry?.changed ?? true;
      }
      const readCache = resolveCache ?? (() => cache.get(key));
      if (!hasChanged) {
        const cached = readCache();
        if (cached !== void 0) {
          await onCacheHit?.(cached);
          return cached;
        }
      }
      const value = await transform();
      const normalized = isProcessResult(value) ? value : { result: value };
      const stored = normalized.cacheValue ?? normalized.result;
      cache.set(key, stored);
      return normalized.result;
    }
  };
  return cache;
}
function initializeCache(cacheConfig) {
  if (typeof cacheConfig === "boolean" || cacheConfig === void 0) {
    return createCache(cacheConfig);
  }
  return cacheConfig;
}

// src/tailwindcss/targets.ts
import process3 from "process";
import { logger as logger2 } from "@weapp-tailwindcss/logger";

// src/tailwindcss/targets/paths.ts
import { existsSync } from "fs";
import path from "path";
var PATCH_INFO_FILENAME = "tailwindcss-target.json";
var PATCH_INFO_CACHE_RELATIVE_PATH = path.join("node_modules", ".cache", "weapp-tailwindcss", PATCH_INFO_FILENAME);
var PATCH_INFO_LEGACY_RELATIVE_PATH = path.join(".tw-patch", PATCH_INFO_FILENAME);
function formatRelativeToBase(targetPath, baseDir) {
  if (!baseDir) {
    return path.normalize(targetPath);
  }
  const relative = path.relative(baseDir, targetPath);
  if (!relative || relative === ".") {
    return ".";
  }
  if (relative.startsWith("..")) {
    return path.normalize(targetPath);
  }
  return path.join(".", relative);
}
function resolveRecordLocation(baseDir) {
  const normalizedBase = path.normalize(baseDir);
  const packageRoot = findNearestPackageRoot(normalizedBase) ?? normalizedBase;
  const packageJsonPath = path.join(packageRoot, "package.json");
  const hasPackageJson = existsSync(packageJsonPath);
  const recordKeySource = hasPackageJson ? packageJsonPath : normalizedBase;
  const recordKey = md5(path.normalize(recordKeySource));
  const recordDir = path.join(packageRoot, "node_modules", ".cache", "weapp-tailwindcss", recordKey);
  const recordPath = path.join(recordDir, PATCH_INFO_FILENAME);
  return {
    normalizedBase,
    packageRoot,
    recordDir,
    recordKey,
    recordPath,
    packageJsonPath: hasPackageJson ? packageJsonPath : void 0
  };
}
function getRecordFileCandidates(baseDir) {
  const { normalizedBase, packageRoot, recordPath } = resolveRecordLocation(baseDir);
  const candidates = /* @__PURE__ */ new Set([
    recordPath,
    path.join(packageRoot, PATCH_INFO_CACHE_RELATIVE_PATH),
    path.join(normalizedBase, PATCH_INFO_CACHE_RELATIVE_PATH),
    path.join(normalizedBase, PATCH_INFO_LEGACY_RELATIVE_PATH)
  ]);
  return [...candidates];
}

// src/tailwindcss/targets/record-io.ts
import { existsSync as existsSync2, readFileSync } from "fs";
import { mkdir, writeFile } from "fs/promises";
import path2 from "path";
import process from "process";
import { logger } from "@weapp-tailwindcss/logger";
var loggedInvalidPatchRecords = /* @__PURE__ */ new Set();
function warnInvalidPatchTargetRecord(baseDir, recordPath, reason) {
  const normalizedPath = path2.normalize(recordPath);
  if (loggedInvalidPatchRecords.has(normalizedPath)) {
    return;
  }
  loggedInvalidPatchRecords.add(normalizedPath);
  const fileDisplay = formatRelativeToBase(normalizedPath, baseDir);
  const baseDisplay = formatRelativeToBase(path2.normalize(baseDir), process.cwd());
  const reasonMessage = reason ? `\uFF1A${reason}` : "";
  logger.warn(
    `\u68C0\u6D4B\u5230\u635F\u574F\u7684 Tailwind CSS \u76EE\u6807\u8BB0\u5F55 ${fileDisplay}${reasonMessage}\u3002\u8BF7\u5728 ${baseDisplay} \u91CD\u65B0\u6267\u884C "weapp-tw patch --record-target" \u6216\u5220\u9664\u8BE5\u6587\u4EF6\u540E\u518D\u8FD0\u884C\u3002`
  );
}
function readPatchTargetRecord(baseDir) {
  if (!baseDir) {
    return void 0;
  }
  const normalizedBase = path2.normalize(baseDir);
  for (const recordPath of getRecordFileCandidates(normalizedBase)) {
    if (!existsSync2(recordPath)) {
      continue;
    }
    try {
      const content = readFileSync(recordPath, "utf8");
      const parsed = JSON.parse(content);
      if (!parsed || typeof parsed.tailwindPackagePath !== "string") {
        warnInvalidPatchTargetRecord(normalizedBase, recordPath, "\u7F3A\u5C11 tailwindPackagePath \u5B57\u6BB5");
        continue;
      }
      return {
        baseDir: normalizedBase,
        path: recordPath,
        record: parsed
      };
    } catch (error) {
      const reason = error instanceof Error ? error.message : String(error);
      warnInvalidPatchTargetRecord(normalizedBase, recordPath, reason);
      continue;
    }
  }
  return void 0;
}
async function saveCliPatchTargetRecord(baseDir, patcher, options) {
  if (!baseDir || !patcher?.packageInfo?.rootPath) {
    return void 0;
  }
  const normalizedBase = path2.normalize(baseDir);
  const location = resolveRecordLocation(normalizedBase);
  const recordPath = options?.recordPath ? path2.normalize(options.recordPath) : location.recordPath;
  const record = {
    tailwindPackagePath: path2.normalize(patcher.packageInfo.rootPath),
    packageVersion: patcher.packageInfo.version,
    recordedAt: (/* @__PURE__ */ new Date()).toISOString(),
    source: options?.source ?? "cli",
    tailwindcssBasedir: normalizedBase,
    cwd: options?.cwd ? path2.normalize(options.cwd) : normalizedBase,
    patchVersion: WEAPP_TW_VERSION,
    packageJsonPath: options?.packageJsonPath ?? location.packageJsonPath,
    recordKey: options?.recordKey ?? location.recordKey
  };
  try {
    await mkdir(path2.dirname(recordPath), { recursive: true });
    await writeFile(recordPath, `${JSON.stringify(record, null, 2)}
`, "utf8");
    return recordPath;
  } catch (error) {
    const baseDisplay = formatRelativeToBase(normalizedBase, process.cwd());
    logger.warn(
      '\u81EA\u52A8\u66F4\u65B0 Tailwind CSS \u8865\u4E01\u8BB0\u5F55\u5931\u8D25\uFF0C\u8BF7\u5728 %s \u8FD0\u884C "weapp-tw patch --cwd %s"\u3002',
      baseDisplay,
      normalizedBase
    );
    logger.debug("failed to persist patch target record %s: %O", recordPath, error);
    return void 0;
  }
}

// src/tailwindcss/targets/recorder.ts
import path3 from "path";
import process2 from "process";
function findPatchTargetRecord(baseDir) {
  const visited = /* @__PURE__ */ new Set();
  const fallback = baseDir ?? process2.cwd();
  let current = path3.resolve(fallback);
  while (!visited.has(current)) {
    const record = readPatchTargetRecord(current);
    if (record) {
      return record;
    }
    const parent = path3.dirname(current);
    if (parent === current) {
      break;
    }
    visited.add(current);
    current = parent;
  }
  return void 0;
}
function createPatchTargetRecorder(baseDir, patcher, options) {
  if (!baseDir || !patcher?.packageInfo?.rootPath || options?.recordTarget === false) {
    return void 0;
  }
  const normalizedBase = path3.normalize(baseDir);
  const recorded = findPatchTargetRecord(normalizedBase);
  const location = resolveRecordLocation(normalizedBase);
  const expectedPath = path3.normalize(patcher.packageInfo.rootPath);
  let reason;
  if (!recorded) {
    reason = "missing";
  } else {
    const normalizedRecorded = path3.normalize(recorded.record.tailwindPackagePath);
    if (normalizedRecorded !== expectedPath) {
      reason = "mismatch";
    } else if (path3.normalize(recorded.path) !== path3.normalize(location.recordPath) || !recorded.record.recordKey || recorded.record.recordKey !== location.recordKey) {
      reason = "migrate";
    } else if (!recorded.record.patchVersion || recorded.record.patchVersion !== WEAPP_TW_VERSION) {
      reason = "stale";
    } else if (options?.cwd && recorded.record.cwd && path3.normalize(recorded.record.cwd) !== path3.normalize(options.cwd)) {
      reason = "metadata";
    } else if (!recorded.record.cwd && options?.cwd) {
      reason = "metadata";
    }
  }
  const shouldPersist = options?.alwaysRecord || !recorded || Boolean(reason);
  if (!shouldPersist) {
    return void 0;
  }
  let message;
  switch (reason) {
    case "mismatch":
      message = "\u68C0\u6D4B\u5230 Tailwind CSS \u76EE\u6807\u8BB0\u5F55\u4E0E\u5F53\u524D\u89E3\u6790\u7ED3\u679C\u4E0D\u4E00\u81F4\uFF0C\u6B63\u5728\u81EA\u52A8\u91CD\u65B0 patch \u5E76\u5237\u65B0\u7F13\u5B58\u3002";
      break;
    case "migrate":
    case "stale":
      message = "\u6B63\u5728\u5237\u65B0\u5F53\u524D\u5B50\u5305\u7684 Tailwind CSS \u8865\u4E01\u8BB0\u5F55\uFF0C\u786E\u4FDD\u7F13\u5B58\u9694\u79BB\u3002";
      break;
    case "missing":
      message = "\u672A\u627E\u5230\u5F53\u524D\u5B50\u5305\u7684 Tailwind CSS \u76EE\u6807\u8BB0\u5F55\uFF0C\u6B63\u5728\u751F\u6210\u3002";
      break;
    default:
      break;
  }
  const onPatched = async () => saveCliPatchTargetRecord(normalizedBase, patcher, {
    cwd: options?.cwd ?? normalizedBase,
    source: options?.source ?? "cli",
    recordPath: location.recordPath,
    recordKey: location.recordKey,
    packageJsonPath: location.packageJsonPath
  });
  return {
    recordPath: location.recordPath,
    message,
    reason,
    onPatched
  };
}

// src/tailwindcss/targets.ts
function logTailwindcssTarget(kind, patcher, baseDir) {
  const packageInfo = patcher?.packageInfo;
  const label = kind === "cli" ? "weapp-tw patch" : "tailwindcss-patcher";
  if (!packageInfo?.rootPath) {
    logger2.warn(
      "%s \u672A\u627E\u5230 Tailwind CSS \u4F9D\u8D56\uFF0C\u8BF7\u68C0\u67E5\u5728 %s \u662F\u5426\u5DF2\u5B89\u88C5 tailwindcss",
      label,
      baseDir ?? process3.cwd()
    );
    return;
  }
  const displayPath = formatRelativeToBase(packageInfo.rootPath, baseDir);
  const version = packageInfo.version ? ` (v${packageInfo.version})` : "";
  logger2.info("%s \u7ED1\u5B9A Tailwind CSS -> %s%s", label, displayPath, version);
}

// src/context/compiler-context-cache.ts
import { Buffer } from "buffer";
import path4 from "path";
import process4 from "process";
import { logger as logger3 } from "@weapp-tailwindcss/logger";
var globalCacheHolder = globalThis;
var compilerContextCache = globalCacheHolder.__WEAPP_TW_COMPILER_CONTEXT_CACHE__ ?? (globalCacheHolder.__WEAPP_TW_COMPILER_CONTEXT_CACHE__ = /* @__PURE__ */ new Map());
function compareNormalizedValues(a, b) {
  const aStr = JSON.stringify(a);
  const bStr = JSON.stringify(b);
  return aStr.localeCompare(bStr);
}
function withCircularGuard(value, stack, factory) {
  if (stack.has(value)) {
    throw new TypeError("Cannot serialize circular structure in compiler context options");
  }
  stack.add(value);
  try {
    return factory();
  } finally {
    stack.delete(value);
  }
}
function encodeTaggedValue(type, value) {
  const record = { __type: type };
  if (value !== void 0) {
    record.value = value;
  }
  return record;
}
function detectCallerLocation() {
  const stack = new Error("compiler-context-cache stack probe").stack;
  if (!stack) {
    return void 0;
  }
  const lines = stack.split("\n");
  for (const line of lines) {
    const match = line.match(/\(([^)]+)\)/u) ?? line.match(/at\s+(\S.*)$/u);
    const location = match?.[1];
    if (!location) {
      continue;
    }
    const candidatePath = location.replace(/:\d+(?::\d+)?$/u, "");
    if (!candidatePath || !path4.isAbsolute(candidatePath)) {
      continue;
    }
    if (candidatePath.includes(`${path4.sep}weapp-tailwindcss${path4.sep}src${path4.sep}`) || candidatePath.includes(`${path4.sep}weapp-tailwindcss${path4.sep}dist${path4.sep}`) || candidatePath.includes(`${path4.sep}node_modules${path4.sep}weapp-tailwindcss${path4.sep}`)) {
      continue;
    }
    return candidatePath;
  }
  return void 0;
}
function getRuntimeCacheScope() {
  return {
    cwd: process4.cwd(),
    npm_package_json: process4.env.npm_package_json,
    npm_config_local_prefix: process4.env.npm_config_local_prefix,
    pnpm_package_name: process4.env.PNPM_PACKAGE_NAME,
    init_cwd: process4.env.INIT_CWD,
    pwd: process4.env.PWD,
    weapp_tailwindcss_basedir: process4.env.WEAPP_TAILWINDCSS_BASEDIR,
    weapp_tailwindcss_base_dir: process4.env.WEAPP_TAILWINDCSS_BASE_DIR,
    tailwindcss_basedir: process4.env.TAILWINDCSS_BASEDIR,
    tailwindcss_base_dir: process4.env.TAILWINDCSS_BASE_DIR,
    uni_input_dir: process4.env.UNI_INPUT_DIR,
    uni_input_root: process4.env.UNI_INPUT_ROOT,
    uni_cli_root: process4.env.UNI_CLI_ROOT,
    uni_app_input_dir: process4.env.UNI_APP_INPUT_DIR,
    caller: detectCallerLocation()
  };
}
function normalizeOptionsValue(rawValue, stack = /* @__PURE__ */ new WeakSet()) {
  if (rawValue === null) {
    return null;
  }
  if (rawValue === void 0) {
    return encodeTaggedValue("Undefined");
  }
  const type = typeof rawValue;
  if (type === "string") {
    return rawValue;
  }
  if (type === "boolean") {
    return rawValue;
  }
  if (type === "number") {
    const numericValue = rawValue;
    if (Number.isNaN(numericValue)) {
      return encodeTaggedValue("Number", "NaN");
    }
    if (!Number.isFinite(numericValue)) {
      return encodeTaggedValue("Number", numericValue > 0 ? "Infinity" : "-Infinity");
    }
    if (Object.is(numericValue, -0)) {
      return encodeTaggedValue("Number", "-0");
    }
    return numericValue;
  }
  if (type === "bigint") {
    return encodeTaggedValue("BigInt", rawValue.toString());
  }
  if (type === "symbol") {
    const symbolValue = rawValue;
    return encodeTaggedValue("Symbol", symbolValue.description ?? String(symbolValue));
  }
  if (type === "function") {
    return encodeTaggedValue("Function", rawValue.toString());
  }
  if (Array.isArray(rawValue)) {
    return withCircularGuard(rawValue, stack, () => rawValue.map((item) => normalizeOptionsValue(item, stack)));
  }
  if (rawValue instanceof Date) {
    return encodeTaggedValue("Date", rawValue.toISOString());
  }
  if (rawValue instanceof RegExp) {
    return {
      __type: "RegExp",
      source: rawValue.source,
      flags: rawValue.flags
    };
  }
  if (typeof Buffer !== "undefined" && Buffer.isBuffer(rawValue)) {
    return encodeTaggedValue("Buffer", rawValue.toString("base64"));
  }
  if (ArrayBuffer.isView(rawValue)) {
    const view = rawValue;
    const buffer = Buffer.from(view.buffer, view.byteOffset, view.byteLength);
    return encodeTaggedValue(view.constructor?.name ?? "ArrayBufferView", buffer.toString("base64"));
  }
  if (rawValue instanceof ArrayBuffer) {
    return encodeTaggedValue("ArrayBuffer", Buffer.from(rawValue).toString("base64"));
  }
  if (rawValue instanceof Set) {
    return withCircularGuard(rawValue, stack, () => {
      const normalizedEntries = Array.from(rawValue, (element) => normalizeOptionsValue(element, stack));
      normalizedEntries.sort(compareNormalizedValues);
      return {
        __type: "Set",
        value: normalizedEntries
      };
    });
  }
  if (rawValue instanceof Map) {
    return withCircularGuard(rawValue, stack, () => {
      const normalizedEntries = Array.from(rawValue.entries()).map(([key, entryValue]) => ({
        key: normalizeOptionsValue(key, stack),
        value: normalizeOptionsValue(entryValue, stack)
      }));
      normalizedEntries.sort((a, b) => compareNormalizedValues(a.key, b.key));
      return {
        __type: "Map",
        value: normalizedEntries.map((entry) => [entry.key, entry.value])
      };
    });
  }
  if (typeof URL !== "undefined" && rawValue instanceof URL) {
    return encodeTaggedValue("URL", rawValue.toString());
  }
  if (rawValue instanceof Error) {
    const errorValue = rawValue;
    return {
      __type: "Error",
      name: errorValue.name,
      message: errorValue.message,
      stack: errorValue.stack ?? ""
    };
  }
  if (rawValue instanceof Promise) {
    return encodeTaggedValue("Promise");
  }
  if (rawValue instanceof WeakMap) {
    return encodeTaggedValue("WeakMap");
  }
  if (rawValue instanceof WeakSet) {
    return encodeTaggedValue("WeakSet");
  }
  if (rawValue && typeof rawValue === "object") {
    return withCircularGuard(rawValue, stack, () => {
      const result = {};
      const entries = Object.entries(rawValue);
      entries.sort(([a], [b]) => a.localeCompare(b));
      for (const [key, entryValue] of entries) {
        result[key] = normalizeOptionsValue(entryValue, stack);
      }
      return result;
    });
  }
  return encodeTaggedValue(typeof rawValue, String(rawValue));
}
function createCompilerContextCacheKey(opts) {
  try {
    const normalized = normalizeOptionsValue({
      options: opts ?? {},
      runtime: getRuntimeCacheScope()
    });
    const serialized = JSON.stringify(normalized);
    return md5(serialized);
  } catch (error) {
    logger3.debug("skip compiler context cache: %O", error);
    return void 0;
  }
}
function withCompilerContextCache(opts, factory) {
  const cacheKey = createCompilerContextCacheKey(opts);
  if (cacheKey) {
    const cached = compilerContextCache.get(cacheKey);
    if (cached) {
      return cached;
    }
  }
  const ctx = factory();
  if (cacheKey) {
    compilerContextCache.set(cacheKey, ctx);
  }
  return ctx;
}

// src/context/custom-attributes.ts
function toCustomAttributesEntities(customAttributes) {
  if (!customAttributes) {
    return [];
  }
  if (isMap(customAttributes)) {
    return [
      ...customAttributes.entries()
    ];
  }
  return Object.entries(customAttributes);
}

// src/context/handlers.ts
import { createStyleHandler } from "@weapp-tailwindcss/postcss";

// src/babel/index.ts
import _babelTraverse from "@babel/traverse";
import { parse, parseExpression } from "@babel/parser";
function _interopDefaultCompat(e) {
  return e && typeof e === "object" && "default" in e ? e.default : e;
}
var traverse = _interopDefaultCompat(_babelTraverse);

// src/utils/nameMatcher.ts
import { escapeStringRegexp } from "@weapp-core/regex";
function buildFuzzyMatcher(fuzzyStrings) {
  if (fuzzyStrings.length === 0) {
    return void 0;
  }
  if (fuzzyStrings.length === 1) {
    const [needle] = fuzzyStrings;
    return (value) => value.includes(needle);
  }
  const unique = [...new Set(fuzzyStrings)];
  const pattern = new RegExp(unique.map(escapeStringRegexp).join("|"));
  return (value) => pattern.test(value);
}
function normaliseRegex(regex) {
  const { source, flags } = regex;
  if (!flags.includes("g")) {
    return regex;
  }
  return new RegExp(source, flags.replace(/g/g, ""));
}
function createNameMatcher(list, { exact = false } = {}) {
  if (!list || list.length === 0) {
    return () => false;
  }
  const exactStrings = exact ? /* @__PURE__ */ new Set() : void 0;
  const fuzzyStrings = [];
  const regexList = [];
  for (const item of list) {
    if (typeof item === "string") {
      if (exact) {
        exactStrings.add(item);
      } else {
        fuzzyStrings.push(item);
      }
    } else {
      regexList.push(normaliseRegex(item));
    }
  }
  const fuzzyMatcher = exact ? void 0 : buildFuzzyMatcher(fuzzyStrings);
  const hasRegex = regexList.length > 0;
  return (value) => {
    if (exact && exactStrings?.has(value)) {
      return true;
    }
    if (fuzzyMatcher?.(value)) {
      return true;
    }
    if (!hasRegex) {
      return false;
    }
    return regexList.some((regex) => regex.test(value));
  };
}

// src/js/babel/parse.ts
import { LRUCache as LRUCache2 } from "lru-cache";
var parseCache = new LRUCache2(
  {
    max: 1024
  }
);
function genCacheKey(source, options) {
  if (typeof options === "string") {
    return source + options;
  }
  return source + JSON.stringify(options, (_, val) => typeof val === "function" ? val.toString() : val);
}
function babelParse(code, opts = {}) {
  const { cache, cacheKey, ...rest } = opts;
  const cacheKeyString = genCacheKey(code, cacheKey ?? rest);
  let result;
  if (cache) {
    result = parseCache.get(cacheKeyString);
  }
  if (!result) {
    const { cache: _cache, cacheKey: _cacheKey, ...parseOptions } = opts;
    result = parse(code, parseOptions);
    if (cache) {
      parseCache.set(cacheKeyString, result);
    }
  }
  return result;
}

// src/js/babel/process.ts
import MagicString from "magic-string";

// src/js/handlers.ts
import { jsStringEscape } from "@ast-core/escape";
import { escapeStringRegexp as escapeStringRegexp2 } from "@weapp-core/regex";
import { splitCode } from "@weapp-tailwindcss/shared/extractors";

// src/wxml/shared.ts
import { escape, MappingChars2String } from "@weapp-core/escape";
function replaceWxml(original, options = {
  keepEOL: false,
  escapeMap: MappingChars2String
}) {
  const { keepEOL, escapeMap, ignoreHead } = options;
  let res = original;
  if (!keepEOL) {
    res = res.replaceAll(/[\n\r]+/g, "");
  }
  res = escape(res, {
    map: escapeMap,
    ignoreHead
  });
  return res;
}

// src/shared/classname-transform.ts
function isArbitraryValueCandidate(candidate) {
  const normalized = candidate.trim();
  if (!normalized.includes("[") || !normalized.includes("]")) {
    return false;
  }
  if (/^(?:https?:)?\/\//.test(normalized)) {
    return false;
  }
  return true;
}
function shouldEnableArbitraryValueFallback({
  classNameSet,
  jsArbitraryValueFallback,
  tailwindcssMajorVersion
}) {
  if (jsArbitraryValueFallback === true) {
    return true;
  }
  if (jsArbitraryValueFallback === false) {
    return false;
  }
  return tailwindcssMajorVersion === 4 && (!classNameSet || classNameSet.size === 0);
}
function resolveClassNameTransformDecision(candidate, {
  alwaysEscape,
  classNameSet,
  escapeMap,
  jsArbitraryValueFallback,
  jsPreserveClass,
  tailwindcssMajorVersion,
  classContext
}) {
  if (alwaysEscape) {
    return "direct";
  }
  if (jsPreserveClass?.(candidate)) {
    return "skip";
  }
  if (classNameSet?.has(candidate)) {
    return "direct";
  }
  if (classNameSet && classNameSet.size > 0) {
    const escapedCandidate = replaceWxml(candidate, { escapeMap });
    if (escapedCandidate !== candidate && classNameSet.has(escapedCandidate)) {
      return "escaped";
    }
  }
  if (classContext && isArbitraryValueCandidate(candidate) && shouldEnableArbitraryValueFallback({
    classNameSet,
    jsArbitraryValueFallback,
    tailwindcssMajorVersion
  })) {
    return "fallback";
  }
  return "skip";
}

// src/utils/decode.ts
var unicodeEscapeRE = /\\u([\dA-Fa-f]{4})/g;
var unicodeEscapeTestRE = /\\u[\dA-Fa-f]{4}/;
function decodeUnicode(value) {
  if (!unicodeEscapeTestRE.test(value)) {
    return value;
  }
  return value.replace(unicodeEscapeRE, (_match, hex) => {
    const codePoint = Number.parseInt(hex, 16);
    return Number.isNaN(codePoint) ? _match : String.fromCharCode(codePoint);
  });
}
function decodeUnicode2(input) {
  if (!unicodeEscapeTestRE.test(input)) {
    return input;
  }
  try {
    return JSON.parse(`"${input}"`);
  } catch (_error) {
    return decodeUnicode(input);
  }
}

// src/js/class-context.ts
var CLASS_LIKE_KEYWORDS = /* @__PURE__ */ new Set([
  "class",
  "classname",
  "hoverclass",
  "virtualhostclass",
  "rootclass"
]);
var CLASS_HELPER_IDENTIFIERS = /* @__PURE__ */ new Set([
  "cn",
  "clsx",
  "classnames",
  "twmerge",
  "cva",
  "tv",
  "cx",
  "r"
]);
function normalizeKeyword(name) {
  return name.replace(/[-_:]/g, "").toLowerCase();
}
function readObjectKeyName(path5) {
  if (path5.isIdentifier()) {
    return path5.node.name;
  }
  if (path5.isStringLiteral()) {
    return path5.node.value;
  }
  if (path5.isTemplateLiteral() && path5.node.expressions.length === 0) {
    return path5.node.quasis[0]?.value.cooked ?? path5.node.quasis[0]?.value.raw;
  }
  return void 0;
}
function isClassLikeObjectProperty(path5, valuePath) {
  if (!path5.isObjectProperty()) {
    return false;
  }
  if (path5.get("value") !== valuePath) {
    return false;
  }
  const keyName = readObjectKeyName(path5.get("key"));
  if (!keyName) {
    return false;
  }
  return CLASS_LIKE_KEYWORDS.has(normalizeKeyword(keyName));
}
function isClassLikeJsxAttribute(path5) {
  if (!path5.isJSXAttribute()) {
    return false;
  }
  const namePath = path5.get("name");
  if (!namePath.isJSXIdentifier()) {
    return false;
  }
  return CLASS_LIKE_KEYWORDS.has(normalizeKeyword(namePath.node.name));
}
function isClassLikeCallExpression(path5, valuePath) {
  if (!path5.isCallExpression()) {
    return false;
  }
  const args = path5.get("arguments");
  if (!args.includes(valuePath)) {
    return false;
  }
  const calleePath = path5.get("callee");
  if (calleePath.isIdentifier()) {
    return CLASS_HELPER_IDENTIFIERS.has(normalizeKeyword(calleePath.node.name));
  }
  if (calleePath.isMemberExpression()) {
    const propertyPath = calleePath.get("property");
    if (propertyPath.isIdentifier()) {
      return CLASS_HELPER_IDENTIFIERS.has(normalizeKeyword(propertyPath.node.name));
    }
    if (propertyPath.isStringLiteral()) {
      return CLASS_HELPER_IDENTIFIERS.has(normalizeKeyword(propertyPath.node.value));
    }
  }
  return false;
}
function isClassContextLiteralPath(path5) {
  let current = path5;
  while (current.parentPath) {
    const parent = current.parentPath;
    if (isClassLikeObjectProperty(parent, current)) {
      return true;
    }
    if (isClassLikeJsxAttribute(parent)) {
      return true;
    }
    if (isClassLikeCallExpression(parent, current)) {
      return true;
    }
    current = parent;
  }
  return false;
}

// src/js/handlers.ts
var debug2 = createDebug("[js:handlers] ");
var patternCache = /* @__PURE__ */ new Map();
var replacementCacheByEscapeMap = /* @__PURE__ */ new WeakMap();
var defaultReplacementCache = /* @__PURE__ */ new Map();
function getPattern(candidate) {
  let cached = patternCache.get(candidate);
  if (!cached) {
    cached = new RegExp(escapeStringRegexp2(candidate));
    patternCache.set(candidate, cached);
  }
  return cached;
}
function getReplacement(candidate, escapeMap) {
  if (!escapeMap) {
    let cached2 = defaultReplacementCache.get(candidate);
    if (cached2 === void 0) {
      cached2 = replaceWxml(candidate, { escapeMap });
      defaultReplacementCache.set(candidate, cached2);
    }
    return cached2;
  }
  let store = replacementCacheByEscapeMap.get(escapeMap);
  if (!store) {
    store = /* @__PURE__ */ new Map();
    replacementCacheByEscapeMap.set(escapeMap, store);
  }
  let cached = store.get(candidate);
  if (cached === void 0) {
    cached = replaceWxml(candidate, { escapeMap });
    store.set(candidate, cached);
  }
  return cached;
}
function hasIgnoreComment(node) {
  return Array.isArray(node.leadingComments) && node.leadingComments.some((comment) => comment.value.includes("weapp-tw") && comment.value.includes("ignore"));
}
function extractLiteralValue(path5, { unescapeUnicode, arbitraryValues }) {
  const allowDoubleQuotes = arbitraryValues?.allowDoubleQuotes;
  let offset = 0;
  let original;
  if (path5.isStringLiteral()) {
    offset = 1;
    original = path5.node.value;
  } else if (path5.isTemplateElement()) {
    original = path5.node.value.raw;
  } else {
    original = "";
  }
  let literal = original;
  if (unescapeUnicode && original.includes("\\u")) {
    literal = decodeUnicode2(original);
  }
  return {
    allowDoubleQuotes,
    literal,
    offset,
    original
  };
}
function replaceHandleValue(path5, options) {
  const {
    escapeMap,
    needEscaped = false
  } = options;
  const { classNameSet, alwaysEscape } = options;
  const fallbackEnabled = shouldEnableArbitraryValueFallback(options);
  const classContext = options.wrapExpression || isClassContextLiteralPath(path5);
  if (!alwaysEscape && !fallbackEnabled && (!classNameSet || classNameSet.size === 0)) {
    return void 0;
  }
  const { literal, original, allowDoubleQuotes, offset } = extractLiteralValue(path5, options);
  if (hasIgnoreComment(path5.node)) {
    return void 0;
  }
  const candidates = splitCode(literal, allowDoubleQuotes);
  if (candidates.length === 0) {
    return void 0;
  }
  let transformed = literal;
  let mutated = false;
  let matchedCandidateCount = 0;
  let escapedDecisionCount = 0;
  let fallbackDecisionCount = 0;
  const escapedSamples = [];
  const skippedSamples = [];
  for (const candidate of candidates) {
    const decision = resolveClassNameTransformDecision(candidate, {
      ...options,
      classContext
    });
    if (decision === "skip") {
      if (skippedSamples.length < 6) {
        skippedSamples.push(candidate);
      }
      continue;
    }
    matchedCandidateCount += 1;
    if (decision === "escaped") {
      escapedDecisionCount += 1;
      if (escapedSamples.length < 6) {
        escapedSamples.push(candidate);
      }
    }
    if (decision === "fallback") {
      fallbackDecisionCount += 1;
    }
    if (!transformed.includes(candidate)) {
      continue;
    }
    const pattern = getPattern(candidate);
    const replacement = getReplacement(candidate, escapeMap);
    const replaced = transformed.replace(pattern, replacement);
    if (replaced !== transformed) {
      transformed = replaced;
      mutated = true;
    }
  }
  const node = path5.node;
  if (!mutated || typeof node.start !== "number" || typeof node.end !== "number") {
    return void 0;
  }
  debug2(
    "runtimeSet size=%d fallbackTriggered=%s candidates=%d matched=%d escapedHits=%d skipped=%d file=%s escapedSamples=%s skippedSamples=%s",
    classNameSet?.size ?? 0,
    fallbackDecisionCount > 0,
    candidates.length,
    matchedCandidateCount,
    escapedDecisionCount,
    skippedSamples.length,
    options.filename ?? "unknown",
    escapedSamples.join(",") || "-",
    skippedSamples.join(",") || "-"
  );
  const start = node.start + offset;
  const end = node.end - offset;
  if (start >= end || transformed === original) {
    return void 0;
  }
  const value = needEscaped ? jsStringEscape(transformed) : transformed;
  return {
    start,
    end,
    value,
    path: path5
  };
}

// src/js/sourceAnalysis.ts
function createModuleSpecifierReplacementToken(path5, replacement) {
  const node = path5.node;
  if (node.value === replacement) {
    return void 0;
  }
  if (typeof node.start !== "number" || typeof node.end !== "number") {
    return void 0;
  }
  const start = node.start + 1;
  const end = node.end - 1;
  if (start >= end) {
    return void 0;
  }
  return {
    start,
    end,
    value: replacement,
    path: path5
  };
}
function collectModuleSpecifierReplacementTokens(analysis, replacements) {
  const tokens = [];
  const applyReplacement = (path5) => {
    const replacement = replacements[path5.node.value];
    if (!replacement) {
      return;
    }
    const token = createModuleSpecifierReplacementToken(path5, replacement);
    if (token) {
      tokens.push(token);
    }
  };
  for (const importPath of analysis.importDeclarations) {
    const source = importPath.get("source");
    if (source.isStringLiteral()) {
      applyReplacement(source);
    }
  }
  for (const exportPath of analysis.exportDeclarations) {
    if (exportPath.isExportNamedDeclaration() || exportPath.isExportAllDeclaration()) {
      const source = exportPath.get("source");
      if (source && !Array.isArray(source) && source.isStringLiteral()) {
        applyReplacement(source);
      }
    }
  }
  for (const literalPath of analysis.requireCallPaths) {
    applyReplacement(literalPath);
  }
  for (const token of analysis.walker.imports) {
    const replacement = replacements[token.source];
    if (replacement) {
      token.source = replacement;
    }
  }
  return tokens;
}

// src/js/babel/process.ts
function processUpdatedSource(rawSource, options, analysis) {
  const { targetPaths, jsTokenUpdater, ignoredPaths } = analysis;
  const replacementTokens = [];
  for (const path5 of targetPaths) {
    if (ignoredPaths.has(path5)) {
      continue;
    }
    const token = replaceHandleValue(
      path5,
      {
        ...options,
        needEscaped: path5.isStringLiteral() ? options.needEscaped ?? true : false
      }
    );
    if (token) {
      replacementTokens.push(token);
    }
  }
  if (options.moduleSpecifierReplacements) {
    replacementTokens.push(
      ...collectModuleSpecifierReplacementTokens(analysis, options.moduleSpecifierReplacements)
    );
  }
  if (jsTokenUpdater.length + replacementTokens.length === 0) {
    return new MagicString(rawSource);
  }
  const ms = new MagicString(rawSource);
  jsTokenUpdater.push(...replacementTokens).filter((token) => !ignoredPaths.has(token.path)).updateMagicString(ms);
  return ms;
}

// src/js/evalTransforms.ts
import { jsStringEscape as jsStringEscape2 } from "@ast-core/escape";
function isEvalPath(path5) {
  if (path5.isCallExpression()) {
    const calleePath = path5.get("callee");
    return calleePath.isIdentifier({ name: "eval" });
  }
  return false;
}
function createEvalReplacementToken(path5, updated) {
  const node = path5.node;
  let offset = 0;
  let original;
  if (path5.isStringLiteral()) {
    offset = 1;
    original = path5.node.value;
  } else if (path5.isTemplateElement()) {
    original = path5.node.value.raw;
  } else {
    original = "";
  }
  if (typeof node.start !== "number" || typeof node.end !== "number") {
    return void 0;
  }
  const start = node.start + offset;
  const end = node.end - offset;
  if (start >= end) {
    return void 0;
  }
  if (original === updated) {
    return void 0;
  }
  const value = path5.isStringLiteral() ? jsStringEscape2(updated) : updated;
  return {
    start,
    end,
    value,
    path: path5
  };
}
function handleEvalStringLiteral(path5, options, updater, handler) {
  const { code } = handler(path5.node.value, {
    ...options,
    needEscaped: false,
    generateMap: false
  });
  if (!code) {
    return;
  }
  const token = createEvalReplacementToken(path5, code);
  if (token) {
    updater.addToken(token);
  }
}
function handleEvalTemplateElement(path5, options, updater, handler) {
  const { code } = handler(path5.node.value.raw, {
    ...options,
    generateMap: false
  });
  if (!code) {
    return;
  }
  const token = createEvalReplacementToken(path5, code);
  if (token) {
    updater.addToken(token);
  }
}
function walkEvalExpression(path5, options, updater, handler) {
  const maybeTraverse = path5?.traverse;
  if (typeof maybeTraverse === "function") {
    try {
      return maybeTraverse.call(path5, {
        StringLiteral(innerPath) {
          handleEvalStringLiteral(innerPath, options, updater, handler);
        },
        TemplateElement(innerPath) {
          handleEvalTemplateElement(innerPath, options, updater, handler);
        }
      });
    } catch (error) {
      const msg = error?.message ?? "";
      const scopeError = /pass a scope and parentPath|traversing a Program\/File/i.test(msg);
      if (!scopeError) {
        throw error;
      }
    }
  }
  const getArgs = path5?.get?.("arguments");
  if (Array.isArray(getArgs)) {
    for (const arg of getArgs) {
      if (arg?.isStringLiteral?.()) {
        handleEvalStringLiteral(arg, options, updater, handler);
        continue;
      }
      if (arg?.isTemplateLiteral?.()) {
        for (const quasi of arg.get("quasis")) {
          handleEvalTemplateElement(quasi, options, updater, handler);
        }
      }
    }
    return;
  }
  const nodeArgs = path5?.node?.arguments;
  if (Array.isArray(nodeArgs)) {
    for (const n of nodeArgs) {
      if (n?.type === "StringLiteral") {
        const stub = {
          node: n,
          isStringLiteral: () => true
        };
        handleEvalStringLiteral(stub, options, updater, handler);
      } else if (n?.type === "TemplateLiteral" && Array.isArray(n.quasis)) {
        for (const q of n.quasis) {
          const stub = {
            node: q,
            isTemplateElement: () => true
          };
          handleEvalTemplateElement(stub, options, updater, handler);
        }
      }
    }
  }
}

// src/js/JsTokenUpdater.ts
var JsTokenUpdater = class {
  constructor({ value } = {}) {
    this.tokens = value ? [...value] : [];
  }
  addToken(token) {
    if (token) {
      this.tokens.push(token);
    }
  }
  push(...args) {
    this.tokens.push(...args);
    return this;
  }
  /**
   * 待写入的 token 数量。
   */
  get length() {
    return this.tokens.length;
  }
  map(callbackfn) {
    this.tokens = this.tokens.map(callbackfn);
    return this;
  }
  filter(callbackfn) {
    this.tokens = this.tokens.filter(callbackfn);
    return this;
  }
  updateMagicString(ms) {
    for (const { start, end, value } of this.tokens) {
      ms.update(start, end, value);
    }
    return ms;
  }
};

// src/js/module-graph/ignored-exports.ts
var IgnoredExportsTracker = class {
  constructor(options) {
    this.options = options;
    this.ignoredExportNames = /* @__PURE__ */ new Map();
  }
  addIgnoredExport(filename, exportName) {
    if (!exportName) {
      return;
    }
    let pending = this.ignoredExportNames.get(filename);
    if (!pending) {
      pending = /* @__PURE__ */ new Set();
      this.ignoredExportNames.set(filename, pending);
    }
    if (pending.has(exportName)) {
      return;
    }
    pending.add(exportName);
    const existing = this.options.modules.get(filename);
    if (existing) {
      this.applyIgnoredExportsToAnalysis(filename, existing.analysis);
    }
  }
  registerIgnoredExportsFromTokens(resolved, tokens) {
    for (const token of tokens) {
      if (token.type === "ImportSpecifier") {
        this.addIgnoredExport(resolved, token.imported);
      } else if (token.type === "ImportDefaultSpecifier") {
        this.addIgnoredExport(resolved, "default");
      }
    }
  }
  applyIgnoredExportsToAnalysis(filename, analysis) {
    const pending = this.ignoredExportNames.get(filename);
    if (!pending || pending.size === 0) {
      return;
    }
    const names = new Set(pending);
    pending.clear();
    const propagate = [];
    for (const exportPath of analysis.exportDeclarations) {
      if (names.size === 0) {
        break;
      }
      if (exportPath.isExportDefaultDeclaration()) {
        if (names.has("default")) {
          analysis.walker.walkExportDefaultDeclaration(exportPath);
          names.delete("default");
        }
        continue;
      }
      if (exportPath.isExportNamedDeclaration()) {
        const source = exportPath.node.source?.value;
        if (typeof source === "string") {
          for (const spec of exportPath.get("specifiers")) {
            if (!spec.isExportSpecifier()) {
              continue;
            }
            const exported = spec.get("exported");
            let exportedName;
            if (exported.isIdentifier()) {
              exportedName = exported.node.name;
            } else if (exported.isStringLiteral()) {
              exportedName = exported.node.value;
            }
            if (!exportedName || !names.has(exportedName)) {
              continue;
            }
            const local = spec.get("local");
            if (local.isIdentifier()) {
              propagate.push({
                specifier: source,
                exportName: local.node.name
              });
              names.delete(exportedName);
            } else if (local.isStringLiteral()) {
              propagate.push({
                specifier: source,
                exportName: local.node.value
              });
              names.delete(exportedName);
            }
          }
          continue;
        }
        const declaration = exportPath.get("declaration");
        if (declaration.isVariableDeclaration()) {
          for (const decl of declaration.get("declarations")) {
            const id = decl.get("id");
            if (id.isIdentifier()) {
              const exportName = id.node.name;
              if (names.has(exportName)) {
                analysis.walker.walkVariableDeclarator(decl);
                names.delete(exportName);
              }
            }
          }
        }
        for (const spec of exportPath.get("specifiers")) {
          if (!spec.isExportSpecifier()) {
            continue;
          }
          const exported = spec.get("exported");
          let exportedName;
          if (exported.isIdentifier()) {
            exportedName = exported.node.name;
          } else if (exported.isStringLiteral()) {
            exportedName = exported.node.value;
          }
          if (!exportedName || !names.has(exportedName)) {
            continue;
          }
          const local = spec.get("local");
          analysis.walker.walkNode(local);
          names.delete(exportedName);
        }
        continue;
      }
      if (exportPath.isExportAllDeclaration()) {
        const source = exportPath.node.source?.value;
        if (typeof source === "string") {
          for (const exportName of names) {
            propagate.push({
              specifier: source,
              exportName
            });
          }
          names.clear();
        }
      }
    }
    for (const { specifier, exportName } of propagate) {
      let resolved;
      try {
        resolved = this.options.resolve(specifier, filename);
      } catch {
        resolved = void 0;
      }
      if (!resolved) {
        pending.add(exportName);
        continue;
      }
      if (this.options.filter && !this.options.filter(resolved, specifier, filename)) {
        pending.add(exportName);
        continue;
      }
      this.addIgnoredExport(resolved, exportName);
    }
    for (const name of names) {
      pending.add(name);
    }
  }
};

// src/js/ModuleGraph.ts
var JsModuleGraph = class {
  constructor(entry, graphOptions) {
    this.modules = /* @__PURE__ */ new Map();
    this.queue = [];
    this.resolve = graphOptions.resolve;
    this.load = graphOptions.load;
    this.filter = graphOptions.filter;
    this.maxDepth = graphOptions.maxDepth ?? Number.POSITIVE_INFINITY;
    const { moduleGraph: _moduleGraph, filename: _ignoredFilename, ...rest } = entry.handlerOptions;
    this.baseOptions = {
      ...rest,
      filename: entry.filename
    };
    this.parserOptions = entry.handlerOptions.babelParserOptions;
    this.rootFilename = entry.filename;
    this.ignoredExports = new IgnoredExportsTracker({
      resolve: this.resolve,
      filter: this.filter,
      modules: this.modules
    });
    this.modules.set(entry.filename, {
      filename: entry.filename,
      source: entry.source,
      analysis: entry.analysis
    });
    this.queue.push({ filename: entry.filename, depth: 0 });
  }
  build() {
    this.collectDependencies();
    const linked = {};
    for (const [filename, state] of this.modules) {
      if (filename === this.rootFilename) {
        continue;
      }
      const childOptions = {
        ...this.baseOptions,
        filename
      };
      const ms = processUpdatedSource(state.source, childOptions, state.analysis);
      const code = ms.toString();
      if (code !== state.source) {
        linked[filename] = { code };
      }
    }
    return linked;
  }
  collectDependencies() {
    while (this.queue.length > 0) {
      const { filename, depth } = this.queue.shift();
      if (depth >= this.maxDepth) {
        continue;
      }
      const state = this.modules.get(filename);
      if (!state) {
        continue;
      }
      const dependencySpecifiers = /* @__PURE__ */ new Map();
      for (const token of state.analysis.walker.imports) {
        if (!dependencySpecifiers.has(token.source)) {
          dependencySpecifiers.set(token.source, []);
        }
        dependencySpecifiers.get(token.source).push(token);
      }
      for (const exportPath of state.analysis.exportDeclarations) {
        if (exportPath.isExportAllDeclaration() || exportPath.isExportNamedDeclaration()) {
          const source = exportPath.node.source?.value;
          if (typeof source === "string" && !dependencySpecifiers.has(source)) {
            dependencySpecifiers.set(source, []);
          }
        }
      }
      for (const [specifier, tokens] of dependencySpecifiers) {
        let resolved;
        try {
          resolved = this.resolve(specifier, filename);
        } catch {
          continue;
        }
        if (!resolved) {
          continue;
        }
        if (this.filter && !this.filter(resolved, specifier, filename)) {
          continue;
        }
        if (tokens.length > 0) {
          this.ignoredExports.registerIgnoredExportsFromTokens(resolved, tokens);
        }
        if (this.modules.has(resolved)) {
          continue;
        }
        let source;
        try {
          source = this.load(resolved);
        } catch {
          continue;
        }
        if (typeof source !== "string") {
          continue;
        }
        let analysis;
        try {
          const ast = babelParse(source, {
            ...this.parserOptions,
            sourceFilename: resolved
          });
          analysis = analyzeSource(ast, {
            ...this.baseOptions,
            filename: resolved
          });
          this.ignoredExports.applyIgnoredExportsToAnalysis(resolved, analysis);
        } catch {
          continue;
        }
        this.modules.set(resolved, {
          filename: resolved,
          source,
          analysis
        });
        this.queue.push({ filename: resolved, depth: depth + 1 });
      }
    }
  }
};

// src/js/node-path-walker/export-handlers.ts
function walkExportDeclaration(ctx, path5) {
  if (path5.isExportDeclaration()) {
    if (path5.isExportNamedDeclaration()) {
      walkExportNamedDeclaration(ctx, path5);
    } else if (path5.isExportDefaultDeclaration()) {
      walkExportDefaultDeclaration(ctx, path5);
    } else if (path5.isExportAllDeclaration()) {
      walkExportAllDeclaration(ctx, path5);
    }
  }
}
function walkExportNamedDeclaration(ctx, path5) {
  const declaration = path5.get("declaration");
  if (declaration.isVariableDeclaration()) {
    for (const decl of declaration.get("declarations")) {
      ctx.walkNode(decl);
    }
  }
  const specifiers = path5.get("specifiers");
  for (const spec of specifiers) {
    if (spec.isExportSpecifier()) {
      const local = spec.get("local");
      if (local.isIdentifier()) {
        ctx.walkNode(local);
      }
    }
  }
}
function walkExportDefaultDeclaration(ctx, path5) {
  const decl = path5.get("declaration");
  if (decl.isIdentifier()) {
    ctx.walkNode(decl);
  } else {
    ctx.walkNode(decl);
  }
}
function walkExportAllDeclaration(ctx, path5) {
  const source = path5.get("source");
  if (source.isStringLiteral()) {
    ctx.imports.add(
      {
        declaration: path5,
        source: source.node.value,
        type: "ExportAllDeclaration"
      }
    );
  }
}

// src/js/node-path-walker/import-tokens.ts
function maybeAddImportToken(imports, arg) {
  if (!(arg.isImportSpecifier() && arg.node.importKind !== "type" || arg.isImportDefaultSpecifier())) {
    return false;
  }
  const importDeclaration = arg.parentPath;
  if (!importDeclaration.isImportDeclaration() || importDeclaration.node.importKind === "type") {
    return false;
  }
  if (arg.isImportSpecifier()) {
    const imported = arg.get("imported");
    if (imported.isIdentifier()) {
      imports.add(
        {
          declaration: importDeclaration,
          specifier: arg,
          imported: imported.node.name,
          local: arg.node.local.name,
          source: importDeclaration.node.source.value,
          type: "ImportSpecifier"
        }
      );
    }
    return true;
  }
  imports.add(
    {
      declaration: importDeclaration,
      specifier: arg,
      local: arg.node.local.name,
      source: importDeclaration.node.source.value,
      type: "ImportDefaultSpecifier"
    }
  );
  return true;
}

// src/js/NodePathWalker.ts
var NodePathWalker = class {
  constructor({ ignoreCallExpressionIdentifiers, callback } = {}) {
    this.ignoreCallExpressionIdentifiers = ignoreCallExpressionIdentifiers ?? [];
    this.callback = callback ?? (() => {
    });
    this.imports = /* @__PURE__ */ new Set();
    this.visited = /* @__PURE__ */ new WeakSet();
    this.isIgnoredCallIdentifier = createNameMatcher(this.ignoreCallExpressionIdentifiers, { exact: true });
  }
  walkVariableDeclarator(path5) {
    const init = path5.get("init");
    this.walkNode(init);
  }
  walkTemplateLiteral(path5) {
    for (const exp of path5.get("expressions")) {
      this.walkNode(exp);
    }
    for (const quasis of path5.get("quasis")) {
      this.callback(quasis);
    }
  }
  walkStringLiteral(path5) {
    this.callback(path5);
  }
  walkBinaryExpression(path5) {
    const left = path5.get("left");
    this.walkNode(left);
    const right = path5.get("right");
    this.walkNode(right);
  }
  walkLogicalExpression(path5) {
    const left = path5.get("left");
    this.walkNode(left);
    const right = path5.get("right");
    this.walkNode(right);
  }
  walkObjectExpression(path5) {
    const props = path5.get("properties");
    for (const prop of props) {
      if (prop.isObjectProperty()) {
        const key = prop.get("key");
        this.walkNode(key);
        const value = prop.get("value");
        this.walkNode(value);
      }
    }
  }
  walkArrayExpression(path5) {
    const elements = path5.get("elements");
    for (const element of elements) {
      this.walkNode(element);
    }
  }
  walkNode(arg) {
    if (this.visited.has(arg)) {
      return;
    }
    this.visited.add(arg);
    if (arg.isIdentifier()) {
      const binding = arg?.scope?.getBinding?.(arg.node.name);
      if (binding) {
        this.walkNode(binding.path);
      }
    } else if (arg.isMemberExpression()) {
      const objectPath = arg.get("object");
      if (objectPath.isIdentifier()) {
        const binding = arg?.scope?.getBinding?.(objectPath.node.name);
        if (binding) {
          if (binding.path.isVariableDeclarator()) {
            this.walkVariableDeclarator(binding.path);
          }
        }
      }
    } else if (arg.isTemplateLiteral()) {
      this.walkTemplateLiteral(arg);
    } else if (arg.isStringLiteral()) {
      this.walkStringLiteral(arg);
    } else if (arg.isBinaryExpression()) {
      this.walkBinaryExpression(arg);
    } else if (arg.isLogicalExpression()) {
      this.walkLogicalExpression(arg);
    } else if (arg.isObjectExpression()) {
      this.walkObjectExpression(arg);
    } else if (arg.isArrayExpression()) {
      this.walkArrayExpression(arg);
    } else if (arg.isVariableDeclarator()) {
      this.walkVariableDeclarator(arg);
    } else if (maybeAddImportToken(this.imports, arg)) {
    }
  }
  /**
   * Walk the arguments of a desired call expression so their bindings can be analysed.
   */
  walkCallExpression(path5) {
    const calleePath = path5.get("callee");
    if (calleePath.isIdentifier() && this.isIgnoredCallIdentifier(calleePath.node.name)) {
      for (const arg of path5.get("arguments")) {
        this.walkNode(arg);
      }
    }
  }
  walkExportDeclaration(path5) {
    walkExportDeclaration(this, path5);
  }
  walkExportNamedDeclaration(path5) {
    walkExportNamedDeclaration(this, path5);
  }
  walkExportDefaultDeclaration(path5) {
    walkExportDefaultDeclaration(this, path5);
  }
  walkExportAllDeclaration(path5) {
    walkExportAllDeclaration(this, path5);
  }
};

// src/js/taggedTemplateIgnore.ts
function createTaggedTemplateIgnore({ matcher, names }) {
  const bindingIgnoreCache = /* @__PURE__ */ new Map();
  const taggedTemplateIgnoreCache = /* @__PURE__ */ new WeakMap();
  const canonicalIgnoreNames = new Set(
    (names ?? []).filter((item) => typeof item === "string")
  );
  const propertyMatches = (propertyPath) => {
    if (!propertyPath) {
      return false;
    }
    if (propertyPath.isIdentifier()) {
      const { name } = propertyPath.node;
      return canonicalIgnoreNames.has(name) || matcher(name);
    }
    if (propertyPath.isStringLiteral()) {
      const { value } = propertyPath.node;
      return canonicalIgnoreNames.has(value) || matcher(value);
    }
    return false;
  };
  const resolvesMemberExpressionToIgnore = (path5, seen) => {
    const propertyPath = path5.get("property");
    if (propertyMatches(propertyPath)) {
      return true;
    }
    const objectPath = path5.get("object");
    if (objectPath.isIdentifier()) {
      const scope = objectPath?.scope;
      const binding = scope?.getBinding?.(objectPath.node.name);
      if (binding) {
        return resolvesToWeappTwIgnore(binding, seen);
      }
    }
    return false;
  };
  const resolvesToWeappTwIgnore = (binding, seen) => {
    const cached = bindingIgnoreCache.get(binding);
    if (cached !== void 0) {
      return cached;
    }
    if (seen.has(binding)) {
      return false;
    }
    seen.add(binding);
    let result = false;
    const bindingPath = binding.path;
    if (bindingPath.isImportSpecifier()) {
      const imported = bindingPath.node.imported;
      if (imported.type === "Identifier" && (canonicalIgnoreNames.has(imported.name) || matcher(imported.name))) {
        result = true;
      } else if (imported.type === "StringLiteral" && (canonicalIgnoreNames.has(imported.value) || matcher(imported.value))) {
        result = true;
      }
    } else if (bindingPath.isVariableDeclarator()) {
      const init = bindingPath.get("init");
      if (init && init.node) {
        if (init.isIdentifier()) {
          const target = binding?.scope?.getBinding?.(init.node.name);
          if (target) {
            result = resolvesToWeappTwIgnore(target, seen);
          }
        } else if (init.isMemberExpression()) {
          result = resolvesMemberExpressionToIgnore(init, seen);
        }
      }
    }
    bindingIgnoreCache.set(binding, result);
    seen.delete(binding);
    return result;
  };
  const getEffectiveTagPath = (tagPath) => {
    let current = tagPath;
    while (true) {
      if (current.isParenthesizedExpression?.() || current.node.type === "ParenthesizedExpression") {
        current = current.get("expression");
        continue;
      }
      if (current.isTSAsExpression() || current.isTSTypeAssertion()) {
        current = current.get("expression");
        continue;
      }
      if (current.isTSNonNullExpression()) {
        current = current.get("expression");
        continue;
      }
      if (current.isTypeCastExpression?.()) {
        current = current.get("expression");
        continue;
      }
      if (current.isSequenceExpression()) {
        const expressions = current.get("expressions");
        const last = expressions[expressions.length - 1];
        if (last) {
          current = last;
          continue;
        }
      }
      if (current.isCallExpression?.() || current.node.type === "CallExpression") {
        const callee = current.get("callee");
        current = callee;
        continue;
      }
      break;
    }
    return current;
  };
  const evaluateTagPath = (tagPath) => {
    if (tagPath.isCallExpression?.() || tagPath.node.type === "CallExpression") {
      const calleePath = tagPath.get("callee");
      return evaluateTagPath(calleePath);
    }
    if (tagPath.isIdentifier()) {
      if (matcher(tagPath.node.name)) {
        return true;
      }
      const binding = tagPath?.scope?.getBinding?.(tagPath.node.name);
      if (binding) {
        return resolvesToWeappTwIgnore(binding, /* @__PURE__ */ new Set());
      }
      return false;
    }
    if (tagPath.isMemberExpression()) {
      return resolvesMemberExpressionToIgnore(tagPath, /* @__PURE__ */ new Set());
    }
    return false;
  };
  const computeIgnore = (tagPath) => {
    const cached = taggedTemplateIgnoreCache.get(tagPath.node);
    if (cached !== void 0) {
      return cached;
    }
    const effectiveTagPath = getEffectiveTagPath(tagPath);
    const effectiveCached = taggedTemplateIgnoreCache.get(effectiveTagPath.node);
    if (effectiveCached !== void 0) {
      taggedTemplateIgnoreCache.set(tagPath.node, effectiveCached);
      return effectiveCached;
    }
    const result = evaluateTagPath(effectiveTagPath);
    taggedTemplateIgnoreCache.set(effectiveTagPath.node, result);
    taggedTemplateIgnoreCache.set(tagPath.node, result);
    return result;
  };
  return {
    shouldIgnore(tagPath) {
      return computeIgnore(tagPath);
    },
    getEffectiveTagPath
  };
}

// src/js/babel.ts
var EXPRESSION_WRAPPER_PREFIX = "(\n";
var EXPRESSION_WRAPPER_SUFFIX = "\n)";
function analyzeSource(ast, options, handler) {
  const jsTokenUpdater = new JsTokenUpdater();
  const ignoredPaths = /* @__PURE__ */ new WeakSet();
  const walker = new NodePathWalker(
    {
      ignoreCallExpressionIdentifiers: options.ignoreCallExpressionIdentifiers,
      callback(path5) {
        if (path5.isStringLiteral() || path5.isTemplateElement()) {
          ignoredPaths.add(path5);
        }
      }
    }
  );
  const isIgnoredTaggedTemplate = createNameMatcher(options.ignoreTaggedTemplateExpressionIdentifiers, { exact: true });
  const taggedTemplateIgnore = createTaggedTemplateIgnore({
    matcher: isIgnoredTaggedTemplate,
    names: options.ignoreTaggedTemplateExpressionIdentifiers
  });
  const needScope = Boolean(options.ignoreCallExpressionIdentifiers && options.ignoreCallExpressionIdentifiers.length > 0);
  const targetPaths = [];
  const importDeclarations = /* @__PURE__ */ new Set();
  const exportDeclarations = /* @__PURE__ */ new Set();
  const requireCallPaths = [];
  const evalHandler = handler ?? jsHandler;
  const traverseOptions = {
    StringLiteral: {
      enter(p) {
        if (isEvalPath(p.parentPath)) {
          return;
        }
        targetPaths.push(p);
      }
    },
    TemplateElement: {
      enter(p) {
        const pp = p.parentPath;
        if (pp.isTemplateLiteral()) {
          const ppp = pp.parentPath;
          if (isEvalPath(ppp)) {
            return;
          }
          if (ppp.isTaggedTemplateExpression()) {
            const tagPath = ppp.get("tag");
            if (taggedTemplateIgnore.shouldIgnore(tagPath)) {
              return;
            }
          }
        }
        targetPaths.push(p);
      }
    },
    CallExpression: {
      enter(p) {
        if (isEvalPath(p)) {
          walkEvalExpression(p, options, jsTokenUpdater, evalHandler);
          return;
        }
        const calleePath = p.get("callee");
        if (calleePath.isIdentifier({ name: "require" }) && !p?.scope?.hasBinding?.("require")) {
          const args = p.get("arguments");
          if (Array.isArray(args) && args.length > 0) {
            const first = args[0];
            if (first?.isStringLiteral()) {
              requireCallPaths.push(first);
            }
          }
        }
        if (needScope) {
          walker.walkCallExpression(p);
        }
      }
    },
    ImportDeclaration: {
      enter(p) {
        importDeclarations.add(p);
      }
    },
    ExportDeclaration: {
      enter(p) {
        exportDeclarations.add(p);
      }
    }
  };
  traverse(ast, { ...traverseOptions, noScope: !needScope });
  return {
    walker,
    jsTokenUpdater,
    ast,
    targetPaths,
    importDeclarations,
    exportDeclarations,
    requireCallPaths,
    ignoredPaths
  };
}
function jsHandler(rawSource, options) {
  const shouldWrapExpression = Boolean(options.wrapExpression);
  const source = shouldWrapExpression ? `${EXPRESSION_WRAPPER_PREFIX}${rawSource}${EXPRESSION_WRAPPER_SUFFIX}` : rawSource;
  let ast;
  try {
    ast = babelParse(source, options.babelParserOptions);
  } catch (error) {
    return {
      code: rawSource,
      error
    };
  }
  const analysis = analyzeSource(ast, options, jsHandler);
  const ms = processUpdatedSource(source, options, analysis);
  if (shouldWrapExpression) {
    const start = 0;
    const end = source.length;
    const prefixLength = EXPRESSION_WRAPPER_PREFIX.length;
    const suffixLength = EXPRESSION_WRAPPER_SUFFIX.length;
    ms.remove(start, start + prefixLength);
    ms.remove(end - suffixLength, end);
  }
  const result = {
    code: ms.toString(),
    get map() {
      return ms.generateMap();
    }
  };
  if (options.moduleGraph && options.filename) {
    const graph = new JsModuleGraph(
      {
        filename: options.filename,
        source: rawSource,
        analysis,
        handlerOptions: options
      },
      options.moduleGraph
    );
    const linked = graph.build();
    if (Object.keys(linked).length > 0) {
      result.linked = linked;
    }
  }
  return result;
}

// src/js/index.ts
function createJsHandler(options) {
  const {
    arbitraryValues,
    escapeMap,
    staleClassNameFallback,
    jsArbitraryValueFallback,
    tailwindcssMajorVersion,
    jsPreserveClass,
    generateMap,
    needEscaped,
    alwaysEscape,
    unescapeUnicode,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
    uniAppX,
    moduleSpecifierReplacements
  } = options;
  function handler(rawSource, classNameSet, options2) {
    const overrideOptions = options2 ?? {};
    const resolvedOptions = defuOverrideArray(
      {
        ...overrideOptions,
        classNameSet
      },
      {
        classNameSet,
        escapeMap,
        staleClassNameFallback,
        jsArbitraryValueFallback,
        tailwindcssMajorVersion,
        arbitraryValues,
        jsPreserveClass,
        generateMap,
        needEscaped,
        alwaysEscape,
        unescapeUnicode,
        babelParserOptions,
        ignoreCallExpressionIdentifiers,
        ignoreTaggedTemplateExpressionIdentifiers,
        uniAppX,
        moduleSpecifierReplacements
      }
    );
    return jsHandler(rawSource, resolvedOptions);
  }
  return handler;
}

// src/wxml/custom-attributes.ts
function regTest(reg, str) {
  reg.lastIndex = 0;
  return reg.test(str);
}
function isPropsMatch(props, attr) {
  if (Array.isArray(props)) {
    let lowerAttr;
    for (const prop of props) {
      if (typeof prop === "string") {
        lowerAttr ?? (lowerAttr = attr.toLowerCase());
        if (prop.toLowerCase() === lowerAttr) {
          return true;
        }
      } else if (regTest(prop, attr)) {
        return true;
      }
    }
    return false;
  } else if (typeof props === "string") {
    return props === attr;
  } else {
    return regTest(props, attr);
  }
}
function createAttributeMatcher(entities) {
  if (!entities || entities.length === 0) {
    return void 0;
  }
  const wildcardAttributeRules = [];
  const tagAttributeRuleMap = /* @__PURE__ */ new Map();
  const regexpAttributeRules = [];
  for (const [selector, props] of entities) {
    if (selector === "*") {
      wildcardAttributeRules.push(props);
    } else if (typeof selector === "string") {
      const list = tagAttributeRuleMap.get(selector);
      if (list) {
        list.push(props);
      } else {
        tagAttributeRuleMap.set(selector, [props]);
      }
    } else {
      regexpAttributeRules.push([selector, props]);
    }
  }
  return (tag, attr) => {
    for (const props of wildcardAttributeRules) {
      if (isPropsMatch(props, attr)) {
        return true;
      }
    }
    const tagRules = tagAttributeRuleMap.get(tag);
    if (tagRules) {
      for (const props of tagRules) {
        if (isPropsMatch(props, attr)) {
          return true;
        }
      }
    }
    for (const [selector, props] of regexpAttributeRules) {
      if (regTest(selector, tag) && isPropsMatch(props, attr)) {
        return true;
      }
    }
    return false;
  };
}

// src/wxml/utils/codegen/legacy-rewriter.ts
import MagicString2 from "magic-string";

// src/wxml/utils/codegen/legacy-visitor.ts
import * as t from "@babel/types";
function shouldSkipLegacyStringLiteral(path5) {
  if (t.isMemberExpression(path5.parent)) {
    return true;
  }
  return Boolean(
    t.isBinaryExpression(path5.parent) && (t.isConditionalExpression(path5.parentPath?.parent) || t.isLogicalExpression(path5.parentPath?.parent))
  );
}
function createLegacyTraverseOptions(options, jsTokenUpdater) {
  return {
    StringLiteral(path5) {
      if (shouldSkipLegacyStringLiteral(path5)) {
        return;
      }
      jsTokenUpdater.addToken(
        replaceHandleValue(
          path5,
          {
            escapeMap: options.escapeMap,
            classNameSet: options.runtimeSet,
            needEscaped: true,
            alwaysEscape: true
          }
        )
      );
    },
    noScope: true
  };
}

// src/wxml/utils/codegen/legacy-rewriter.ts
function rewriteLegacyExpression(match, options) {
  const ms = new MagicString2(match);
  const ast = parseExpression(match);
  const jsTokenUpdater = new JsTokenUpdater();
  traverse(ast, createLegacyTraverseOptions(options, jsTokenUpdater));
  jsTokenUpdater.updateMagicString(ms);
  return ms.toString();
}

// src/wxml/utils/codegen.ts
function generateCode(match, options = {}) {
  try {
    const { jsHandler: jsHandler2, runtimeSet } = options;
    if (jsHandler2 && runtimeSet) {
      const runHandler = (wrap) => jsHandler2(match, runtimeSet, wrap ? { wrapExpression: true } : void 0);
      const initial = runHandler(options.wrapExpression);
      if (!initial.error || options.wrapExpression) {
        return initial.code;
      }
      const fallback = runHandler(true);
      return fallback.code;
    } else {
      return rewriteLegacyExpression(match, options);
    }
  } catch {
    return match;
  }
}

// src/wxml/utils/custom-template.ts
import { Parser } from "htmlparser2";
import MagicString4 from "magic-string";

// src/wxml/utils/template-fragments.ts
import MagicString3 from "magic-string";

// src/wxml/whitespace.ts
var WHITESPACE_CODES = /* @__PURE__ */ new Set([
  9,
  // 制表符 \t
  10,
  // 换行符 \n
  11,
  // 垂直制表符 \v
  12,
  // 换页符 \f
  13,
  // 回车符 \r
  32,
  // 空格
  160,
  // 不间断空格 \u00A0
  65279
  // 零宽无断行空格 \uFEFF
]);
function isWhitespace(char) {
  if (char.length === 0) {
    return false;
  }
  return WHITESPACE_CODES.has(char.charCodeAt(0));
}
function isAllWhitespace(value) {
  for (let i = 0; i < value.length; i++) {
    if (!WHITESPACE_CODES.has(value.charCodeAt(i))) {
      return false;
    }
  }
  return true;
}

// src/wxml/Tokenizer.ts
var Tokenizer = class {
  constructor() {
    this.reset();
  }
  processChar(char, index) {
    switch (this.state) {
      case 0 /* START */:
        if (isWhitespace(char)) {
        } else if (char === "{") {
          this.state = 2 /* OPEN_BRACE */;
          this.bufferStartIndex = index;
          this.buffer += char;
          this.expressionBuffer = char;
          this.expressionStartIndex = index;
        } else {
          this.state = 1 /* TEXT */;
          this.bufferStartIndex = index;
          this.buffer += char;
        }
        break;
      case 1 /* TEXT */:
        if (isWhitespace(char)) {
          this.tokens.push({ start: this.bufferStartIndex, end: index, value: this.buffer, expressions: this.expressions });
          this.buffer = "";
          this.expressions = [];
          this.state = 0 /* START */;
        } else if (char === "{") {
          this.buffer += char;
          this.expressionBuffer = char;
          this.expressionStartIndex = index;
          this.state = 2 /* OPEN_BRACE */;
        } else {
          this.buffer += char;
        }
        break;
      case 2 /* OPEN_BRACE */:
        if (char === "}") {
          this.buffer += char;
          this.expressionBuffer += char;
          this.state = 3 /* POTENTIAL_CLOSE */;
        } else {
          this.buffer += char;
          this.expressionBuffer += char;
        }
        break;
      case 3 /* POTENTIAL_CLOSE */:
        if (char === "}") {
          this.buffer += char;
          this.expressionBuffer += char;
          this.expressions.push({
            start: this.expressionStartIndex,
            end: index + 1,
            value: this.expressionBuffer
          });
          this.expressionBuffer = "";
          this.state = 4 /* BRACES_COMPLETE */;
        } else {
          this.buffer += char;
          this.expressionBuffer += char;
          this.state = 2 /* OPEN_BRACE */;
        }
        break;
      case 4 /* BRACES_COMPLETE */:
        if (isWhitespace(char)) {
          this.tokens.push({
            start: this.bufferStartIndex,
            end: index,
            value: this.buffer,
            expressions: this.expressions
          });
          this.buffer = "";
          this.expressions = [];
          this.state = 0 /* START */;
        } else if (char === "{") {
          this.expressionStartIndex = index;
          this.expressionBuffer = char;
          this.buffer += char;
          this.state = 2 /* OPEN_BRACE */;
        } else {
          this.buffer += char;
          this.state = 1 /* TEXT */;
        }
        break;
      default:
        throw new Error("Unexpected state");
    }
  }
  run(input) {
    this.reset();
    for (let i = 0; i < input.length; i++) {
      const char = input[i];
      this.processChar(char, i);
    }
    if (this.buffer.length > 0) {
      this.tokens.push({
        start: this.bufferStartIndex,
        end: input.length,
        value: this.buffer,
        expressions: this.expressions
      });
    }
    const tokens = this.tokens;
    this.reset();
    return tokens;
  }
  reset() {
    this.state = 0 /* START */;
    this.buffer = "";
    this.tokens = [];
    this.bufferStartIndex = 0;
    this.expressionBuffer = "";
    this.expressionStartIndex = 0;
    this.expressions = [];
  }
};

// src/wxml/utils/fragment-helpers.ts
function updateWhitespaceGap(ms, start, end, options) {
  const gap = ms.slice(start, end);
  if (isAllWhitespace(gap)) {
    ms.update(start, end, replaceWxml(gap, {
      keepEOL: false,
      escapeMap: options.escapeMap,
      ignoreHead: true
    }));
  }
}
function updateWxmlSegment(ms, start, end, options, keepEOL, ignoreHead) {
  ms.update(start, end, replaceWxml(ms.slice(start, end), {
    keepEOL,
    escapeMap: options.escapeMap,
    ignoreHead
  }));
}
function updateExpressionSegment(ms, exp, options) {
  const code = `{{${generateCode(exp.value.slice(2, -2), options)}}}`;
  ms.update(exp.start, exp.end, code);
}

// src/wxml/utils/fragment-updater.ts
function handleEachClassFragment(ms, tokens, options = {}) {
  let previousEnd = 0;
  for (const token of tokens) {
    if (token.start > previousEnd) {
      updateWhitespaceGap(ms, previousEnd, token.start, options);
    }
    let p = token.start;
    if (token.expressions.length > 0) {
      for (const exp of token.expressions) {
        if (exp.start > token.start && p < exp.start) {
          updateWxmlSegment(ms, p, exp.start, options, true, p > 0);
        }
        updateExpressionSegment(ms, exp, options);
        p = exp.end;
      }
      if (token.end > p) {
        updateWxmlSegment(ms, p, token.end, options, false, true);
      }
    } else {
      updateWxmlSegment(ms, token.start, token.end, options, false, false);
    }
    previousEnd = token.end;
  }
  if (tokens.length > 0) {
    const lastToken = tokens[tokens.length - 1];
    if (lastToken.end < ms.original.length) {
      updateWhitespaceGap(ms, lastToken.end, ms.original.length, options);
    }
  }
}

// src/wxml/utils/template-fragments.ts
function templateReplacer(original, options = {}) {
  const ms = new MagicString3(original);
  const tokenizer = new Tokenizer();
  const tokens = tokenizer.run(ms.original);
  handleEachClassFragment(ms, tokens, options);
  return ms.toString();
}

// src/wxml/utils/custom-template.ts
async function customTemplateHandler(rawSource, options) {
  const {
    customAttributesEntities = [],
    disabledDefaultTemplateHandler,
    inlineWxs,
    runtimeSet,
    jsHandler: jsHandler2
  } = options ?? {};
  const matchCustomAttribute = createAttributeMatcher(customAttributesEntities);
  const s = new MagicString4(rawSource);
  let tag = "";
  const wxsArray = [];
  const parser = new Parser(
    {
      onopentagname(name) {
        tag = name;
      },
      onattribute(name, value, quote) {
        if (!value) {
          return;
        }
        const lowerName = name.toLowerCase();
        const shouldHandleDefault = !disabledDefaultTemplateHandler && (lowerName === "class" || lowerName === "hover-class" || lowerName === "virtualhostclass");
        const shouldHandleCustom = matchCustomAttribute?.(tag, name) ?? false;
        if (!shouldHandleDefault && !shouldHandleCustom) {
          return;
        }
        s.update(
          parser.startIndex + name.length + 2,
          // !important
          // htmlparser2 9.0.0: parser.endIndex
          // htmlparser2 9.1.0: parser.endIndex - 1
          // https://github.com/sonofmagic/weapp-tailwindcss/issues/269
          parser.endIndex - 1,
          templateReplacer(value, {
            ...options,
            quote
          })
        );
      },
      ontext(data) {
        if (inlineWxs && tag === "wxs") {
          wxsArray.push({
            data,
            endIndex: parser.endIndex + 1,
            startIndex: parser.startIndex
          });
        }
      },
      onclosetag() {
        tag = "";
      }
    },
    {
      xmlMode: true
    }
  );
  parser.write(s.original);
  parser.end();
  for (const { data, endIndex, startIndex } of wxsArray) {
    const { code } = await jsHandler2(data, runtimeSet);
    s.update(startIndex, endIndex, code);
  }
  return s.toString();
}

// src/wxml/utils.ts
function createTemplateHandler(options = {}) {
  return (rawSource, opt = {}) => {
    return customTemplateHandler(rawSource, defuOverrideArray(opt, options));
  };
}

// src/context/handlers.ts
function resolveRuntimePackageReplacements(option) {
  if (!option) {
    return void 0;
  }
  const mapping = option === true ? DEFAULT_RUNTIME_PACKAGE_REPLACEMENTS : option;
  const normalized = {};
  for (const [from, to] of Object.entries(mapping)) {
    if (!from || typeof to !== "string" || to.length === 0) {
      continue;
    }
    normalized[from] = to;
  }
  return Object.keys(normalized).length > 0 ? normalized : void 0;
}
function createHandlersFromContext(ctx, customAttributesEntities, cssCalcOptions, tailwindcssMajorVersion) {
  const {
    cssPreflight,
    cssPreflightRange,
    escapeMap,
    cssChildCombinatorReplaceValue,
    injectAdditionalCssVarScope,
    cssSelectorReplacement,
    rem2rpx,
    postcssOptions,
    cssRemoveProperty,
    cssRemoveHoverPseudoClass,
    cssPresetEnv,
    uniAppX,
    px2rpx,
    unitsToPx,
    arbitraryValues,
    jsPreserveClass,
    staleClassNameFallback,
    jsArbitraryValueFallback,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
    inlineWxs,
    disabledDefaultTemplateHandler,
    replaceRuntimePackages
  } = ctx;
  const moduleSpecifierReplacements = resolveRuntimePackageReplacements(replaceRuntimePackages);
  const styleHandler = createStyleHandler({
    cssPreflight,
    cssPreflightRange,
    escapeMap,
    cssChildCombinatorReplaceValue,
    injectAdditionalCssVarScope,
    cssSelectorReplacement,
    rem2rpx,
    postcssOptions,
    cssRemoveProperty,
    cssRemoveHoverPseudoClass,
    cssPresetEnv,
    uniAppX,
    cssCalc: cssCalcOptions,
    px2rpx,
    unitsToPx
  });
  const jsHandler2 = createJsHandler({
    escapeMap,
    arbitraryValues,
    jsPreserveClass,
    staleClassNameFallback,
    jsArbitraryValueFallback: jsArbitraryValueFallback ?? "auto",
    tailwindcssMajorVersion,
    generateMap: true,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
    uniAppX,
    moduleSpecifierReplacements
  });
  const templateHandler = createTemplateHandler({
    customAttributesEntities,
    escapeMap,
    inlineWxs,
    jsHandler: jsHandler2,
    disabledDefaultTemplateHandler
  });
  return {
    styleHandler,
    jsHandler: jsHandler2,
    templateHandler
  };
}

// src/context/logger.ts
import { logger as logger4 } from "@weapp-tailwindcss/logger";
var loggerLevelMap = {
  error: 0,
  warn: 1,
  info: 3,
  silent: -999
};
function applyLoggerLevel(logLevel) {
  logger4.level = loggerLevelMap[logLevel ?? "info"] ?? loggerLevelMap.info;
}

// src/context/index.ts
async function clearTailwindcssPatcherCache(patcher, options) {
  if (!patcher) {
    return;
  }
  const cacheOptions = patcher.options?.cache;
  if (cacheOptions == null || typeof cacheOptions === "object" && cacheOptions.enabled === false) {
    return;
  }
  if (typeof patcher.clearCache === "function") {
    try {
      await patcher.clearCache({ scope: "all" });
    } catch (error) {
      logger5.debug("failed to clear tailwindcss patcher cache via clearCache(): %O", error);
    }
  }
  if (!options?.removeDirectory) {
    return;
  }
  const cachePaths = /* @__PURE__ */ new Map();
  const normalizedCacheOptions = typeof cacheOptions === "object" ? cacheOptions : void 0;
  if (normalizedCacheOptions?.path) {
    cachePaths.set(normalizedCacheOptions.path, false);
  }
  const privateCachePath = patcher?.cacheStore?.options?.path;
  if (privateCachePath) {
    cachePaths.set(privateCachePath, false);
  }
  if (options?.removeDirectory && normalizedCacheOptions?.dir) {
    cachePaths.set(normalizedCacheOptions.dir, true);
  }
  if (!cachePaths.size) {
    return;
  }
  for (const [cachePath, recursive] of cachePaths.entries()) {
    try {
      await rm(cachePath, { force: true, recursive });
    } catch (error) {
      const err = error;
      if (err?.code === "ENOENT") {
        continue;
      }
      logger5.debug("failed to clear tailwindcss patcher cache: %s %O", cachePath, err);
    }
  }
}
function createInternalCompilerContext(opts) {
  const ctx = defuOverrideArray(
    opts,
    getDefaultOptions(),
    {}
  );
  ctx.escapeMap = ctx.customReplaceDictionary;
  applyLoggerLevel(ctx.logLevel);
  const twPatcher = createTailwindcssPatcherFromContext(ctx);
  logTailwindcssTarget("runtime", twPatcher, ctx.tailwindcssBasedir);
  if (twPatcher.packageInfo?.version) {
    logger5.success(`\u5F53\u524D\u4F7F\u7528 ${pc.cyanBright("Tailwind CSS")} \u7248\u672C\u4E3A: ${pc.underline(pc.bold(pc.green(twPatcher.packageInfo.version)))}`);
  } else {
    logger5.warn(`${pc.cyanBright("Tailwind CSS")} \u672A\u5B89\u88C5\uFF0C\u5DF2\u8DF3\u8FC7\u7248\u672C\u68C0\u6D4B\u4E0E\u8865\u4E01\u5E94\u7528\u3002`);
  }
  warnMissingCssEntries(ctx, twPatcher);
  const cssCalcOptions = applyV4CssCalcDefaults(ctx.cssCalc, twPatcher);
  ctx.cssCalc = cssCalcOptions;
  const customAttributesEntities = toCustomAttributesEntities(ctx.customAttributes);
  const { styleHandler, jsHandler: jsHandler2, templateHandler } = createHandlersFromContext(
    ctx,
    customAttributesEntities,
    cssCalcOptions,
    twPatcher.majorVersion
  );
  ctx.styleHandler = styleHandler;
  ctx.jsHandler = jsHandler2;
  ctx.templateHandler = templateHandler;
  ctx.cache = initializeCache(ctx.cache);
  ctx.twPatcher = twPatcher;
  const refreshTailwindcssPatcher = async (options) => {
    const previousPatcher = ctx.twPatcher;
    if (options?.clearCache !== false) {
      await clearTailwindcssPatcherCache(previousPatcher);
    }
    invalidateRuntimeClassSet(previousPatcher);
    const nextPatcher = createTailwindcssPatcherFromContext(ctx);
    Object.assign(previousPatcher, nextPatcher);
    ctx.twPatcher = previousPatcher;
    return previousPatcher;
  };
  ctx.refreshTailwindcssPatcher = refreshTailwindcssPatcher;
  Object.defineProperty(ctx.twPatcher, refreshTailwindcssPatcherSymbol, {
    value: refreshTailwindcssPatcher,
    configurable: true
  });
  return ctx;
}
function getCompilerContext(opts) {
  return withCompilerContextCache(opts, () => createInternalCompilerContext(opts));
}

export {
  createDebug,
  getRuntimeClassSetSignature,
  createTailwindPatchPromise,
  refreshTailwindRuntimeState,
  ensureRuntimeClassSet,
  collectRuntimeClassSet,
  pluginName,
  vitePluginName,
  WEAPP_TW_REQUIRED_NODE_VERSION,
  createPatchTargetRecorder,
  logTailwindcssTarget,
  toCustomAttributesEntities,
  replaceWxml,
  createAttributeMatcher,
  generateCode,
  clearTailwindcssPatcherCache,
  getCompilerContext
};
