// weapp-tailwindcss@4.10.3/dist/core.mjs
import {
  setupPatchRecorder
} from "./chunk-LICQ6EGN.mjs";
import {
  ensureRuntimeClassSet,
  getCompilerContext
} from "./chunk-E7775SFS.mjs";
import "./chunk-QOTLDKI4.mjs";
import "./chunk-3WUHHFLF.mjs";
import "./chunk-OOHJLO5M.mjs";
import "./chunk-TNYEOBAC.mjs";

// src/core.ts
import process from "process";
import { defuOverrideArray } from "@weapp-tailwindcss/shared";
function createContext(options = {}) {
  const opts = getCompilerContext(options);
  const { templateHandler, styleHandler, jsHandler, twPatcher: initialTwPatcher, refreshTailwindcssPatcher } = opts;
  const patchRecorderState = setupPatchRecorder(initialTwPatcher, opts.tailwindcssBasedir, {
    source: "runtime",
    cwd: opts.tailwindcssBasedir ?? process.cwd()
  });
  let runtimeSet = /* @__PURE__ */ new Set();
  const runtimeState = {
    twPatcher: initialTwPatcher,
    patchPromise: patchRecorderState.patchPromise,
    refreshTailwindcssPatcher,
    onPatchCompleted: patchRecorderState.onPatchCompleted
  };
  function withRuntimeTailwindMajorVersion(options2) {
    const resolvedOptions = {
      ...options2 ?? {}
    };
    if (typeof resolvedOptions.tailwindcssMajorVersion === "number") {
      return resolvedOptions;
    }
    const majorVersion = runtimeState.twPatcher.majorVersion;
    if (typeof majorVersion === "number") {
      resolvedOptions.tailwindcssMajorVersion = majorVersion;
    }
    return resolvedOptions;
  }
  async function transformWxss(rawCss, options2) {
    await runtimeState.patchPromise;
    const result = await styleHandler(rawCss, defuOverrideArray(options2, {
      isMainChunk: true
    }));
    runtimeSet = await ensureRuntimeClassSet(runtimeState, {
      forceRefresh: true,
      forceCollect: true
    });
    return result;
  }
  async function transformJs(rawJs, options2 = {}) {
    await runtimeState.patchPromise;
    if (options2?.runtimeSet) {
      runtimeSet = options2.runtimeSet;
    } else if (runtimeSet.size === 0) {
      runtimeSet = await ensureRuntimeClassSet(runtimeState, {
        forceCollect: true
      });
    }
    return await jsHandler(rawJs, runtimeSet, withRuntimeTailwindMajorVersion(options2));
  }
  async function transformWxml(rawWxml, options2) {
    await runtimeState.patchPromise;
    if (!options2?.runtimeSet && runtimeSet.size === 0) {
      runtimeSet = await ensureRuntimeClassSet(runtimeState, {
        forceCollect: true
      });
    }
    const runtimeJsHandler = jsHandler;
    return templateHandler(rawWxml, defuOverrideArray(options2, {
      runtimeSet,
      jsHandler: (source, runtime, handlerOptions) => {
        return runtimeJsHandler(source, runtime, withRuntimeTailwindMajorVersion(handlerOptions));
      }
    }));
  }
  return {
    transformWxss,
    transformWxml,
    transformJs
  };
}
export {
  createContext
};
