// weapp-tailwindcss@4.10.3/dist/core.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }

var _chunkGCRL3ZYPjs = require('./chunk-GCRL3ZYP.js');



var _chunkDEIJXHGJjs = require('./chunk-DEIJXHGJ.js');
require('./chunk-RYC23C3K.js');
require('./chunk-5U24PLVV.js');
require('./chunk-DYLQ6UOI.js');
require('./chunk-YUTKX7JZ.js');

// src/core.ts
var _process = require('process'); var _process2 = _interopRequireDefault(_process);
var _shared = require('@weapp-tailwindcss/shared');
function createContext(options = {}) {
  const opts = _chunkDEIJXHGJjs.getCompilerContext.call(void 0, options);
  const { templateHandler, styleHandler, jsHandler, twPatcher: initialTwPatcher, refreshTailwindcssPatcher } = opts;
  const patchRecorderState = _chunkGCRL3ZYPjs.setupPatchRecorder.call(void 0, initialTwPatcher, opts.tailwindcssBasedir, {
    source: "runtime",
    cwd: _nullishCoalesce(opts.tailwindcssBasedir, () => ( _process2.default.cwd()))
  });
  let runtimeSet = /* @__PURE__ */ new Set();
  const runtimeState = {
    twPatcher: initialTwPatcher,
    patchPromise: patchRecorderState.patchPromise,
    refreshTailwindcssPatcher,
    onPatchCompleted: patchRecorderState.onPatchCompleted
  };
  function withRuntimeTailwindMajorVersion(options2) {
    const resolvedOptions = {
      ..._nullishCoalesce(options2, () => ( {}))
    };
    if (typeof resolvedOptions.tailwindcssMajorVersion === "number") {
      return resolvedOptions;
    }
    const majorVersion = runtimeState.twPatcher.majorVersion;
    if (typeof majorVersion === "number") {
      resolvedOptions.tailwindcssMajorVersion = majorVersion;
    }
    return resolvedOptions;
  }
  async function transformWxss(rawCss, options2) {
    await runtimeState.patchPromise;
    const result = await styleHandler(rawCss, _shared.defuOverrideArray.call(void 0, options2, {
      isMainChunk: true
    }));
    runtimeSet = await _chunkDEIJXHGJjs.ensureRuntimeClassSet.call(void 0, runtimeState, {
      forceRefresh: true,
      forceCollect: true
    });
    return result;
  }
  async function transformJs(rawJs, options2 = {}) {
    await runtimeState.patchPromise;
    if (_optionalChain([options2, 'optionalAccess', _ => _.runtimeSet])) {
      runtimeSet = options2.runtimeSet;
    } else if (runtimeSet.size === 0) {
      runtimeSet = await _chunkDEIJXHGJjs.ensureRuntimeClassSet.call(void 0, runtimeState, {
        forceCollect: true
      });
    }
    return await jsHandler(rawJs, runtimeSet, withRuntimeTailwindMajorVersion(options2));
  }
  async function transformWxml(rawWxml, options2) {
    await runtimeState.patchPromise;
    if (!_optionalChain([options2, 'optionalAccess', _2 => _2.runtimeSet]) && runtimeSet.size === 0) {
      runtimeSet = await _chunkDEIJXHGJjs.ensureRuntimeClassSet.call(void 0, runtimeState, {
        forceCollect: true
      });
    }
    const runtimeJsHandler = jsHandler;
    return templateHandler(rawWxml, _shared.defuOverrideArray.call(void 0, options2, {
      runtimeSet,
      jsHandler: (source, runtime, handlerOptions) => {
        return runtimeJsHandler(source, runtime, withRuntimeTailwindMajorVersion(handlerOptions));
      }
    }));
  }
  return {
    transformWxss,
    transformWxml,
    transformJs
  };
}


exports.createContext = createContext;
