// weapp-tailwindcss@4.10.3/dist/escape.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/escape.ts
var escape_exports = {};
__export(escape_exports, {
  escape: () => replaceWxml,
  isAllowedClassName: () => import_escape2.isAllowedClassName,
  unescape: () => import_escape2.unescape,
  weappTwIgnore: () => weappTwIgnore
});
module.exports = __toCommonJS(escape_exports);

// src/wxml/shared.ts
var import_escape = require("@weapp-core/escape");
function replaceWxml(original, options = {
  keepEOL: false,
  escapeMap: import_escape.MappingChars2String
}) {
  const { keepEOL, escapeMap, ignoreHead } = options;
  let res = original;
  if (!keepEOL) {
    res = res.replaceAll(/[\n\r]+/g, "");
  }
  res = (0, import_escape.escape)(res, {
    map: escapeMap,
    ignoreHead
  });
  return res;
}

// src/escape.ts
var import_escape2 = require("@weapp-core/escape");
var weappTwIgnore = String.raw;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  escape,
  isAllowedClassName,
  unescape,
  weappTwIgnore
});
