// weapp-tailwindcss@4.10.3/dist/presets.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } }


var _chunkRYC23C3Kjs = require('./chunk-RYC23C3K.js');



var _chunkDYLQ6UOIjs = require('./chunk-DYLQ6UOI.js');
require('./chunk-YUTKX7JZ.js');

// src/presets/shared.ts
function normalizeCssEntries(entries) {
  if (!entries) {
    return void 0;
  }
  const normalized = Array.isArray(entries) ? entries : [entries];
  return normalized.length > 0 ? normalized : void 0;
}
function createBasePreset(options = {}) {
  const {
    base,
    cssEntries,
    resolve,
    rawOptions,
    ...userOptions
  } = options;
  const baseDir = _chunkRYC23C3Kjs.resolveTailwindcssBasedir.call(void 0, base);
  const normalizedCssEntries = normalizeCssEntries(cssEntries);
  const tailwindConfig = {
    v2: { cwd: baseDir },
    v3: { cwd: baseDir },
    v4: {
      base: baseDir,
      cssEntries: normalizedCssEntries
    }
  };
  if (normalizedCssEntries && normalizedCssEntries.length > 0) {
    tailwindConfig.version = 4;
  }
  const patchTailwindConfig = resolve ? {
    ...tailwindConfig,
    resolve
  } : tailwindConfig;
  const preset = {
    tailwindcssBasedir: baseDir,
    tailwindcss: tailwindConfig,
    tailwindcssPatcherOptions: {
      projectRoot: baseDir,
      tailwindcss: patchTailwindConfig
    }
  };
  const mergedUserOptions = _chunkDYLQ6UOIjs.defuOverrideArray.call(void 0, userOptions, _nullishCoalesce(rawOptions, () => ( {})));
  return _chunkDYLQ6UOIjs.defuOverrideArray.call(void 0, 
    mergedUserOptions,
    preset
  );
}

// src/presets/uni-app.ts
var _process = require('process'); var _process2 = _interopRequireDefault(_process);
function uniApp(options = {}) {
  const uniPlatform = _process2.default.env.UNI_PLATFORM;
  const disableInH5OrApp = uniPlatform === "h5" || uniPlatform === "app" || uniPlatform === "app-plus";
  return createBasePreset({
    disabled: disableInH5OrApp ? true : void 0,
    ...options
  });
}

// src/presets/uni-app-x.ts

function uniAppX(options) {
  _chunkRYC23C3Kjs.logger.info(`UNI_PLATFORM: ${_process2.default.env.UNI_PLATFORM}`);
  const utsPlatform = _chunkDYLQ6UOIjs.resolveUniUtsPlatform.call(void 0, );
  const uniPlatform = _chunkDYLQ6UOIjs.resolveUniUtsPlatform.call(void 0, _process2.default.env.UNI_PLATFORM);
  _chunkRYC23C3Kjs.logger.info(`UNI_UTS_PLATFORM: ${_nullishCoalesce(utsPlatform.raw, () => ( "undefined"))}`);
  const isApp = utsPlatform.isApp || uniPlatform.isApp;
  const cssEntries = normalizeCssEntries(options.cssEntries);
  return _chunkDYLQ6UOIjs.defuOverrideArray.call(void 0, 
    _nullishCoalesce(options.rawOptions, () => ( {})),
    {
      uniAppX: isApp,
      rem2rpx: options.rem2rpx,
      unitsToPx: options.unitsToPx,
      tailwindcssBasedir: options.base,
      tailwindcssPatcherOptions: {
        projectRoot: options.base,
        tailwindcss: {
          cwd: options.base,
          resolve: options.resolve,
          v3: {
            cwd: options.base
          },
          v4: {
            base: options.base,
            cssEntries
          }
        }
      },
      cssPreflight: {
        "border-style": false
      },
      cssPresetEnv: {
        features: {
          "custom-properties": {
            preserve: false
          }
        }
      },
      ...options.customAttributes ? { customAttributes: options.customAttributes } : {}
    }
  );
}





exports.createBasePreset = createBasePreset; exports.normalizeCssEntries = normalizeCssEntries; exports.uniApp = uniApp; exports.uniAppX = uniAppX;
