// weapp-tailwindcss@4.10.3/dist/presets.mjs
import {
  logger,
  resolveTailwindcssBasedir
} from "./chunk-QOTLDKI4.mjs";
import {
  defuOverrideArray,
  resolveUniUtsPlatform
} from "./chunk-OOHJLO5M.mjs";
import "./chunk-TNYEOBAC.mjs";

// src/presets/shared.ts
function normalizeCssEntries(entries) {
  if (!entries) {
    return void 0;
  }
  const normalized = Array.isArray(entries) ? entries : [entries];
  return normalized.length > 0 ? normalized : void 0;
}
function createBasePreset(options = {}) {
  const {
    base,
    cssEntries,
    resolve,
    rawOptions,
    ...userOptions
  } = options;
  const baseDir = resolveTailwindcssBasedir(base);
  const normalizedCssEntries = normalizeCssEntries(cssEntries);
  const tailwindConfig = {
    v2: { cwd: baseDir },
    v3: { cwd: baseDir },
    v4: {
      base: baseDir,
      cssEntries: normalizedCssEntries
    }
  };
  if (normalizedCssEntries && normalizedCssEntries.length > 0) {
    tailwindConfig.version = 4;
  }
  const patchTailwindConfig = resolve ? {
    ...tailwindConfig,
    resolve
  } : tailwindConfig;
  const preset = {
    tailwindcssBasedir: baseDir,
    tailwindcss: tailwindConfig,
    tailwindcssPatcherOptions: {
      projectRoot: baseDir,
      tailwindcss: patchTailwindConfig
    }
  };
  const mergedUserOptions = defuOverrideArray(userOptions, rawOptions ?? {});
  return defuOverrideArray(
    mergedUserOptions,
    preset
  );
}

// src/presets/uni-app.ts
import process from "process";
function uniApp(options = {}) {
  const uniPlatform = process.env.UNI_PLATFORM;
  const disableInH5OrApp = uniPlatform === "h5" || uniPlatform === "app" || uniPlatform === "app-plus";
  return createBasePreset({
    disabled: disableInH5OrApp ? true : void 0,
    ...options
  });
}

// src/presets/uni-app-x.ts
import process2 from "process";
function uniAppX(options) {
  logger.info(`UNI_PLATFORM: ${process2.env.UNI_PLATFORM}`);
  const utsPlatform = resolveUniUtsPlatform();
  const uniPlatform = resolveUniUtsPlatform(process2.env.UNI_PLATFORM);
  logger.info(`UNI_UTS_PLATFORM: ${utsPlatform.raw ?? "undefined"}`);
  const isApp = utsPlatform.isApp || uniPlatform.isApp;
  const cssEntries = normalizeCssEntries(options.cssEntries);
  return defuOverrideArray(
    options.rawOptions ?? {},
    {
      uniAppX: isApp,
      rem2rpx: options.rem2rpx,
      unitsToPx: options.unitsToPx,
      tailwindcssBasedir: options.base,
      tailwindcssPatcherOptions: {
        projectRoot: options.base,
        tailwindcss: {
          cwd: options.base,
          resolve: options.resolve,
          v3: {
            cwd: options.base
          },
          v4: {
            base: options.base,
            cssEntries
          }
        }
      },
      cssPreflight: {
        "border-style": false
      },
      cssPresetEnv: {
        features: {
          "custom-properties": {
            preserve: false
          }
        }
      },
      ...options.customAttributes ? { customAttributes: options.customAttributes } : {}
    }
  );
}
export {
  createBasePreset,
  normalizeCssEntries,
  uniApp,
  uniAppX
};
