// weapp-tailwindcss@4.10.3/dist/webpack4.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }











var _chunkWF636Q5Ejs = require('./chunk-WF636Q5E.js');






var _chunk6Z4GEN2Yjs = require('./chunk-6Z4GEN2Y.js');


var _chunkLTJQUORKjs = require('./chunk-LTJQUORK.js');


var _chunkGCRL3ZYPjs = require('./chunk-GCRL3ZYP.js');






var _chunkDEIJXHGJjs = require('./chunk-DEIJXHGJ.js');
require('./chunk-RYC23C3K.js');
require('./chunk-5U24PLVV.js');


var _chunkDYLQ6UOIjs = require('./chunk-DYLQ6UOI.js');
require('./chunk-YUTKX7JZ.js');

// src/bundlers/webpack/BaseUnifiedPlugin/v4.ts
var _process = require('process'); var _process2 = _interopRequireDefault(_process);

// src/bundlers/webpack/BaseUnifiedPlugin/v4-assets.ts
var _path = require('path'); var _path2 = _interopRequireDefault(_path);

var _webpacksources = require('webpack-sources');
function resolveWebpackStaleClassNameFallback(option, _compiler) {
  if (typeof option === "boolean") {
    return option;
  }
  return false;
}
function setupWebpackV4EmitHook(options) {
  const {
    compiler,
    options: compilerOptions,
    appType,
    runtimeState,
    debug: debug2
  } = options;
  compiler.hooks.emit.tapPromise(_chunkDEIJXHGJjs.pluginName, async (compilation) => {
    await runtimeState.patchPromise;
    compilerOptions.onStart();
    debug2("start");
    for (const chunk of compilation.chunks) {
      if (chunk.id && chunk.hash) {
        compilerOptions.cache.calcHashValueChanged(chunk.id, chunk.hash);
      }
    }
    const assetHashByChunk = _chunkWF636Q5Ejs.createAssetHashByChunkMap.call(void 0, compilation.chunks);
    const assets = compilation.assets;
    const entries = Object.entries(assets);
    const outputDir = _optionalChain([compiler, 'access', _ => _.options, 'optionalAccess', _2 => _2.output, 'optionalAccess', _3 => _3.path]) ? _path2.default.resolve(compiler.options.output.path) : _process2.default.cwd();
    const jsAssets = /* @__PURE__ */ new Map();
    for (const [file] of entries) {
      if (compilerOptions.jsMatcher(file) || compilerOptions.wxsMatcher(file)) {
        const absolute = _chunk6Z4GEN2Yjs.toAbsoluteOutputPath.call(void 0, file, outputDir);
        jsAssets.set(absolute, file);
      }
    }
    const moduleGraphOptions = {
      resolve(specifier, importer) {
        return _chunk6Z4GEN2Yjs.resolveOutputSpecifier.call(void 0, specifier, importer, outputDir, (candidate) => jsAssets.has(candidate));
      },
      load: (id) => {
        const assetName = jsAssets.get(id);
        if (!assetName) {
          return void 0;
        }
        const assetSource = compilation.assets[assetName];
        if (!assetSource) {
          return void 0;
        }
        const source = assetSource.source();
        return typeof source === "string" ? source : source.toString();
      },
      filter(id) {
        return jsAssets.has(id);
      }
    };
    const applyLinkedResults = (linked) => {
      if (!linked) {
        return;
      }
      for (const [id, { code }] of Object.entries(linked)) {
        const assetName = jsAssets.get(id);
        if (!assetName) {
          continue;
        }
        const assetSource = compilation.assets[assetName];
        if (!assetSource) {
          continue;
        }
        const previousSource = assetSource.source();
        const previous = typeof previousSource === "string" ? previousSource : previousSource.toString();
        if (previous === code) {
          continue;
        }
        const source = new (0, _webpacksources.ConcatSource)(code);
        compilation.updateAsset(assetName, source);
        compilerOptions.onUpdate(assetName, previous, code);
        debug2("js linked handle: %s", assetName);
      }
    };
    const groupedEntries = _chunkDYLQ6UOIjs.getGroupedEntries.call(void 0, entries, compilerOptions);
    const staleClassNameFallback = resolveWebpackStaleClassNameFallback(compilerOptions.staleClassNameFallback, compiler);
    const runtimeSet = await _chunkDEIJXHGJjs.ensureRuntimeClassSet.call(void 0, runtimeState, {
      // webpack 的 script-only 热更新可能不会触发 runtime classset loader，
      // 这里强制收集可避免沿用上轮 class set，保证 JS 仅按最新集合精确命中。
      forceCollect: true,
      allowEmpty: false
    });
    debug2("get runtimeSet, class count: %d", runtimeSet.size);
    const tasks = [];
    if (Array.isArray(groupedEntries.html)) {
      for (const element of groupedEntries.html) {
        const [file, originalSource] = element;
        const rawSource = originalSource.source().toString();
        const cacheKey = file;
        const chunkHash = assetHashByChunk.get(file);
        tasks.push(
          _chunkLTJQUORKjs.processCachedTask.call(void 0, {
            cache: compilerOptions.cache,
            cacheKey,
            hashKey: `${file}:asset`,
            rawSource,
            hash: chunkHash,
            applyResult(source) {
              compilation.updateAsset(file, source);
            },
            onCacheHit() {
              debug2("html cache hit: %s", file);
            },
            transform: async () => {
              const wxml = await compilerOptions.templateHandler(rawSource, {
                runtimeSet
              });
              const source = new (0, _webpacksources.ConcatSource)(wxml);
              compilerOptions.onUpdate(file, rawSource, wxml);
              debug2("html handle: %s", file);
              return {
                result: source
              };
            }
          })
        );
      }
    }
    const jsTaskFactories = [];
    if (Array.isArray(groupedEntries.js)) {
      for (const [file] of groupedEntries.js) {
        const cacheKey = _chunkWF636Q5Ejs.getCacheKey.call(void 0, file);
        const assetSource = compilation.assets[file];
        if (!assetSource) {
          continue;
        }
        const initialValue = assetSource.source();
        const initialRawSource = typeof initialValue === "string" ? initialValue : initialValue.toString();
        const absoluteFile = _chunk6Z4GEN2Yjs.toAbsoluteOutputPath.call(void 0, file, outputDir);
        const chunkHash = assetHashByChunk.get(file);
        jsTaskFactories.push(async () => {
          await _chunkLTJQUORKjs.processCachedTask.call(void 0, {
            cache: compilerOptions.cache,
            cacheKey,
            hashKey: `${file}:asset`,
            rawSource: initialRawSource,
            hash: chunkHash,
            applyResult(source) {
              compilation.updateAsset(file, source);
            },
            onCacheHit() {
              debug2("js cache hit: %s", file);
            },
            transform: async () => {
              const currentAsset = compilation.assets[file];
              const currentValue = _optionalChain([currentAsset, 'optionalAccess', _4 => _4.source, 'call', _5 => _5()]);
              const currentSource = typeof currentValue === "string" ? currentValue : _nullishCoalesce(_optionalChain([currentValue, 'optionalAccess', _6 => _6.toString, 'call', _7 => _7()]), () => ( ""));
              const { code, linked } = await compilerOptions.jsHandler(currentSource, runtimeSet, {
                staleClassNameFallback,
                tailwindcssMajorVersion: runtimeState.twPatcher.majorVersion,
                filename: absoluteFile,
                moduleGraph: moduleGraphOptions,
                babelParserOptions: {
                  sourceFilename: absoluteFile
                }
              });
              const source = new (0, _webpacksources.ConcatSource)(code);
              compilerOptions.onUpdate(file, currentSource, code);
              debug2("js handle: %s", file);
              applyLinkedResults(linked);
              return {
                result: source
              };
            }
          });
        });
      }
    }
    if (Array.isArray(groupedEntries.css)) {
      for (const element of groupedEntries.css) {
        const [file, originalSource] = element;
        const rawSource = originalSource.source().toString();
        const cacheKey = file;
        const chunkHash = assetHashByChunk.get(file);
        tasks.push(
          _chunkLTJQUORKjs.processCachedTask.call(void 0, {
            cache: compilerOptions.cache,
            cacheKey,
            hashKey: `${file}:asset`,
            rawSource,
            hash: chunkHash,
            applyResult(source) {
              compilation.updateAsset(file, source);
            },
            onCacheHit() {
              debug2("css cache hit: %s", file);
            },
            transform: async () => {
              await runtimeState.patchPromise;
              const { css } = await compilerOptions.styleHandler(rawSource, {
                isMainChunk: compilerOptions.mainCssChunkMatcher(file, appType),
                postcssOptions: {
                  options: {
                    from: file
                  }
                },
                majorVersion: runtimeState.twPatcher.majorVersion
              });
              const source = new (0, _webpacksources.ConcatSource)(css);
              compilerOptions.onUpdate(file, rawSource, css);
              debug2("css handle: %s", file);
              return {
                result: source
              };
            }
          })
        );
      }
    }
    _chunk6Z4GEN2Yjs.pushConcurrentTaskFactories.call(void 0, tasks, jsTaskFactories);
    await Promise.all(tasks);
    debug2("end");
    compilerOptions.onEnd();
  });
}

// src/bundlers/webpack/BaseUnifiedPlugin/v4-loaders.ts
var _fs = require('fs'); var _fs2 = _interopRequireDefault(_fs);


function setupWebpackV4Loaders(options) {
  const {
    compiler,
    options: compilerOptions,
    appType,
    weappTailwindcssPackageDir: weappTailwindcssPackageDir2,
    shouldRewriteCssImports,
    runtimeLoaderPath,
    runtimeCssImportRewriteLoaderPath,
    getClassSetInLoader,
    debug: debug2
  } = options;
  const isMpxApp = _chunkWF636Q5Ejs.isMpx.call(void 0, appType);
  if (shouldRewriteCssImports && isMpxApp) {
    _chunkWF636Q5Ejs.ensureMpxTailwindcssAliases.call(void 0, compiler, weappTailwindcssPackageDir2);
  }
  const runtimeClassSetLoader = _nullishCoalesce(runtimeLoaderPath, () => ( _path2.default.resolve(__dirname, "./weapp-tw-runtime-classset-loader.js")));
  const runtimeCssImportRewriteLoader = shouldRewriteCssImports ? _nullishCoalesce(runtimeCssImportRewriteLoaderPath, () => ( _path2.default.resolve(__dirname, "./weapp-tw-css-import-rewrite-loader.js"))) : void 0;
  const runtimeClassSetLoaderExists = _fs2.default.existsSync(runtimeClassSetLoader);
  const runtimeCssImportRewriteLoaderExists = runtimeCssImportRewriteLoader ? _fs2.default.existsSync(runtimeCssImportRewriteLoader) : false;
  const runtimeLoaderRewriteOptions = shouldRewriteCssImports ? {
    pkgDir: weappTailwindcssPackageDir2,
    appType
  } : void 0;
  const classSetLoaderOptions = {
    getClassSet: getClassSetInLoader
  };
  const { findRewriteAnchor, findClassSetAnchor } = _chunkWF636Q5Ejs.createLoaderAnchorFinders.call(void 0, appType);
  const cssImportRewriteLoaderOptions = runtimeLoaderRewriteOptions ? {
    rewriteCssImports: runtimeLoaderRewriteOptions
  } : void 0;
  if (runtimeCssImportRewriteLoader && shouldRewriteCssImports && cssImportRewriteLoaderOptions && isMpxApp) {
    _chunkWF636Q5Ejs.injectMpxCssRewritePreRules.call(void 0, compiler, runtimeCssImportRewriteLoader, cssImportRewriteLoaderOptions);
  }
  const createRuntimeClassSetLoaderEntry = () => ({
    loader: runtimeClassSetLoader,
    options: classSetLoaderOptions,
    ident: null,
    type: null
  });
  const createCssImportRewriteLoaderEntry = () => {
    if (!runtimeCssImportRewriteLoader) {
      return null;
    }
    return {
      loader: runtimeCssImportRewriteLoader,
      options: cssImportRewriteLoaderOptions,
      ident: null,
      type: null
    };
  };
  compiler.hooks.compilation.tap(_chunkDEIJXHGJjs.pluginName, (compilation) => {
    compilation.hooks.normalModuleLoader.tap(_chunkDEIJXHGJjs.pluginName, (_loaderContext, module) => {
      const hasRuntimeLoader = runtimeClassSetLoaderExists || runtimeCssImportRewriteLoaderExists;
      if (!hasRuntimeLoader) {
        return;
      }
      if (shouldRewriteCssImports && _chunkWF636Q5Ejs.isMpx.call(void 0, appType) && typeof _loaderContext.resolve === "function") {
        _chunkWF636Q5Ejs.patchMpxLoaderResolve.call(void 0, _loaderContext, weappTailwindcssPackageDir2, true);
      }
      const loaderEntries = module.loaders || [];
      let rewriteAnchorIdx = findRewriteAnchor(loaderEntries);
      const classSetAnchorIdx = findClassSetAnchor(loaderEntries);
      const isCssModule = _chunkWF636Q5Ejs.isCssLikeModuleResource.call(void 0, module.resource, compilerOptions.cssMatcher, appType);
      if (_process2.default.env.WEAPP_TW_LOADER_DEBUG && isCssModule) {
        debug2("loader hook css module: %s loaders=%o anchors=%o", module.resource, loaderEntries.map((x) => x.loader), { rewriteAnchorIdx, classSetAnchorIdx });
      }
      if (_process2.default.env.WEAPP_TW_LOADER_DEBUG && typeof module.resource === "string" && module.resource.includes("app.css")) {
        debug2("app.css module loaders=%o anchors=%o", loaderEntries.map((x) => x.loader), { rewriteAnchorIdx, classSetAnchorIdx });
      } else if (_process2.default.env.WEAPP_TW_LOADER_DEBUG && typeof module.resource === "string" && module.resource.endsWith(".css")) {
        debug2("css module seen: %s loaders=%o anchors=%o", module.resource, loaderEntries.map((x) => x.loader), { rewriteAnchorIdx, classSetAnchorIdx });
      }
      if (rewriteAnchorIdx === -1 && classSetAnchorIdx === -1 && !isCssModule) {
        return;
      }
      const anchorlessInsert = (entry, position) => {
        if (position === "after") {
          loaderEntries.push(entry);
        } else {
          loaderEntries.unshift(entry);
        }
      };
      if (runtimeLoaderRewriteOptions && runtimeCssImportRewriteLoaderExists && cssImportRewriteLoaderOptions && runtimeCssImportRewriteLoader) {
        const existingIndex = loaderEntries.findIndex(
          (entry) => _optionalChain([entry, 'access', _8 => _8.loader, 'optionalAccess', _9 => _9.includes, 'optionalCall', _10 => _10(runtimeCssImportRewriteLoader)])
        );
        const rewriteEntry = existingIndex !== -1 ? loaderEntries.splice(existingIndex, 1)[0] : createCssImportRewriteLoaderEntry();
        if (rewriteEntry) {
          const anchorIndex = findRewriteAnchor(loaderEntries);
          if (anchorIndex === -1) {
            anchorlessInsert(rewriteEntry, "after");
          } else {
            loaderEntries.splice(anchorIndex + 1, 0, rewriteEntry);
          }
          rewriteAnchorIdx = findRewriteAnchor(loaderEntries);
        }
      }
      if (runtimeClassSetLoaderExists && !_chunkWF636Q5Ejs.hasLoaderEntry.call(void 0, loaderEntries, runtimeClassSetLoader)) {
        const anchorIndex = findClassSetAnchor(loaderEntries);
        if (anchorIndex === -1) {
          anchorlessInsert(createRuntimeClassSetLoaderEntry(), "before");
        } else {
          const insertIndex = anchorIndex === -1 ? rewriteAnchorIdx : anchorIndex;
          loaderEntries.splice(insertIndex, 0, createRuntimeClassSetLoaderEntry());
        }
      }
    });
  });
}

// src/bundlers/webpack/BaseUnifiedPlugin/v4.ts
var debug = _chunkDEIJXHGJjs.createDebug.call(void 0, );
var weappTailwindcssPackageDir = _chunk6Z4GEN2Yjs.resolvePackageDir.call(void 0, "weapp-tailwindcss");
var UnifiedWebpackPluginV4 = class {
  constructor(options = {}) {
    this.options = _chunkDEIJXHGJjs.getCompilerContext.call(void 0, options);
    this.appType = this.options.appType;
  }
  apply(compiler) {
    compiler.options = compiler.options || {};
    const {
      disabled,
      onLoad,
      runtimeLoaderPath,
      runtimeCssImportRewriteLoaderPath,
      twPatcher: initialTwPatcher,
      refreshTailwindcssPatcher
    } = this.options;
    const disabledOptions = _chunk6Z4GEN2Yjs.resolveDisabledOptions.call(void 0, disabled);
    const isTailwindcssV4 = (_nullishCoalesce(initialTwPatcher.majorVersion, () => ( 0))) >= 4;
    const shouldRewriteCssImports = isTailwindcssV4 && this.options.rewriteCssImports !== false && !disabledOptions.rewriteCssImports;
    const isMpxApp = _chunkWF636Q5Ejs.isMpx.call(void 0, this.appType);
    if (shouldRewriteCssImports) {
      _chunkWF636Q5Ejs.applyTailwindcssCssImportRewrite.call(void 0, compiler, {
        pkgDir: weappTailwindcssPackageDir,
        enabled: true,
        appType: this.appType
      });
      _chunkWF636Q5Ejs.setupMpxTailwindcssRedirect.call(void 0, weappTailwindcssPackageDir, isMpxApp);
    }
    if (disabledOptions.plugin) {
      return;
    }
    const patchRecorderState = _chunkGCRL3ZYPjs.setupPatchRecorder.call(void 0, initialTwPatcher, this.options.tailwindcssBasedir, {
      source: "runtime",
      cwd: _nullishCoalesce(this.options.tailwindcssBasedir, () => ( _process2.default.cwd()))
    });
    const runtimeState = {
      twPatcher: initialTwPatcher,
      patchPromise: patchRecorderState.patchPromise,
      refreshTailwindcssPatcher,
      onPatchCompleted: patchRecorderState.onPatchCompleted
    };
    let runtimeSetPrepared = false;
    let runtimeSetSignature;
    compiler.hooks.compilation.tap(_chunkDEIJXHGJjs.pluginName, () => {
      runtimeSetPrepared = false;
    });
    async function getClassSetInLoader() {
      if (runtimeSetPrepared) {
        return;
      }
      const signature = _chunkDEIJXHGJjs.getRuntimeClassSetSignature.call(void 0, runtimeState.twPatcher);
      const forceRefresh = signature !== runtimeSetSignature;
      runtimeSetPrepared = true;
      await _chunkDEIJXHGJjs.ensureRuntimeClassSet.call(void 0, runtimeState, {
        forceRefresh,
        forceCollect: true,
        clearCache: forceRefresh,
        allowEmpty: true
      });
      runtimeSetSignature = signature;
    }
    onLoad();
    setupWebpackV4Loaders({
      compiler,
      options: this.options,
      appType: this.appType,
      weappTailwindcssPackageDir,
      shouldRewriteCssImports,
      runtimeLoaderPath,
      runtimeCssImportRewriteLoaderPath,
      getClassSetInLoader,
      debug
    });
    setupWebpackV4EmitHook({
      compiler,
      options: this.options,
      appType: this.appType,
      runtimeState,
      debug
    });
  }
};



exports.UnifiedWebpackPluginV4 = UnifiedWebpackPluginV4; exports.weappTailwindcssPackageDir = weappTailwindcssPackageDir;
