// @weapp-core/escape@7.0.0/dist/index.mjs
const SYMBOL_TABLE = {
  BACKQUOTE: "`",
  TILDE: "~",
  EXCLAM: "!",
  AT: "@",
  NUMBERSIGN: "#",
  DOLLAR: "$",
  PERCENT: "%",
  CARET: "^",
  AMPERSAND: "&",
  ASTERISK: "*",
  PARENLEFT: "(",
  PARENRIGHT: ")",
  MINUS: "-",
  UNDERSCORE: "_",
  EQUAL: "=",
  PLUS: "+",
  BRACKETLEFT: "[",
  BRACELEFT: "{",
  BRACKETRIGHT: "]",
  BRACERIGHT: "}",
  SEMICOLON: ";",
  COLON: ":",
  QUOTE: "'",
  DOUBLEQUOTE: '"',
  BACKSLASH: "\\",
  BAR: "|",
  COMMA: ",",
  LESS: "<",
  PERIOD: ".",
  GREATER: ">",
  SLASH: "/",
  QUESTION: "?",
  SPACE: " ",
  DOT: ".",
  HASH: "#"
};
const ESCAPE_CHARACTER_CODES = [
  ["[", "b"],
  ["]", "B"],
  ["(", "p"],
  [")", "P"],
  ["#", "h"],
  ["!", "e"],
  ["/", "f"],
  ["\\", "r"],
  [".", "d"],
  [":", "c"],
  ["%", "v"],
  [",", "m"],
  ["'", "a"],
  ['"', "q"],
  ["*", "x"],
  ["&", "n"],
  ["@", "t"],
  ["{", "k"],
  ["}", "K"],
  ["+", "u"],
  [";", "j"],
  ["<", "l"],
  ["~", "w"],
  ["=", "z"],
  [">", "g"],
  ["?", "Q"],
  ["^", "y"],
  ["`", "i"],
  ["|", "o"],
  ["$", "s"]
];
function buildMapping(prefix) {
  return ESCAPE_CHARACTER_CODES.reduce((acc, [char, code]) => {
    acc[char] = `${prefix}${code}`;
    return acc;
  }, {});
}
const SIMPLE_TOKEN_PREFIX = "_";
const COMPLEX_TOKEN_PREFIX = "_c";
const ComplexMappingChars2String = buildMapping(COMPLEX_TOKEN_PREFIX);
const ComplexMappingChars2StringEntries = Object.entries(ComplexMappingChars2String);
const MappingChars2String = buildMapping(SIMPLE_TOKEN_PREFIX);
const MappingChars2StringEntries = Object.entries(MappingChars2String);
const MAX_ASCII_CHAR_CODE = 127;

const ESCAPE_PREFIX = "_";
const HYPHEN = "-";

const DEFAULT_ESCAPE_MAPPING = MappingChars2String;
const COMPLEX_ESCAPE_MAPPING = ComplexMappingChars2String;
new Set(Object.keys(DEFAULT_ESCAPE_MAPPING));
const escapeMappingCache = /* @__PURE__ */ new WeakMap();
const tokenBucketCache = /* @__PURE__ */ new WeakMap();
const inverseMappingCache = /* @__PURE__ */ new WeakMap();
function hasOwnKey(object, key) {
  return Object.prototype.hasOwnProperty.call(object, key);
}
function createEscapeMapping(customMap) {
  if (!customMap) {
    return DEFAULT_ESCAPE_MAPPING;
  }
  if (customMap === DEFAULT_ESCAPE_MAPPING || customMap === COMPLEX_ESCAPE_MAPPING) {
    return customMap;
  }
  const cached = escapeMappingCache.get(customMap);
  if (cached) {
    return cached;
  }
  if (Object.keys(customMap).length === 0) {
    escapeMappingCache.set(customMap, DEFAULT_ESCAPE_MAPPING);
    return DEFAULT_ESCAPE_MAPPING;
  }
  const merged = { ...DEFAULT_ESCAPE_MAPPING, ...customMap };
  escapeMappingCache.set(customMap, merged);
  return merged;
}
function createUnescapeMapping(customMap) {
  return customMap;
}
function buildInverseMapping(mapping) {
  const inverse = {};
  const tokens = /* @__PURE__ */ new Set();
  for (const [key, value] of Object.entries(mapping)) {
    inverse[value] = key;
    tokens.add(value);
  }
  const sortedTokens = Array.from(tokens).sort((a, b) => b.length - a.length);
  return { inverse, tokens: sortedTokens };
}
function createInverseMapping(mapping) {
  const cached = inverseMappingCache.get(mapping);
  if (cached) {
    return cached;
  }
  const built = buildInverseMapping(mapping);
  inverseMappingCache.set(mapping, built);
  return built;
}
function createTokenBuckets(mapping, tokens) {
  if (tokens.length === 0) {
    return void 0;
  }
  const cached = tokenBucketCache.get(mapping);
  if (cached) {
    return cached;
  }
  const buckets = tokens.reduce((acc, token) => {
    if (!token) {
      return acc;
    }
    const first = token[0];
    if (!first) {
      return acc;
    }
    const bucket = acc[first];
    if (bucket) {
      bucket.push(token);
    } else {
      acc[first] = [token];
    }
    return acc;
  }, {});
  tokenBucketCache.set(mapping, buckets);
  return buckets;
}
function primeInverseCache(mapping) {
  inverseMappingCache.set(mapping, buildInverseMapping(mapping));
}
primeInverseCache(DEFAULT_ESCAPE_MAPPING);
primeInverseCache(COMPLEX_ESCAPE_MAPPING);

const DIGIT_MIN_CODE = 48;
const DIGIT_MAX_CODE = 57;
function isAsciiNumber(code) {
  return code >= DIGIT_MIN_CODE && code <= DIGIT_MAX_CODE;
}
function isAllowedClassName(className) {
  return /^[\w-]+$/.test(className);
}

const HIGH_SURROGATE_MIN = 55296;
const HIGH_SURROGATE_MAX = 56319;
const LOW_SURROGATE_MIN = 56320;
const LOW_SURROGATE_MAX = 57343;
const DEFAULT_ESCAPE_TABLE = (() => {
  const table = Array.from({ length: MAX_ASCII_CHAR_CODE + 1 });
  for (const [char, token] of Object.entries(DEFAULT_ESCAPE_MAPPING)) {
    table[char.codePointAt(0)] = token;
  }
  return table;
})();
function readCodePoint(str, index) {
  const first = str.charCodeAt(index);
  if (first < HIGH_SURROGATE_MIN || first > HIGH_SURROGATE_MAX || index + 1 >= str.length) {
    return { codePoint: first, size: 1 };
  }
  const second = str.charCodeAt(index + 1);
  if (second < LOW_SURROGATE_MIN || second > LOW_SURROGATE_MAX) {
    return { codePoint: first, size: 1 };
  }
  return {
    codePoint: (first - HIGH_SURROGATE_MIN << 10) + (second - LOW_SURROGATE_MIN) + 65536,
    size: 2
  };
}
function escapeLeadingCharacter(char, nextChar, ignoreHead) {
  if (ignoreHead) {
    return char;
  }
  const code = char.codePointAt(0);
  if (isAsciiNumber(code)) {
    return `${ESCAPE_PREFIX}${char}`;
  }
  if (char === HYPHEN) {
    if (!nextChar) {
      return `${ESCAPE_PREFIX}${char}`;
    }
    if (isAsciiNumber(nextChar.codePointAt(0))) {
      return `${ESCAPE_PREFIX}${char}`;
    }
  }
  return char;
}
function escape(selectors, options) {
  const length = selectors.length;
  if (length === 0) {
    return "";
  }
  const map = createEscapeMapping(options?.map);
  const ignoreHead = options?.ignoreHead ?? false;
  const usingDefaultMap = map === DEFAULT_ESCAPE_MAPPING;
  let cursor = 0;
  let lastUnchangedIndex = 0;
  let buffer;
  while (cursor < length) {
    const { codePoint, size } = readCodePoint(selectors, cursor);
    const isHead = cursor === 0;
    let replacement;
    if (codePoint > MAX_ASCII_CHAR_CODE) {
      replacement = `u_x${codePoint.toString(16)}_`;
    } else if (usingDefaultMap) {
      if (isHead) {
        const mapped = DEFAULT_ESCAPE_TABLE[codePoint];
        if (mapped !== void 0) {
          replacement = mapped;
        } else {
          const nextIndex = cursor + size;
          const next = nextIndex < length ? readCodePoint(selectors, nextIndex) : void 0;
          const char = selectors[cursor];
          const nextChar = next ? selectors.slice(nextIndex, nextIndex + next.size) : void 0;
          const escapedHead = escapeLeadingCharacter(char, nextChar, ignoreHead);
          if (escapedHead !== char) {
            replacement = escapedHead;
          }
        }
      } else {
        const mapped = DEFAULT_ESCAPE_TABLE[codePoint];
        if (mapped !== void 0) {
          replacement = mapped;
        }
      }
    } else {
      const char = selectors[cursor];
      const mapped = hasOwnKey(map, char) ? map[char] : void 0;
      if (mapped !== void 0) {
        replacement = mapped;
      } else if (isHead) {
        const nextIndex = cursor + size;
        const next = nextIndex < length ? readCodePoint(selectors, nextIndex) : void 0;
        const nextChar = next ? selectors.slice(nextIndex, nextIndex + next.size) : void 0;
        const escapedHead = escapeLeadingCharacter(char, nextChar, ignoreHead);
        if (escapedHead !== char) {
          replacement = escapedHead;
        }
      }
    }
    if (replacement !== void 0) {
      if (!buffer) {
        buffer = [];
      }
      if (lastUnchangedIndex !== cursor) {
        buffer.push(selectors.slice(lastUnchangedIndex, cursor));
      }
      buffer.push(replacement);
      lastUnchangedIndex = cursor + size;
    }
    cursor += size;
  }
  if (!buffer) {
    return selectors;
  }
  if (lastUnchangedIndex < length) {
    buffer.push(selectors.slice(lastUnchangedIndex));
  }
  return buffer.join("");
}

function toEscapeOptions(options) {
  return options ?? {};
}
function toUnescapeOptions(options) {
  return options ?? {};
}

const UNICODE_PREFIX = "u_x";
const UNICODE_SUFFIX = "_";
const HEX_RADIX = 16;
function isHexDigit(char) {
  const code = char.codePointAt(0);
  return code >= 48 && code <= 57 || code >= 65 && code <= 70 || code >= 97 && code <= 102;
}
function decodeUnicodeSequence(value, index) {
  if (!value.startsWith(UNICODE_PREFIX, index)) {
    return void 0;
  }
  let cursor = index + UNICODE_PREFIX.length;
  while (cursor < value.length) {
    const nextChar = value[cursor];
    if (!nextChar || !isHexDigit(nextChar)) {
      break;
    }
    cursor += 1;
  }
  if (cursor === index + UNICODE_PREFIX.length) {
    return void 0;
  }
  if (cursor >= value.length || value[cursor] !== UNICODE_SUFFIX) {
    return void 0;
  }
  const hex = value.slice(index + UNICODE_PREFIX.length, cursor);
  const codePoint = Number.parseInt(hex, HEX_RADIX);
  if (Number.isNaN(codePoint) || codePoint <= MAX_ASCII_CHAR_CODE || codePoint > 1114111) {
    return void 0;
  }
  return {
    char: String.fromCodePoint(codePoint),
    length: cursor - index + 1
  };
}

const ESCAPE_PREFIX_CODE = ESCAPE_PREFIX.codePointAt(0);
const LOWER_U_CODE = "u".codePointAt(0);
const DEFAULT_UNESCAPE_TABLE = (() => {
  const inverse = {};
  for (const [char, token] of Object.entries(DEFAULT_ESCAPE_MAPPING)) {
    inverse[token] = char;
  }
  return inverse;
})();
function unescapeDefault(value, ignoreHead) {
  const length = value.length;
  const hasLeadingEscape = !ignoreHead && value[0] === ESCAPE_PREFIX;
  const hasUnicodeMarker = value.includes("u_x");
  if (!hasLeadingEscape && !hasUnicodeMarker) {
    return value;
  }
  let cursor = 0;
  let result = "";
  while (cursor < length) {
    const currentChar = value[cursor];
    const currentCode = value.codePointAt(cursor);
    if (currentCode === LOWER_U_CODE) {
      const decodedUnicode = decodeUnicodeSequence(value, cursor);
      if (decodedUnicode) {
        result += decodedUnicode.char;
        cursor += decodedUnicode.length;
        continue;
      }
    }
    result += currentChar;
    cursor++;
  }
  if (ignoreHead || result.length === 0 || !result.startsWith(ESCAPE_PREFIX)) {
    return result;
  }
  const nextChar = result[1];
  if (!nextChar) {
    return result.slice(1);
  }
  let shouldStrip = isAsciiNumber(nextChar.codePointAt(0));
  if (!shouldStrip && nextChar === HYPHEN) {
    const thirdChar = result[2];
    shouldStrip = !thirdChar || isAsciiNumber(thirdChar.codePointAt(0));
  }
  return shouldStrip ? result.slice(1) : result;
}
function unescape(value, options) {
  const length = value.length;
  if (length === 0) {
    return "";
  }
  const map = createUnescapeMapping(options?.map);
  const ignoreHead = options?.ignoreHead ?? false;
  if (!map) {
    return unescapeDefault(value, ignoreHead);
  }
  const usingDefaultMap = map === DEFAULT_ESCAPE_MAPPING;
  const { inverse, tokens } = usingDefaultMap ? { inverse: {}, tokens: [] } : createInverseMapping(map);
  const tokenBuckets = usingDefaultMap ? void 0 : createTokenBuckets(map, tokens);
  let cursor = 0;
  let lastUnchangedIndex = 0;
  let buffer;
  while (cursor < length) {
    const currentCode = value.charCodeAt(cursor);
    const currentChar = value[cursor];
    let replacement;
    let advance = 1;
    if (replacement === void 0 && currentCode === LOWER_U_CODE) {
      const decodedUnicode = decodeUnicodeSequence(value, cursor);
      if (decodedUnicode) {
        replacement = decodedUnicode.char;
        advance = decodedUnicode.length;
      }
    }
    if (replacement === void 0) {
      if (usingDefaultMap && currentCode === ESCAPE_PREFIX_CODE) {
        const mapped = DEFAULT_UNESCAPE_TABLE[value.slice(cursor, cursor + 2)];
        if (mapped !== void 0) {
          replacement = mapped;
          advance = 2;
        }
      } else if (tokenBuckets) {
        const bucket = tokenBuckets[currentChar];
        if (bucket) {
          for (const token of bucket) {
            if (value.startsWith(token, cursor)) {
              replacement = inverse[token];
              advance = token.length;
              break;
            }
          }
        }
      }
    }
    if (replacement !== void 0) {
      if (!buffer) {
        buffer = [];
      }
      if (lastUnchangedIndex !== cursor) {
        buffer.push(value.slice(lastUnchangedIndex, cursor));
      }
      buffer.push(replacement);
      lastUnchangedIndex = cursor + advance;
      cursor += advance;
      continue;
    }
    cursor += 1;
  }
  const result = buffer ? (() => {
    if (lastUnchangedIndex < length) {
      buffer.push(value.slice(lastUnchangedIndex));
    }
    return buffer.join("");
  })() : value;
  if (ignoreHead || result.length === 0 || !result.startsWith(ESCAPE_PREFIX)) {
    return result;
  }
  const nextChar = result[1];
  if (!nextChar) {
    return result.slice(1);
  }
  let shouldStrip = isAsciiNumber(nextChar.codePointAt(0));
  if (!shouldStrip && nextChar === HYPHEN) {
    const thirdChar = result[2];
    shouldStrip = !thirdChar || isAsciiNumber(thirdChar.codePointAt(0));
  }
  return shouldStrip ? result.slice(1) : result;
}

const __areExportsLoaded = true;

export { ComplexMappingChars2String, ComplexMappingChars2StringEntries, MAX_ASCII_CHAR_CODE, MappingChars2String, MappingChars2StringEntries, SYMBOL_TABLE, __areExportsLoaded, escape, isAllowedClassName, isAsciiNumber, toEscapeOptions, toUnescapeOptions, unescape };
