// weapp-tailwindcss@4.10.3/dist/css-macro/postcss.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } }



var _chunkSR4GC2F4js = require('../chunk-SR4GC2F4.js');
require('../chunk-YUTKX7JZ.js');

// src/css-macro/postcss.ts
var creator = () => {
  return {
    postcssPlugin: "postcss-weapp-tw-css-macro-plugin",
    prepare() {
      return {
        AtRule(atRule, helper) {
          if (atRule.name === "media") {
            const values = [];
            _chunkSR4GC2F4js.matchCustomPropertyFromValue.call(void 0, atRule.params, (arr) => {
              values.push(arr[1]);
            });
            if (values.length > 0) {
              const isNegative = atRule.params.includes("not");
              const text = values.join(" ");
              const comment = isNegative ? _chunkSR4GC2F4js.ifndef.call(void 0, text) : _chunkSR4GC2F4js.ifdef.call(void 0, text);
              const nextNodes = [
                helper.comment({
                  raws: {
                    before: "\n"
                  },
                  text: comment.start
                }),
                ...(_nullishCoalesce(atRule.nodes, () => ( []))).map((node) => node.clone()),
                helper.comment({
                  raws: {
                    before: "\n"
                  },
                  text: comment.end
                })
              ];
              atRule.replaceWith(nextNodes);
            }
          }
        },
        CommentExit(comment) {
          if (/#(?:ifn?def|endif)/.test(comment.text)) {
            comment.raws.left = " ";
            comment.raws.right = " ";
          }
        }
      };
    }
  };
};
creator.postcss = true;
var postcss_default = creator;


exports.default = postcss_default;

module.exports = exports.default;
