// weapp-tailwindcss@4.10.3/dist/css-macro/postcss.mjs
import {
  ifdef,
  ifndef,
  matchCustomPropertyFromValue
} from "../chunk-3XVUGMTY.mjs";
import "../chunk-TNYEOBAC.mjs";

// src/css-macro/postcss.ts
var creator = () => {
  return {
    postcssPlugin: "postcss-weapp-tw-css-macro-plugin",
    prepare() {
      return {
        AtRule(atRule, helper) {
          if (atRule.name === "media") {
            const values = [];
            matchCustomPropertyFromValue(atRule.params, (arr) => {
              values.push(arr[1]);
            });
            if (values.length > 0) {
              const isNegative = atRule.params.includes("not");
              const text = values.join(" ");
              const comment = isNegative ? ifndef(text) : ifdef(text);
              const nextNodes = [
                helper.comment({
                  raws: {
                    before: "\n"
                  },
                  text: comment.start
                }),
                ...(atRule.nodes ?? []).map((node) => node.clone()),
                helper.comment({
                  raws: {
                    before: "\n"
                  },
                  text: comment.end
                })
              ];
              atRule.replaceWith(nextNodes);
            }
          }
        },
        CommentExit(comment) {
          if (/#(?:ifn?def|endif)/.test(comment.text)) {
            comment.raws.left = " ";
            comment.raws.right = " ";
          }
        }
      };
    }
  };
};
creator.postcss = true;
var postcss_default = creator;
export {
  postcss_default as default
};
