// postcss-units-to-px@0.2.0/dist/index.cjs
const require_defaults = require('./defaults-CTsBq2Vl.cjs');
let postcss_plugin_shared = require("postcss-plugin-shared");

//#region package.json
var name = "postcss-units-to-px";

//#endregion
//#region src/shared.ts
/**
* PostCSS plugin name used by this package.
*
* @default package.json name
*/
const postcssPlugin = name;
/**
* Merge user options with defaults for postcss-units-to-px.
*
* Defaults:
* - unitMap: { rem: 16, em: 16, vw: 3.75, vh: 6.67, vmin: 3.75, vmax: 6.67, rpx: 0.5 }
* - unitPrecision: 5
* - propList: ['*']
* - replace: true
* - mediaQuery: false
* - minValue: 0
* - exclude: [/node_modules/i]
* - disabled: false
*
* @example
* const config = getConfig({ unitPrecision: 4 })
*/
const getConfig = (0, postcss_plugin_shared.createConfigGetter)(require_defaults.defaultOptions);

//#endregion
//#region src/plugin.ts
const DEFAULT_NUMBER_PATTERN = String.raw`\d+(?:\.\d+)?|\.\d+`;
function createAnyUnitRegex() {
	const parts = [
		String.raw`"[^"]+"`,
		String.raw`'[^']+'`,
		String.raw`url\([^)]+\)`,
		String.raw`var\([^)]+\)`,
		String.raw`(${DEFAULT_NUMBER_PATTERN})([a-zA-Z%]+)`
	];
	return new RegExp(parts.join("|"), "g");
}
function normalizeRule(rule) {
	return rule === void 0 ? null : rule;
}
function isPlainObject(value) {
	return Object.prototype.toString.call(value) === "[object Object]";
}
function normalizeMatchers(unitMap) {
	const entries = [];
	const addEntry = (matcher, rule) => {
		const normalizedRule = normalizeRule(rule);
		if (typeof matcher === "string") {
			const unit = matcher.trim().toLowerCase();
			if (!unit) return;
			entries.push({
				matcher: unit,
				rule: normalizedRule,
				type: "string",
				unit
			});
			return;
		}
		if (matcher instanceof RegExp) {
			entries.push({
				matcher,
				rule: normalizedRule,
				type: "regex"
			});
			return;
		}
		if (typeof matcher === "function") entries.push({
			matcher,
			rule: normalizedRule,
			type: "function"
		});
	};
	if (Array.isArray(unitMap)) for (const entry of unitMap) {
		if (!entry) continue;
		const [matcher, rule] = entry;
		addEntry(matcher, rule);
	}
	else if (unitMap instanceof Map) for (const [matcher, rule] of unitMap.entries()) addEntry(matcher, rule);
	else for (const [unit, rule] of Object.entries(unitMap)) addEntry(unit, rule);
	const hasComplexMatcher = entries.some((entry) => entry.type !== "string");
	if (hasComplexMatcher) return {
		entries,
		unitMap: void 0,
		hasComplexMatcher
	};
	const normalized = /* @__PURE__ */ new Map();
	for (const entry of entries) {
		if (entry.type !== "string") continue;
		if (!normalized.has(entry.unit)) normalized.set(entry.unit, entry.rule);
	}
	return {
		entries,
		unitMap: normalized,
		hasComplexMatcher
	};
}
function getUnitsForRegex(unitMap) {
	return Array.from(unitMap.keys()).sort((a, b) => b.length - a.length);
}
function createReplace(getRule, unitPrecision, minValue, transform, context) {
	const shouldRound = unitPrecision >= 0 && unitPrecision <= 100;
	return function replace(m, $1, $2) {
		if (!$1) return m;
		const value = Number($1);
		if (Number.isNaN(value)) return m;
		if (value < minValue) return m;
		const unit = (typeof $2 === "string" ? $2 : m.slice($1.length)).toLowerCase();
		const rule = getRule(unit);
		let pxValue;
		if (rule === void 0) return m;
		if (rule === false) return m;
		if (typeof rule === "function") pxValue = rule(value, context);
		else if (typeof rule === "number") pxValue = value * rule;
		else if (transform) pxValue = transform(value, unit, context);
		else return m;
		if (pxValue === void 0 || Number.isNaN(pxValue)) return m;
		const roundedValue = shouldRound ? (0, postcss_plugin_shared.toFixed)(pxValue, unitPrecision) : pxValue;
		if (roundedValue === 0) return "0";
		return `${roundedValue}px`;
	};
}
/**
* Convert multiple CSS units to px based on a unit map and optional transform.
*
* Defaults:
* - unitMap: { rem: 16, em: 16, vw: 3.75, vh: 6.67, vmin: 3.75, vmax: 6.67, rpx: 0.5 }
* - unitPrecision: 5
* - propList: ['*']
* - replace: true
* - mediaQuery: false
* - minValue: 0
* - exclude: [/node_modules/i]
* - disabled: false
*
* @example
* import postcss from 'postcss'
* import unitsToPx from 'postcss-units-to-px'
*
* const result = await postcss([unitsToPx({
*   unitMap: { rem: 16, vw: 3.75 },
*   transform: (value, unit) => (unit === 'em' ? value * 12 : value * 16),
* })]).process('.a{margin:1rem 1vw}', { from: undefined })
*/
const plugin = (options = {}) => {
	const { transform, unitPrecision, minValue, selectorBlackList, propList, replace, mediaQuery, exclude, disabled } = getConfig(options);
	const userUnitMap = options.unitMap;
	const mergedUnitMap = userUnitMap === void 0 ? require_defaults.defaultUnitMap : isPlainObject(userUnitMap) ? {
		...require_defaults.defaultUnitMap,
		...userUnitMap
	} : userUnitMap;
	if (disabled || transform === false) return { postcssPlugin };
	const { entries, unitMap: normalizedUnitMap, hasComplexMatcher } = normalizeMatchers(mergedUnitMap);
	if (entries.length === 0) return { postcssPlugin };
	let unitRegex;
	let getRule;
	if (hasComplexMatcher) {
		unitRegex = createAnyUnitRegex();
		getRule = (unit) => {
			for (const entry of entries) {
				if (entry.type === "string") {
					if (entry.unit === unit) return entry.rule;
					continue;
				}
				if (entry.type === "regex") {
					if (entry.matcher.global || entry.matcher.sticky) entry.matcher.lastIndex = 0;
					if (entry.matcher.test(unit)) return entry.rule;
					continue;
				}
				if (entry.type === "function") {
					if (entry.matcher(unit)) return entry.rule;
				}
			}
		};
	} else {
		const units = getUnitsForRegex(normalizedUnitMap ?? /* @__PURE__ */ new Map());
		if (units.length === 0) return { postcssPlugin };
		unitRegex = (0, postcss_plugin_shared.createUnitRegex)({ units });
		getRule = (unit) => normalizedUnitMap?.get(unit);
	}
	return {
		postcssPlugin,
		Once(css) {
			(0, postcss_plugin_shared.walkAndReplaceValues)({
				root: css,
				unitRegex,
				propList,
				selectorBlackList,
				exclude,
				replace,
				mediaQuery,
				createReplacer: (context) => {
					return createReplace(getRule, unitPrecision, minValue, transform, context);
				},
				shouldProcessAtRule: (atRule) => atRule.name === "media"
			});
		}
	};
};
plugin.postcss = true;
var plugin_default = plugin;

//#endregion
//#region src/index.ts
var src_default = plugin_default;

//#endregion
module.exports = src_default;
