// postcss-plugin-shared@1.1.1/dist/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }// src/numbers.ts
function toFixed(number, precision) {
  if (number === 0) {
    return 0;
  }
  const multiplier = 10 ** precision;
  const sign = Math.sign(number);
  const absNumber = Math.abs(number);
  const rounded = Math.round((absNumber + Number.EPSILON) * multiplier) / multiplier;
  return sign * rounded;
}

// src/option-merge.ts
var _defu = require('defu');
var defu = _defu.createDefu.call(void 0, (obj, key, value) => {
  if (Array.isArray(obj[key]) && Array.isArray(value)) {
    obj[key] = value;
    return true;
  }
});
function mergeOptions(options, defaults) {
  return defu(_nullishCoalesce(options, () => ( {})), defaults);
}
function createConfigGetter(defaults) {
  return function getConfig(options) {
    return mergeOptions(options, defaults);
  };
}

// src/selectors.ts
function blacklistedSelector(blacklist, selector) {
  if (typeof selector !== "string") {
    return false;
  }
  return blacklist.some((rule) => {
    if (typeof rule === "string") {
      return selector.includes(rule);
    }
    return Boolean(selector.match(rule));
  });
}
function maybeBlacklistedSelector(blacklist, selector) {
  if (typeof selector !== "string") {
    return void 0;
  }
  return blacklistedSelector(blacklist, selector);
}
function createSelectorBlacklistMatcher(blacklist, options = {}) {
  const useCache = _nullishCoalesce(options.cache, () => ( true));
  const cache = useCache ? /* @__PURE__ */ new WeakMap() : void 0;
  return function isBlacklisted(rule) {
    if (blacklist.length === 0) {
      return false;
    }
    if (cache) {
      const cached = cache.get(rule);
      if (cached !== void 0) {
        return cached;
      }
      const value = Boolean(maybeBlacklistedSelector(blacklist, rule.selector));
      cache.set(rule, value);
      return value;
    }
    return Boolean(maybeBlacklistedSelector(blacklist, rule.selector));
  };
}
function createPropListMatcher(propList) {
  const hasWild = propList.includes("*");
  return function satisfyPropList(prop) {
    if (hasWild) {
      return true;
    }
    return propList.some((rule) => {
      if (typeof rule === "string") {
        return prop.includes(rule);
      }
      return Boolean(prop.match(rule));
    });
  };
}
function filterPropList(list) {
  return {
    exact: list.filter((m) => /^[^*!]+$/.test(m)),
    contain: list.filter((m) => /^\*.+\*$/.test(m)).map((m) => m.substring(1, m.length - 1)),
    endWith: list.filter((m) => /^\*[^*]+$/.test(m)).map((m) => m.substring(1)),
    startWith: list.filter((m) => /^[^*!]+\*$/.test(m)).map((m) => m.substring(0, m.length - 1)),
    notExact: list.filter((m) => /^![^*].*$/.test(m)).map((m) => m.substring(1)),
    notContain: list.filter((m) => /^!\*.+\*$/.test(m)).map((m) => m.substring(2, m.length - 1)),
    notEndWith: list.filter((m) => /^!\*[^*]+$/.test(m)).map((m) => m.substring(2)),
    notStartWith: list.filter((m) => /^![^*]+\*$/.test(m)).map((m) => m.substring(1, m.length - 1))
  };
}
function createAdvancedPropListMatcher(propList) {
  const hasWild = propList.includes("*");
  const matchAll = hasWild && propList.length === 1;
  const lists = filterPropList(propList);
  return function satisfyPropList(prop) {
    if (matchAll) {
      return true;
    }
    const shouldInclude = hasWild || lists.exact.includes(prop) || lists.contain.some((m) => prop.includes(m)) || lists.startWith.some((m) => prop.startsWith(m)) || lists.endWith.some((m) => prop.endsWith(m));
    const shouldExclude = lists.notExact.includes(prop) || lists.notContain.some((m) => prop.includes(m)) || lists.notStartWith.some((m) => prop.startsWith(m)) || lists.notEndWith.some((m) => prop.endsWith(m));
    return shouldInclude && !shouldExclude;
  };
}
function createExcludeMatcher(exclude) {
  return function isExcluded(filepath) {
    if (filepath === void 0) {
      return false;
    }
    return Array.isArray(exclude) ? exclude.some((rule) => {
      if (typeof rule === "string") {
        return filepath.includes(rule);
      }
      return Boolean(filepath.match(rule));
    }) : exclude(filepath);
  };
}

// src/postcss.ts
function declarationExists(decls, prop, value) {
  return decls.some((node) => {
    if (node.type !== "decl") {
      return false;
    }
    const decl = node;
    return decl.prop === prop && decl.value === value;
  });
}
function walkAndReplaceValues(options) {
  const {
    root,
    unitRegex,
    propList,
    selectorBlackList = [],
    exclude = [],
    replace = true,
    skipDuplicate = true,
    mediaQuery = false,
    createReplacer,
    shouldProcessDecl,
    shouldProcessAtRule
  } = options;
  const source = root.source;
  const input = _optionalChain([source, 'optionalAccess', _ => _.input]);
  if (!input) {
    return;
  }
  const filePath = input.file;
  const excludeFn = createExcludeMatcher(exclude);
  if (filePath && excludeFn(filePath)) {
    return;
  }
  const satisfyPropList = createPropListMatcher(propList);
  const unitTestRegex = new RegExp(unitRegex.source, unitRegex.flags.replace("g", ""));
  const isBlacklisted = createSelectorBlacklistMatcher(selectorBlackList, { cache: true });
  const baseContext = {
    root,
    input,
    ...filePath === void 0 ? {} : { filePath }
  };
  root.walkDecls((decl) => {
    if (!satisfyPropList(decl.prop)) {
      return;
    }
    if (shouldProcessDecl && !shouldProcessDecl(decl)) {
      return;
    }
    if (!unitTestRegex.test(decl.value)) {
      return;
    }
    const rule = decl.parent;
    if (selectorBlackList.length > 0 && isBlacklisted(rule)) {
      return;
    }
    const atRule = _optionalChain([rule, 'access', _2 => _2.parent, 'optionalAccess', _3 => _3.type]) === "atrule" ? rule.parent : void 0;
    const context = {
      ...baseContext,
      decl,
      rule,
      prop: decl.prop,
      selector: rule.selector,
      ...atRule ? { atRule } : {}
    };
    const replacer = createReplacer(context);
    const nextValue = decl.value.replace(unitRegex, replacer);
    if (nextValue === decl.value) {
      return;
    }
    if (skipDuplicate && declarationExists(rule, decl.prop, nextValue)) {
      return;
    }
    if (replace) {
      decl.value = nextValue;
      return;
    }
    decl.cloneAfter({ value: nextValue });
  });
  if (!mediaQuery) {
    return;
  }
  const shouldProcess = _nullishCoalesce(shouldProcessAtRule, () => ( ((atRule) => atRule.name === "media")));
  root.walkAtRules((atRule) => {
    if (!shouldProcess(atRule)) {
      return;
    }
    if (!unitTestRegex.test(atRule.params)) {
      return;
    }
    const context = {
      ...baseContext,
      atRule
    };
    const replacer = createReplacer(context);
    const nextParams = atRule.params.replace(unitRegex, replacer);
    if (nextParams !== atRule.params) {
      atRule.params = nextParams;
    }
  });
}

// src/regex.ts
function createUnitRegex(options) {
  const {
    units,
    numberPattern = String.raw`\d+(?:\.\d+)?|\.\d+`,
    skipDoubleQuotes = true,
    skipSingleQuotes = true,
    skipUrl = true,
    skipVar = true,
    ignoreCase = false
  } = options;
  const parts = [];
  if (skipDoubleQuotes) {
    parts.push(String.raw`"[^"]+"`);
  }
  if (skipSingleQuotes) {
    parts.push(String.raw`'[^']+'`);
  }
  if (skipUrl) {
    parts.push(String.raw`url\([^)]+\)`);
  }
  if (skipVar) {
    parts.push(String.raw`var\([^)]+\)`);
  }
  const unitPart = units.map((u) => u.replace(/[\\^$.*+?()[\]{}|]/g, String.raw`\$&`)).join("|");
  parts.push(String.raw`(${numberPattern})(?:${unitPart})`);
  return new RegExp(parts.join("|"), `g${ignoreCase ? "i" : ""}`);
}
var remRegex = createUnitRegex({ units: ["rem"] });
var pxRegex = createUnitRegex({ units: ["px"] });















exports.blacklistedSelector = blacklistedSelector; exports.createAdvancedPropListMatcher = createAdvancedPropListMatcher; exports.createConfigGetter = createConfigGetter; exports.createExcludeMatcher = createExcludeMatcher; exports.createPropListMatcher = createPropListMatcher; exports.createSelectorBlacklistMatcher = createSelectorBlacklistMatcher; exports.createUnitRegex = createUnitRegex; exports.declarationExists = declarationExists; exports.maybeBlacklistedSelector = maybeBlacklistedSelector; exports.mergeOptions = mergeOptions; exports.pxRegex = pxRegex; exports.remRegex = remRegex; exports.toFixed = toFixed; exports.walkAndReplaceValues = walkAndReplaceValues;
