// miaoda-taro-utils@0.0.4/dist/chatStream/index.mjs
import ky from "ky";
import { createParser } from "eventsource-parser";
import Taro from "@tarojs/taro";
const EventStreamContentType = "text/event-stream";
const DEFAULT_TIMEOUT = 3e5;
const SSE_DATA_SEPARATOR = "\\ ";
function isH5Environment() {
  try {
    return Taro.getEnv() === Taro.ENV_TYPE.WEB;
  } catch {
    return typeof window !== "undefined" && typeof window.fetch === "function";
  }
}
function decodeUtf8(bytes) {
  try {
    const charStr = Array.from(bytes, (byte) => String.fromCharCode(byte)).join(
      ""
    );
    return decodeURIComponent(escape(charStr));
  } catch (e) {
    console.warn("UTF-8 decode error:", e);
    return Array.from(bytes, (byte) => String.fromCharCode(byte)).join("");
  }
}
function createH5SSE(options) {
  const {
    url,
    headers = {},
    data,
    timeout,
    onData,
    onEvent,
    onCompleted,
    onAborted
  } = options;
  const controller = new AbortController();
  const createSSEHook = (sseOptions) => async (request, _options, response) => {
    if (!response.ok || !response.body) {
      return;
    }
    let completed = false;
    const innerOnCompleted = (error) => {
      var _a;
      if (completed)
        return;
      completed = true;
      (_a = sseOptions.onCompleted) == null ? void 0 : _a.call(sseOptions, error);
    };
    const reader = response.body.getReader();
    const decoder = new TextDecoder("utf8");
    const parser = createParser({
      onEvent: (event) => {
        var _a;
        if (event.data) {
          (_a = sseOptions.onEvent) == null ? void 0 : _a.call(sseOptions, event);
          event.data.split(SSE_DATA_SEPARATOR).forEach(sseOptions.onData);
        }
      }
    });
    const read = () => {
      reader.read().then((result) => {
        if (result.done) {
          innerOnCompleted();
          return;
        }
        parser.feed(decoder.decode(result.value, { stream: true }));
        read();
      }).catch((error) => {
        var _a;
        if (request.signal.aborted) {
          (_a = sseOptions.onAborted) == null ? void 0 : _a.call(sseOptions);
          return;
        }
        innerOnCompleted(error);
      });
    };
    read();
    return response;
  };
  const sseHook = createSSEHook({ onData, onEvent, onCompleted, onAborted });
  ky.post(url, {
    json: data,
    headers,
    signal: controller.signal,
    timeout,
    hooks: {
      afterResponse: [sseHook]
    }
  }).catch((error) => {
    if (!controller.signal.aborted) {
      onCompleted == null ? void 0 : onCompleted(error);
    }
  });
  return {
    abort: () => controller.abort()
  };
}
function createWechatSSE(options) {
  const {
    url,
    headers = {},
    data,
    timeout,
    onData,
    onEvent,
    onCompleted,
    onAborted
  } = options;
  let requestTask = null;
  let completed = false;
  let aborted = false;
  const innerOnCompleted = (error) => {
    if (completed)
      return;
    completed = true;
    onCompleted == null ? void 0 : onCompleted(error);
  };
  try {
    requestTask = Taro.request({
      url,
      method: "POST",
      header: headers,
      data,
      enableChunked: true,
      responseType: "text",
      timeout,
      success: () => !aborted && innerOnCompleted(),
      fail: (err) => {
        if (!aborted) {
          innerOnCompleted(new Error(err.errMsg || "Request failed"));
        }
      }
    });
    const parser = createParser({
      onEvent: (event) => {
        if (aborted || !event.data)
          return;
        onEvent == null ? void 0 : onEvent(event);
        event.data.split(SSE_DATA_SEPARATOR).forEach(onData);
      }
    });
    requestTask.onChunkReceived((res) => {
      if (aborted)
        return;
      parser.feed(decodeUtf8(new Uint8Array(res.data)));
    });
  } catch (err) {
    innerOnCompleted(err);
  }
  return {
    abort: () => {
      if (requestTask && !aborted) {
        aborted = true;
        try {
          requestTask.abort();
          onAborted == null ? void 0 : onAborted();
        } catch (e) {
          console.error("Abort request error:", e);
        }
        requestTask = null;
      }
    }
  };
}
function createChatDataHandler(onUpdate) {
  let currentContent = "";
  return (data) => {
    var _a, _b, _c;
    try {
      const parsed = JSON.parse(data);
      if ((_c = (_b = (_a = parsed.choices) == null ? void 0 : _a[0]) == null ? void 0 : _b.delta) == null ? void 0 : _c.content) {
        currentContent += parsed.choices[0].delta.content;
        onUpdate(currentContent);
      }
    } catch {
      console.warn("Failed to parse SSE data:", data);
    }
  };
}
const sendChatStream = (options) => {
  const {
    messages,
    onUpdate,
    onComplete,
    onError,
    timeout = DEFAULT_TIMEOUT
  } = options;
  const handleCompleted = (error) => error ? onError == null ? void 0 : onError(error) : onComplete == null ? void 0 : onComplete();
  const handleAborted = () => {
    console.log("Stream aborted");
    onComplete == null ? void 0 : onComplete();
  };
  const requestData = {
    messages: messages.map((msg) => ({
      role: msg.role,
      content: msg.content
    })),
    enable_thinking: false
  };
  const requestHeaders = {
    "X-App-Id": options.appId,
    "Content-Type": "application/json"
  };
  const commonOptions = {
    url: options.endpoint,
    headers: requestHeaders,
    data: requestData,
    timeout,
    onData: onUpdate,
    onCompleted: handleCompleted,
    onAborted: handleAborted
  };
  const sseInstance = isH5Environment() ? createH5SSE(commonOptions) : createWechatSSE(commonOptions);
  return { abort: () => sseInstance.abort() };
};
export {
  EventStreamContentType,
  createChatDataHandler,
  createH5SSE,
  createWechatSSE,
  sendChatStream
};
