// postcss-rem-to-responsive-pixel@7.0.1/dist/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }// src/regex.ts
var _postcsspluginshared = require('postcss-plugin-shared');

// src/shared.ts













// src/defaults.ts
var defaultOptions = {
  rootValue: 16,
  unitPrecision: 5,
  selectorBlackList: [],
  propList: ["font", "font-size", "line-height", "letter-spacing"],
  replace: true,
  mediaQuery: false,
  minRemValue: 0,
  exclude: [/node_modules/i],
  transformUnit: "px",
  disabled: false,
  processorStage: "Once"
};

// src/shared.ts
var postcssPlugin = "postcss-rem-to-responsive-pixel";
var getConfig = _postcsspluginshared.createConfigGetter.call(void 0, defaultOptions);
function createRemReplace(rootValue, unitPrecision, minRemValue, transformUnit = "px") {
  return function(m, $1) {
    if (!$1) {
      return m;
    }
    const rems = Number.parseFloat($1);
    if (rems < minRemValue) {
      return m;
    }
    const fixedVal = _postcsspluginshared.toFixed.call(void 0, rems * rootValue, unitPrecision);
    return fixedVal === 0 ? "0" : fixedVal + transformUnit;
  };
}

// src/index.ts
var plugin = (options = {}) => {
  const {
    exclude,
    mediaQuery,
    minRemValue,
    propList,
    replace,
    rootValue,
    selectorBlackList,
    transformUnit,
    unitPrecision,
    disabled,
    processorStage
  } = getConfig(options);
  if (disabled) {
    return {
      postcssPlugin
    };
  }
  const isRootValueFn = typeof rootValue === "function";
  const staticReplace = isRootValueFn ? void 0 : createRemReplace(rootValue, unitPrecision, minRemValue, transformUnit);
  return {
    postcssPlugin,
    [processorStage]: (css) => {
      const source = css.source;
      const input = _optionalChain([source, 'optionalAccess', _ => _.input]);
      if (!input) {
        return;
      }
      const pxReplace = isRootValueFn ? createRemReplace(
        rootValue(input),
        unitPrecision,
        minRemValue,
        transformUnit
      ) : staticReplace;
      _postcsspluginshared.walkAndReplaceValues.call(void 0, {
        root: css,
        unitRegex: _postcsspluginshared.remRegex,
        propList,
        selectorBlackList,
        exclude,
        replace,
        mediaQuery,
        createReplacer: () => pxReplace,
        shouldProcessDecl: (decl) => decl.value.includes("rem"),
        shouldProcessAtRule: (atRule) => atRule.name === "media" && atRule.params.includes("rem")
      });
    }
  };
};
plugin.postcss = true;
var index_default = plugin;


exports.default = index_default;

module.exports = exports.default;
