// postcss-rem-to-responsive-pixel@7.0.1/dist/index.mjs
// src/regex.ts
import { pxRegex, remRegex } from "postcss-plugin-shared";

// src/shared.ts
import {
  createConfigGetter,
  createExcludeMatcher,
  createPropListMatcher,
  createSelectorBlacklistMatcher,
  declarationExists,
  maybeBlacklistedSelector,
  pxRegex as pxRegex2,
  remRegex as remRegex2,
  toFixed,
  walkAndReplaceValues
} from "postcss-plugin-shared";

// src/defaults.ts
var defaultOptions = {
  rootValue: 16,
  unitPrecision: 5,
  selectorBlackList: [],
  propList: ["font", "font-size", "line-height", "letter-spacing"],
  replace: true,
  mediaQuery: false,
  minRemValue: 0,
  exclude: [/node_modules/i],
  transformUnit: "px",
  disabled: false,
  processorStage: "Once"
};

// src/shared.ts
var postcssPlugin = "postcss-rem-to-responsive-pixel";
var getConfig = createConfigGetter(defaultOptions);
function createRemReplace(rootValue, unitPrecision, minRemValue, transformUnit = "px") {
  return function(m, $1) {
    if (!$1) {
      return m;
    }
    const rems = Number.parseFloat($1);
    if (rems < minRemValue) {
      return m;
    }
    const fixedVal = toFixed(rems * rootValue, unitPrecision);
    return fixedVal === 0 ? "0" : fixedVal + transformUnit;
  };
}

// src/index.ts
var plugin = (options = {}) => {
  const {
    exclude,
    mediaQuery,
    minRemValue,
    propList,
    replace,
    rootValue,
    selectorBlackList,
    transformUnit,
    unitPrecision,
    disabled,
    processorStage
  } = getConfig(options);
  if (disabled) {
    return {
      postcssPlugin
    };
  }
  const isRootValueFn = typeof rootValue === "function";
  const staticReplace = isRootValueFn ? void 0 : createRemReplace(rootValue, unitPrecision, minRemValue, transformUnit);
  return {
    postcssPlugin,
    [processorStage]: (css) => {
      const source = css.source;
      const input = source?.input;
      if (!input) {
        return;
      }
      const pxReplace = isRootValueFn ? createRemReplace(
        rootValue(input),
        unitPrecision,
        minRemValue,
        transformUnit
      ) : staticReplace;
      walkAndReplaceValues({
        root: css,
        unitRegex: remRegex,
        propList,
        selectorBlackList,
        exclude,
        replace,
        mediaQuery,
        createReplacer: () => pxReplace,
        shouldProcessDecl: (decl) => decl.value.includes("rem"),
        shouldProcessAtRule: (atRule) => atRule.name === "media" && atRule.params.includes("rem")
      });
    }
  };
};
plugin.postcss = true;
var index_default = plugin;
export {
  index_default as default
};
